<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>比较转录组分析（五）—— 转录本定量与差异表达分析 | Juse's Blog</title><meta name="keywords" content="生物信息学,生信,转录组,比较转录组学,差异表达分析"><meta name="author" content="Juse"><meta name="copyright" content="Juse"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="使用 Trinity 包装的 Pipeline 进行转录本定量与差异表达分析。">
<meta property="og:type" content="article">
<meta property="og:title" content="比较转录组分析（五）—— 转录本定量与差异表达分析">
<meta property="og:url" content="https://biojuse.com/2022/12/11/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%BA%94%EF%BC%89%E2%80%94%E2%80%94%20%E8%BD%AC%E5%BD%95%E6%9C%AC%E5%AE%9A%E9%87%8F%E4%B8%8E%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/index.html">
<meta property="og:site_name" content="Juse's Blog">
<meta property="og:description" content="使用 Trinity 包装的 Pipeline 进行转录本定量与差异表达分析。">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://biojuse.com/pic/cover2.jpg">
<meta property="article:published_time" content="2022-12-11T07:25:00.000Z">
<meta property="article:modified_time" content="2024-08-22T12:44:00.132Z">
<meta property="article:author" content="Juse">
<meta property="article:tag" content="生物信息学">
<meta property="article:tag" content="生信">
<meta property="article:tag" content="转录组">
<meta property="article:tag" content="比较转录组学">
<meta property="article:tag" content="差异表达分析">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://biojuse.com/pic/cover2.jpg"><link rel="shortcut icon" href="/pic/webiron.png"><link rel="canonical" href="https://biojuse.com/2022/12/11/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%BA%94%EF%BC%89%E2%80%94%E2%80%94%20%E8%BD%AC%E5%BD%95%E6%9C%AC%E5%AE%9A%E9%87%8F%E4%B8%8E%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/"><link rel="preconnect" href="//cdn.jsdelivr.net"><link rel="preconnect" href="//busuanzi.ibruce.info"><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":false,"languages":{"hits_empty":"找不到您查询的内容：${query}"}},
  translate: {"defaultEncoding":1,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: true,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '比较转录组分析（五）—— 转录本定量与差异表达分析',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2024-08-22 20:44:00'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><link rel="stylesheet" href="/css/mycss.css"><link rel="stylesheet" href="/css/bg.css"><link rel="stylesheet" href="/css/hp.css"><script src="https://cdn1.tianli0.top/npm/vue@2.6.14/dist/vue.min.js"></script><script src="https://cdn1.tianli0.top/npm/element-ui@2.15.6/lib/index.js"></script><link rel="stylesheet" href="https://cdn1.tianli0.top/npm/element-ui@2.15.6/packages/theme-chalk/lib/index.css"><!-- hexo injector head_end start --><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Zfour/Butterfly-double-row-display@1.00/cardlistpost.min.css">
<style>#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags:before {content:"\A";
  white-space: pre;}#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags > .article-meta__separator{display:none}</style>
<link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiperstyle.css" media="print" onload="this.media='all'"><!-- hexo injector head_end end --><meta name="generator" content="Hexo 5.4.2"><link rel="alternate" href="/atom.xml" title="Juse's Blog" type="application/atom+xml">
</head><body><div id="loading-box"><div class="loading-left-bg"></div><div class="loading-right-bg"></div><div class="spinner-box"><div class="configure-border-1"><div class="configure-core"></div></div><div class="configure-border-2"><div class="configure-core"></div></div><div class="loading-word">加载中...</div></div></div><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/pic/cover.gif" data-original="/pic/gitav.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">53</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">18</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">10</div></a></div><hr><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fa fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> 文章</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fa fa-archive"></i><span> 归档</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fa fa-tags"></i><span> 标签</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fa fa-folder-open"></i><span> 分类</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/messageboard/"><i class="fa-fw fa fa-paper-plane"></i><span> 留言板</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fa fa-link"></i><span> 友链</span></a></div><div class="menus_item"><a class="site-page" href="/myself/"><i class="fa-fw fa fa-id-card"></i><span> 我自己</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fa fa-sitemap"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-shoe-prints"></i><span> 杂项</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/talk/"><i class="fa-fw fa fa-book"></i><span> 琐碎的日常</span></a></li><li><a class="site-page child" href="/new/"><i class="fa-fw fa fa-bell"></i><span> 更新日志</span></a></li><li><a class="site-page child" href="/paper/"><i class="fa-fw fa fa-sticky-note"></i><span> 文献笔记阁</span></a></li></ul></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/pic/cover2.jpg')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Juse's Blog</a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fa fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> 文章</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fa fa-archive"></i><span> 归档</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fa fa-tags"></i><span> 标签</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fa fa-folder-open"></i><span> 分类</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/messageboard/"><i class="fa-fw fa fa-paper-plane"></i><span> 留言板</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fa fa-link"></i><span> 友链</span></a></div><div class="menus_item"><a class="site-page" href="/myself/"><i class="fa-fw fa fa-id-card"></i><span> 我自己</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fa fa-sitemap"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-shoe-prints"></i><span> 杂项</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/talk/"><i class="fa-fw fa fa-book"></i><span> 琐碎的日常</span></a></li><li><a class="site-page child" href="/new/"><i class="fa-fw fa fa-bell"></i><span> 更新日志</span></a></li><li><a class="site-page child" href="/paper/"><i class="fa-fw fa fa-sticky-note"></i><span> 文献笔记阁</span></a></li></ul></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">比较转录组分析（五）—— 转录本定量与差异表达分析</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2022-12-11T07:25:00.000Z" title="发表于 2022-12-11 15:25:00">2022-12-11</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2024-08-22T12:44:00.132Z" title="更新于 2024-08-22 20:44:00">2024-08-22</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%AD%A6%E4%B9%A0/">学习</a><i class="fas fa-angle-right post-meta-separator"></i><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%AD%A6%E4%B9%A0/%E7%94%9F%E4%BF%A1/">生信</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">字数总计:</span><span class="word-count">4.6k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">阅读时长:</span><span>16分钟</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="比较转录组分析（五）—— 转录本定量与差异表达分析"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><hr>
<blockquote>
<div id="portalContainer"></div>
<script src="/js/ctlink.js"></script>
<script> generatePortalLinks(5);</script>
</blockquote>
<h2 id="前言"><a href="#前言" class="headerlink" title="前言"></a>前言</h2><p>之前看知乎看到过一篇关于写生物论文的文章，里面有一段很有意思的话，具体的内容我已经忘了，但意思大致是指<u>写生物论文最好 “水” 的就是差异表达分析</u>，我看到的第一反应是<strong>乐</strong>，第二反应是这可能确实是客观存在的。</p>
<p>不过会的人多用的人多并不代表这个分析烂大街没有含金量，相反，这么多人用应当是证实了这一分析的价值所在。</p>
<h3 id="更新日志"><a href="#更新日志" class="headerlink" title="更新日志"></a>更新日志</h3><p>2022.12.12 在 DEG 挑选部分中补充了些关于使用 voom 的内容。</p>
<p>2022.12.21 更改了关于 log2FoldChange 阈值的说法。</p>
<p>2022.12.22 补充了题外话中关于软件差异的新发现。</p>
<p>2023.04.20 补充了题外话中关于软件差异的新发现。</p>
<p>2023.07.03 精简了内容。</p>
<h3 id="出现的一些词"><a href="#出现的一些词" class="headerlink" title="出现的一些词"></a>出现的一些词</h3><p><strong>log2FoldChange</strong>：差异倍数，常以 1 作为阈值，有时也可为 1.2 或 1.5（一些单细胞研究中常见这些值）。</p>
<p>例：若对照组表达量为 1，处理组表达量为 4，则 log2FC = log2(4/1) = 2。正数时为上调，负数时为下调。</p>
<p><strong>生物学重复</strong>：经过相同方式处理的不同样品，提高生物学重复可以显著提高检测结果的准确性，一般而言一个较为可靠的分析需要<strong>每个处理至少三个重复</strong>。</p>
<p>例：3 个小鼠放在低温下处理，3 个小鼠放在常温下处理，然后比较不同温度下的小鼠转录组差异。</p>
<p>关于生物学重复的重要性可以看看这篇综述：</p>
<blockquote>
<p>Todd, Erica V et al. “The power and promise of RNA-seq in ecology and evolution.” <em>Molecular ecology</em> vol. 25,6 (2016): 1224-41. doi:10.1111/mec.13526</p>
</blockquote>
<p><strong>原始表达矩阵</strong>：进行差异表达分析时需要用到的各基因表达矩阵，里面记录了每个测序数据对应的基因表达量，被称为原始是因为还没进行标准化（不同测序数据的测序深度可能不同，会导致整体表达量的差异），标准化后可减弱测序深度差异带来的影响。</p>
<p><strong>探索性研究（Exploratory research）</strong>和<strong>描述性研究（Descriptive Research）</strong>的差别：</p>
<ul>
<li>Exploratory research is one which aims at providing insights into and an understanding of the problem faced by the researcher. Descriptive research, on the other hand, aims at describing something, mainly functions and characteristics.</li>
</ul>
<p><img src="/pic/cover.gif" data-original="/pic/dif.png"></p>
<ul>
<li>Conclusion: Therefore exploratory research results in insights or hypothesis, regardless of the method adopted, the most important thing is that it should remain flexible so that all the facets of the problem can be studied, as and when they arise. Conversely, descriptive research is a comparative design which is prepared according to the study and resources available. Such study minimises bias and maximises reliability.</li>
</ul>
<blockquote>
<p>Reference: <a target="_blank" rel="noopener" href="https://keydifferences.com/difference-between-exploratory-and-descriptive-research.html">Difference Between Exploratory and Descriptive Research (with Comparison Chart) - Key Differences</a></p>
</blockquote>
<p><strong>批次效应</strong>：<strong>不同时间、不同操作者、不同试剂、不同仪器</strong>导致的实验误差。</p>
<p>如果批次效应很严重，就可能会和真实的生物学差异相混淆，让结果难以捉摸。</p>
<h2 id="差异表达分析过程"><a href="#差异表达分析过程" class="headerlink" title="差异表达分析过程"></a>差异表达分析过程</h2><p>可以看到很多用于进行差异表达分析的比对软件，但这里首先要着重说明 <code>Hisat2</code> 这个软件是用于 <strong>有参</strong> 情况的。而对于无参转录组而言主流的比对软件是 <code>Bowtie2</code> ，不过这两个软件也同样是由同一队人开发。</p>
<p>具体的实现将依靠 Trinity 成熟的内置脚本进行，教程可见 <a target="_blank" rel="noopener" href="https://github.com/trinityrnaseq/trinityrnaseq/wiki/Trinity-Transcript-Quantification%EF%BC%8C%E8%BF%99%E9%87%8C%E4%B8%BB%E8%A6%81%E5%B0%B1%E6%98%AF%E6%8A%8A%E5%AE%83%E7%BF%BB%E6%88%90%E4%B8%AD%E6%96%87%E3%80%82">https://github.com/trinityrnaseq/trinityrnaseq/wiki/Trinity-Transcript-Quantification，这里主要就是把它翻成中文。</a></p>
<h3 id="估计转录本丰度"><a href="#估计转录本丰度" class="headerlink" title="估计转录本丰度"></a>估计转录本丰度</h3><p>使用的脚本是 Trinity 的 <code>align_and_estimate_abundance.pl</code>，不知道位置可以使用 <code>locate</code> 命令查找（没找到可能是版本不够）。</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">nohup</span> align_and_estimate_abundance.pl --transcripts cdhit.fasta \</span></span><br><span class="line"><span class="language-bash">--seqType fq --samples_file sample.txt --est_method RSEM \</span></span><br><span class="line"><span class="language-bash">--aln_method bowtie2 --thread_count 15 --debug --prep_reference \</span></span><br><span class="line"><span class="language-bash"><span class="comment">#--SS_lib_type RF \</span></span></span><br><span class="line"><span class="language-bash">&gt;bowtie.out 2&gt;&amp;1</span></span><br></pre></td></tr></tbody></table></figure>

<p>此处 <code>nohup</code> 用于避免终端意外退出导致运行中断，如果服务器或主机没有该程序也可去除，在这里：</p>
<ul>
<li><p><code>--transcripts</code> 为转录组组装的位置。</p>
</li>
<li><p><code>--seqType</code> 用 fq 就行。</p>
</li>
<li><p><code>--samples_file</code> 在组装的时候我们用到过，如果不用该参数则替换成相应的 <code>--left</code> 和 <code>--right</code>。</p>
</li>
</ul>
<p>tip：如果在组装时使用过 <code>--trimmomatic</code> ，那么这里的双端数据应该替换成质控后的位置，此外，也有可能参考转录组的组装数据并不是用来做差异表达分析的，因此可以直接在 <code>--transcripts</code> 指定以后，<code>sample.txt</code> 中放进行差异表达分析的数据就行。</p>
<p>在这里再放一下 <code>sample.txt</code> 所应该有的格式：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cond_A    cond_A_rep1    A_rep1_left.fq    A_rep1_right.fq</span><br><span class="line">cond_A    cond_A_rep2    A_rep2_left.fq    A_rep2_right.fq</span><br><span class="line">cond_A    cond_A_rep3    A_rep3_left.fq    A_rep3_right.fq</span><br><span class="line">cond_B    cond_B_rep1    B_rep1_left.fq    B_rep1_right.fq</span><br><span class="line">cond_B    cond_B_rep2    B_rep2_left.fq    B_rep2_right.fq</span><br><span class="line">cond_B    cond_B_rep3    B_rep3_left.fq    B_rep3_right.fq</span><br></pre></td></tr></tbody></table></figure>

<p>可把<strong>第一列看成不同的处理</strong>，<strong>第二列看成每个处理的生物学重复</strong>，第三第四列是对应的双端文件（也可为单端）。</p>
<ul>
<li><p><code>--est_method</code> 是丰度估计方式，分三个软件，有 <code>RSEM</code> <code>kallisto</code> 和 <code>salmon</code>，这里用 <code>RSEM</code>。</p>
</li>
<li><p><code>--aln_method</code> 是 <code>--est_method</code> 选择 <code>RSEM</code> 时需要配置的，可以选择 <code>bowtie</code> 和 <code>bowtie2</code>，此处使用后者。</p>
</li>
<li><p><code>--thread_count</code> 线程数，越大越快，但大到某个值以后再加效果有限，所以不必用太多。</p>
</li>
<li><p><code>--debug</code> 运行时开启该参数，在运行结束后不会删除中间文件，如果所剩空间不多可以不用加这个参数。</p>
</li>
<li><p><code>--prep_reference</code> 自动为转录组建立索引。</p>
</li>
<li><p><code>--SS_lib_type</code> 当测序数据为<strong>链特异性</strong>时需要添加，否则会导致定量不准确。</p>
</li>
<li><p><code>&gt;bowtie.out 2&gt;&amp;1</code>  最后这一行是为了将运行过程中的屏幕输出储存到 bowtie.out 这个文件中，方便后续查看比对率等信息。</p>
</li>
</ul>
<p>运行结束以后，会跑出来所有的比对结果，以上述 sample.txt 为例，跑出来的各个文件应该长下面这样：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">ls</span></span></span><br><span class="line">cond_A_rep1 	cond_A_rep2 	cond_A_rep3</span><br><span class="line">cond_B_rep1 	cond_B_rep2 	cond_B_rep3</span><br></pre></td></tr></tbody></table></figure>

<p>跑出来的都是文件夹，里面的 <code>RSEM.isoforms.results</code> 就是各个重复的定量结果。</p>
<p><img src="/pic/cover.gif" data-original="/pic/dljg.png"></p>
<p>然后，可以用一串命令将这些定量结果统计起来：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">ls</span> */RSEM.isoforms.results &gt; countfile.txt</span></span><br></pre></td></tr></tbody></table></figure>

<p>所有的定量结果位置都已统计在 countfile.txt 中，此后运行下列命令：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">abundance_estimates_to_matrix.pl --est_method RSEM \</span></span><br><span class="line"><span class="language-bash">--gene_trans_map none --quant_files countfile.txt \</span></span><br><span class="line"><span class="language-bash">--name_sample_by_basedir</span></span><br></pre></td></tr></tbody></table></figure>

<p>此处因为我们选择在 isoform 水平上分析，因此 <code>--gene_trans_map</code> 选用 <code>none</code>。</p>
<p>使用 <code>--name_sample_by_basedir</code> 以让表达矩阵中的样本名与文件夹一致。</p>
<p>运行结束后得到的 <code>RSEM.isoform.counts.matrix</code> 就是我们在进行差异表达分析时需要用到的<strong>原始表达矩阵</strong>了。</p>
<p>接着我们再运行：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">cut</span> -f 1,2 sample.txt &gt; sample5.txt</span></span><br></pre></td></tr></tbody></table></figure>

<p>这里的 sample5.txt 可用于后续的样本质量检查和差异表达分析。</p>
<h3 id="检查样本质量"><a href="#检查样本质量" class="headerlink" title="检查样本质量"></a>检查样本质量</h3><p>如果觉得需要进行：<a target="_blank" rel="noopener" href="https://github.com/trinityrnaseq/trinityrnaseq/wiki/QC-Samples-and-Biological-Replicates">https://github.com/trinityrnaseq/trinityrnaseq/wiki/QC-Samples-and-Biological-Replicates</a></p>
<p>我个人觉得所有图里最有价值的是 <code>sample_cor_matrix</code> 和 <code>principal_components</code> 的图（相关性热图 &amp; 主成分分析）。</p>
<p><img src="/pic/cover.gif" data-original="/pic/sample_cor_matrix.png"></p>
<p><img src="/pic/cover.gif" data-original="/pic/PCA.png"></p>
<p>运行基本都是照官网教的做，而且这个脚本的可设置参数非常多，里面的某些默认参数似乎需要根据自己的数据情况进行微调才更严谨。</p>
<p>这一步最好还是跑一下，正常的情况应该是：</p>
<p>①、在 <code>sample_cor_matrix</code> 中，不同处理的各个生物学重复应该要分别聚在一块。</p>
<p>②、在 <code>principal_components</code> 中，一个处理的各个生物学重复的点应该在邻近的区域内，不同处理的点应该要分开。</p>
<p>例子：</p>
<p><img src="/pic/cover.gif" data-original="/pic/cor.jpg"></p>
<p><img src="/pic/cover.gif" data-original="/pic/pca.jpg"></p>
<p>如果<strong>不同处理混到一起了</strong>有可能是下面的几个情况之一：</p>
<p>①、样本的变异系数很大，也就是说这个生物即使是在同一处理下，不同个体之间也可能表现出较大的差异，所以彼此之间很离散。</p>
<p>②、实验方法上的问题，可能样本处理不当等；又或者是样本编号错误。</p>
<p>如果<strong>同一处理下有很明显分开的两个区块</strong>则可能是：</p>
<p>①、批次效应，可见 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/0bec17eb301f">批次效应 from 简书</a>。</p>
<p>②、巧合。    </p>
<h3 id="差异表达分析"><a href="#差异表达分析" class="headerlink" title="差异表达分析"></a>差异表达分析</h3><p>可选择的软件：<code>edgeR</code>、<code>DESeq2</code>、<code>limma-voom</code>。</p>
<p>前两者的使用应该是最广泛的，具体使用哪一个可以根据具体情况确定，例如：</p>
<p>①、如果差异表达基因鉴定数量较少，可以选择得到更多差异表达基因的软件，亦或者是取多个软件的<strong>并集</strong>。</p>
<p>②、如果差异表达基因鉴定数量很多，可以取多个软件的<strong>交集</strong>。</p>
<p>关于取并集和交集，两个方法都有人用，探索性（exploratory）的研究用前者也行，但如果是描述性（descriptive）的研究用后者会更合适，因为总的来说肯定是后者更为精准可靠（假阳性率更低）。</p>
<p>首先，要先检查一下自己的 linux 中是否有所需要的包，不确定可以直接安装试试：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">R <span class="comment">#进入 R</span></span> </span><br><span class="line"><span class="meta prompt_">&gt; </span><span class="language-bash"><span class="keyword">if</span> (!requireNamespace(<span class="string">"BiocManager"</span>, quietly = TRUE))</span></span><br><span class="line">    install.packages("BiocManager")</span><br><span class="line"><span class="meta prompt_">&gt; </span><span class="language-bash">BiocManager::install(c(<span class="string">"edgeR"</span>, <span class="string">"limma"</span>, <span class="string">"DESeq2"</span>, <span class="string">"ctc"</span>, <span class="string">"Biobase"</span>, <span class="string">"gplots"</span>, <span class="string">"ape"</span>, <span class="string">"argparse"</span>))</span></span><br></pre></td></tr></tbody></table></figure>

<p>安装好以后输出 <code>q()</code> 退出 R，并使用 Trinity 的 pipeline 进行分析，首先你需要这样一份文件：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cond_A	cond_B</span><br><span class="line">cond_C	cond_B</span><br></pre></td></tr></tbody></table></figure>

<p>中间是制表符分列，命名成 contrast.txt。可以是多行也可以是一行，里面不同的 cond 代表不同的处理，每一行代表一次比对，排在后一列的是<strong>对照组</strong>。</p>
<p>以上述文本为例，后续的差异表达分析将以 cond_B 作为对照组，以 cond_A 和 cond_C 作为处理组分别进行两次比对并输出结果。</p>
<p>记住<strong>处理组和对照组千万不要放反了，不然后续的分析会全部按照真实情况的相反面进行</strong>，我依然记得在之前生信课上有一个小组汇报的很好，但是由于这一步做错了她们把上调基因当下调分析，下调基因当上调分析，虽然听她们讲的很好但是事实上完全是反着来讲的。</p>
<p>将 <code>contrast.txt</code> 弄好后运行下列命令：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">DifferentialExpression/run_DE_analysis.pl  \ <span class="comment"># 老样子，不知道位置可以用 locate 确定。</span></span></span><br><span class="line">--matrix RSEM.isoform.counts.matrix \ # 如果使用的别的定量方法就需要进行相应的替换</span><br><span class="line">--method DESeq2 \</span><br><span class="line">--min_reps_min_cpm 3,1 \</span><br><span class="line">--samples_file sample5.txt --contrasts contrast.txt</span><br></pre></td></tr></tbody></table></figure>

<ul>
<li><p><code>--matrix</code> <strong>一定要用原始表达矩阵</strong>！每个软件都有它自己的标准化步骤嵌在分析流程中，所以不需要担心。</p>
</li>
<li><p><code>--method</code> 有 <code>DESeq2</code> <code>edgeR</code> 和 <code>limma-voom</code>。</p>
</li>
<li><p><code>--min_reps_min_cpm</code> 筛选基因，例如此处的意思是筛选至少在三个样本中出现的且 <strong>count-per-millon</strong> 不低于 1 的基因。</p>
</li>
</ul>
<p>运行结束后会得到一个表格，此后可以用这个表格进行火山图绘制或挑选差异表达基因。</p>
<h3 id="DEG-挑选"><a href="#DEG-挑选" class="headerlink" title="DEG 挑选"></a>DEG 挑选</h3><p>在差异表达分析结果中以 <code>|log2FC| &gt; 1</code>, <code>q-value (or FDR) &lt; 0.05</code> 作为阈值进行筛选。</p>
<p>挑选方法：自己手动在 excel 里面筛选或者用 python：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"></span><br><span class="line">raw_file = <span class="built_in">open</span>(sys.argv[<span class="number">1</span>],<span class="string">"r"</span>)</span><br><span class="line">out_up_file = <span class="built_in">open</span>(sys.argv[<span class="number">2</span>],<span class="string">"w"</span>)</span><br><span class="line">out_down_file = <span class="built_in">open</span>(sys.argv[<span class="number">3</span>],<span class="string">"w"</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> line <span class="keyword">in</span> raw_file:</span><br><span class="line">	string = line.split(<span class="string">"\t"</span>)</span><br><span class="line">	<span class="keyword">if</span> string[<span class="number">0</span>] == <span class="string">'sampleA'</span>:</span><br><span class="line">		<span class="keyword">continue</span></span><br><span class="line">	<span class="keyword">if</span> <span class="built_in">len</span>(string) == <span class="number">11</span>:</span><br><span class="line">		<span class="keyword">if</span> <span class="built_in">abs</span>(<span class="built_in">float</span>(string[<span class="number">6</span>])) &gt;= <span class="number">1.0</span> <span class="keyword">and</span> <span class="built_in">float</span>(string[-<span class="number">1</span>]) &lt;= <span class="number">0.05</span>:</span><br><span class="line">			<span class="keyword">if</span> <span class="built_in">float</span>(string[<span class="number">6</span>]) &gt; <span class="number">0</span>:</span><br><span class="line">				out_up_file.write(string[<span class="number">0</span>] + <span class="string">'\n'</span>)</span><br><span class="line">			<span class="keyword">else</span>:</span><br><span class="line">				out_down_file.write(string[<span class="number">0</span>] + <span class="string">'\n'</span>)</span><br><span class="line">	<span class="keyword">else</span>:</span><br><span class="line">		<span class="keyword">if</span> <span class="built_in">abs</span>(<span class="built_in">float</span>(string[<span class="number">3</span>])) &gt;= <span class="number">1.0</span> <span class="keyword">and</span> <span class="built_in">float</span>(string[-<span class="number">1</span>]) &lt;= <span class="number">0.05</span>:</span><br><span class="line">			<span class="keyword">if</span> <span class="built_in">float</span>(string[<span class="number">3</span>]) &gt; <span class="number">0</span>:</span><br><span class="line">				out_up_file.write(string[<span class="number">0</span>] + <span class="string">'\n'</span>)</span><br><span class="line">			<span class="keyword">else</span>:</span><br><span class="line">				out_down_file.write(string[<span class="number">0</span>] + <span class="string">'\n'</span>)</span><br></pre></td></tr></tbody></table></figure>

<p>这个我针对上述脚本输出的文件所写的一个提取差异表达基因的脚本，如果想要更改筛选标准可以在里面做相应的调整，或者添加可指定的参数项。</p>
<p>使用例：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">python de.py xxx.DE_results DEG_up.txt DEG_down.txt</span></span><br><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">cat</span> *up.txt | <span class="built_in">sort</span> | <span class="built_in">uniq</span> &gt; DEG_all_up.txt</span></span><br><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">cat</span> *down.txt | <span class="built_in">sort</span> | <span class="built_in">uniq</span> &gt; DEG_all_down.txt</span></span><br><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">wc</span> -l DEGS* &gt; stast.txt <span class="comment"># 统计差异表达基因的数量</span></span></span><br></pre></td></tr></tbody></table></figure>

<p>输出的 <code>DEG_up.txt</code> 和 <code>DEG_down.txt</code> 里分别是上调和下调的差异表达基因列表（可根据自身需求调整输出的文件名称），后面几步可选择进行，用处主要是合并所有软件得到的差异表达基因以及统计有多少上调和下调的基因。</p>
<p>获得差异表达基因以后，就可以准备进行后续的富集分析了。</p>
<h2 id="后记"><a href="#后记" class="headerlink" title="后记"></a>后记</h2><p>下一篇预告：比较转录组分析（六）—— GO 富集分析与可视化。</p>
<p>火山图教程预告（随缘）：基于 GraphPad Prism 的火山图绘制。</p>
<p>使用 R 绘制火山图的话可以看看这一篇教程：<a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/82785739">R 数据可视化 02 | 火山图 from 知乎</a>。</p>
<h3 id="①、关于三个转录本丰度估计软件的差异"><a href="#①、关于三个转录本丰度估计软件的差异" class="headerlink" title="①、关于三个转录本丰度估计软件的差异"></a>①、关于三个转录本丰度估计软件的差异</h3><p>RSEM 是需要比对结果进行定量的，可以调用的软件有 STAR Bowtie2 Bowtie 等，文里使用了 Bowtie2。其他两个是不需要进行比对就可以直接进行定量的工具。具体哪个最为准确的问题，感兴趣可以看一下这篇论文：<a target="_blank" rel="noopener" href="https://www.sciencedirect.com/science/article/pii/S1874939919300501#bb0230">Interoperable RNA-Seq analysis in the cloud</a>。这篇文章对 salmon 和 kallisto 评价是比较高的（不过和 qPCR 的结果最不相似<del>，事实上和其他的也就差了一丢丢，我个人不排除误差问题</del>），RSEM 的结果和它们相似但是由于运行效能的问题被 diss 了。</p>
<p>再找，然后在 <a target="_blank" rel="noopener" href="https://www.nature.com/articles/nbt.3519/">kallisto 的论文</a>中找到了这张图：</p>
<p><img src="/pic/cover.gif" data-original="/pic/bdrj.jpg"></p>
<p>简单来说就是把每个转录本定量结果的中位数差异作为精确性的指标，然后可以看到 Bowtie2 + RSEM 的组合是最为精准的，kallisto 虽然次了一点点但也不遑多让。而当考虑到运行时间时，Bowtie2 + RSEM 的组合很不幸地被 kallisto 碾压了，还不是一点点的程度。可以看到 kallisto 的定量是非常迅速且也保证了精确度的。所以运行内存有限并且追求精确度的情况下，kallisto 毫无疑问是更好的选择。</p>
<p>此外在一篇讨论 salmon 精度的文章中，Bowtie2 + RSEM 在某些情况下的精准度比其高（当数据和参考转录组一致的情况下），当参考转录组和 reads 并不完全匹配时，则 salmon 更为精确些。</p>
<p>文章：<a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02151-8">Alignment and mapping methodology influence transcript abundance estimation</a></p>
<p>关于不同软件具体的差异在哪里可以精读一下文章，这里我就只搬个结论作为参考。</p>
<p>软件的效率（速度）是一个非常重要的参考依据，如果想用有限的内存得到最好的效果，那么 kallisto 和 salmon 的优先级要高很多。</p>
<h3 id="②、不同差异表达分析软件的差异"><a href="#②、不同差异表达分析软件的差异" class="headerlink" title="②、不同差异表达分析软件的差异"></a>②、不同差异表达分析软件的差异</h3><p>找文章，直接找文章：</p>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-022-02648-4">Exaggerated false positives by popular differential expression methods when analyzing human population samples</a></p>
<p>from Genome Biology</p>
</blockquote>
<p>这里就不放图了，简单说一下几个结论：</p>
<p>①、在大样本量时，DESeq2 和 edgeR 具有惊人的假阳性率，两者中 edgeR 更为明显。</p>
<blockquote>
<p>造成这些现象的原因是：However, while the three parametric methods were initially designed to address the small-sample-size issue, these population-level studies had much larger sample sizes (at least dozens) and thus no longer needed restrictive parametric assumptions. Moreover, violation of parametric assumptions would lead to ill-behaved <em>p</em>-values and likely failed FDR control, an issue independent of the sample size.</p>
<p>简单来说，就是这三种方法都有设置相应的假设以应对低样本量的情况（这个假设允许混合有相近表达情况的基因数据以估计方差），然而在大样本量的研究中这一假设不再被需要，如果继续坚持这一假设会导致错误的估计。</p>
</blockquote>
<p>②、相比之下大多数情况中 limma-voom 表现更为良好（不管小样本量大样本量）。</p>
<p>③、推荐在大样本量时使用 wilcoxon 秩和检验，这种非参数的方法在面对异常值时会显得更加可靠。但与此同时，由于其是一种非基于回归的方法，因此无法针对混杂的因素进行调整，因此使用该方法进行 DEG 鉴定前需先标准化或者使用概率指数模型调整协变量。（文中的方法是先用 edgeR 进行过滤和标准化然后再使用 wilcoxon 秩和检验 + Benjamini&amp;Hochberg 矫正）</p>
<p>tip: 有趣的是我发现网上有关于第一篇文章的介绍，写的非常好，传送门：<a target="_blank" rel="noopener" href="https://www.plob.org/article/27145.html">别再用 DEseq2 和 edgeR 进行大样本差异表达基因分析了</a>。</p>
<p><strong>补充（2023.04.20）：</strong></p>
<p>事实上我在大概一月份左右就发现了新东西，但是一直没有更新，今天心血来潮突然就想起来，于是乎就进行了补充。</p>
<p>新的发现就是：</p>
<ul>
<li><p>样本多（重复多）时，edgeR 多于 DESeq2。</p>
</li>
<li><p>样本少（重复少）时，edgeR 少于 DESeq2。</p>
</li>
</ul>
<p>这可能与它们的算法差异有关，这一点可以看 DESeq2 论文原文片段：</p>
<blockquote>
<p>Besides the need to account for the specifics of count data, such as non-normality and a dependence of the variance on the mean, a core challenge is the small number of samples in typical HTS experiments – often as few as two or three replicates per condition. Inferential methods that treat each gene separately suffer here from lack of power, due to the high uncertainty of within-group variance estimates. In high-throughput assays, this limitation can be overcome by pooling information across genes, specifically, by exploiting assumptions about the similarity of the variances of different genes measured in the same experiment.</p>
<p>Many methods for differential expression analysis of RNA-seq data perform such information sharing across genes for variance (or, equivalently, dispersion) estimation. <em>edgeR</em> moderates the dispersion estimate for each gene toward a common estimate across all genes, or toward a local estimate from genes with similar expression strength, using a weighted conditional likelihood. Our <em>DESeq</em> method detects and corrects dispersion estimates that are too low through modeling of the dependence of the dispersion on the average expression strength over all samples.</p>
</blockquote>
<h3 id="③、关于批次效应"><a href="#③、关于批次效应" class="headerlink" title="③、关于批次效应"></a>③、关于批次效应</h3><p>一个会毁坏整个实验的事情。</p>
<p>事实往往比言语更有力：</p>
<p><img src="/pic/cover.gif" data-original="/pic/batch1.png"></p>
<p>这是一张 NMDS 图，乍一看好像有挺多值得琢磨的东西，比如这三组里面红色组和绿色组的势头是不一样的。</p>
<p><img src="/pic/cover.gif" data-original="/pic/batch2.png"></p>
<p>然而事实上，红圈所框住的地方所对应的样本与其他样本的采集时间是不一致的（相隔了两个季节），导致虽然这些样本的分组和同组其他样本是一样的，但它们的 NMDS 结果却截然不同并且出现分层。</p>
<p>这就是批次效应，也是很多研究者们想要尽力避免的问题。所以在设计实验时就应当注意，需要把除了自己关注的因素以外的所有其他干扰因素给排除或者尽力降到最低，避免出现部分结论无法解释的现象以及避免产生批次效应。</p>
<!-- hexo injector body_end start --> <script data-pjax="">if(document.getElementById('recent-posts') && (location.pathname ==='/'|| '/' ==='all')){
    var parent = document.getElementById('recent-posts');
    var child = '<div class="recent-post-item" style="width:100%;height: auto"><div id="catalog_magnet"><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/生信/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📚 Juseの生信分享 (23)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/绘图/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">👩‍💻 Juseの绘图经验 (4)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/杂项/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🤹 Juseの琐碎教程 (13)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/文献阅读/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📕 Juseの文献笔记 (2)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/软件开发/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🧰 Juseの软件开发 (8)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/博客/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📑 Juseの博客轻改 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><a class="magnet_link_more"  target="_blank" rel="noopener" href="https://jusetiz.github.io/categories" style="flex:1;text-align: center;margin-bottom: 10px;">查看更多...</a></div></div>';
    console.log('已挂载magnet')
    parent.insertAdjacentHTML("afterbegin",child)}
     </script><style>#catalog_magnet{flex-wrap: wrap;display: flex;width:100%;justify-content:space-between;padding: 10px 10px 0 10px;align-content: flex-start;}.magnet_item{flex-basis: calc(33.333333333333336% - 5px);background: #f2f2f2;margin-bottom: 10px;border-radius: 8px;transition: all 0.2s ease-in-out;}.magnet_item:hover{background: #b30070}.magnet_link_more{color:#555}.magnet_link{color:black}.magnet_link:hover{color:white}@media screen and (max-width: 600px) {.magnet_item {flex-basis: 100%;}}.magnet_link_context{display:flex;padding: 10px;font-size:16px;transition: all 0.2s ease-in-out;}.magnet_link_context:hover{padding: 10px 20px;}</style>
    <style></style><script data-pjax="">
  function butterfly_swiper_injector_config(){
    var parent_div_git = document.getElementById('recent-posts');
    var item_html = '<div class="recent-post-item" style="height: auto;width: 100%"><div class="blog-slider swiper-container-fade swiper-container-horizontal" id="swiper_container"><div class="blog-slider__wrp swiper-wrapper" style="transition-duration: 0ms;"><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-09-01</span><a class="blog-slider__title" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt="">JuseKit（八） —— 计算转录组组装指标</a><div class="blog-slider__text">本次更新新增了计算转录组组装指标的功能，经实测比 Trinity 快且结果一致。</div><a class="blog-slider__button" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-08-04</span><a class="blog-slider__title" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt="">JuseKit（七） —— 绘制 GO 富集分析气泡图</a><div class="blog-slider__text">本次更新新增 GO 富集气泡图绘制方法，只需文件一键执行！</div><a class="blog-slider__button" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-07-12</span><a class="blog-slider__title" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt="">DiscoVista 可视化系统发育不一致</a><div class="blog-slider__text">使用合适的工具将系统发育的不一致更加清晰地呈现出来。</div><a class="blog-slider__button" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/05/04/关于 PAML 的一二三事/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-05-04</span><a class="blog-slider__title" href="2023/05/04/关于 PAML 的一二三事/" alt="">关于 PAML 的一二三事</a><div class="blog-slider__text">PAML 进行正选择检测、选择压力分析和物种分歧时间估计的详细介绍，包括诸多需要注意的事项。</div><a class="blog-slider__button" href="2023/05/04/关于 PAML 的一二三事/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-07-04</span><a class="blog-slider__title" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt="">JuseKit（六） —— 绘制火山图</a><div class="blog-slider__text">此次更新新增了绘制火山图的功能，仅需输入对应格式的文件即可一键绘制火山图。</div><a class="blog-slider__button" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-05-14</span><a class="blog-slider__title" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt="">JuseKit（五） —— 用于系统发育分析的序列过滤</a><div class="blog-slider__text">本次更新带来序列过滤功能，可根据序列长度及物种序列数过滤比对。</div><a class="blog-slider__button" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/wxyd1.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-04-06</span><a class="blog-slider__title" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt="">系统发育转录组学是否靠谱？</a><div class="blog-slider__text">向来只闻 Phylogenomics，那么 Phylotranscriptomics 又如何呢？</div><a class="blog-slider__button" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/slwd.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-03-13</span><a class="blog-slider__title" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt="">单基因 dN、dS 和 ω 的滑动窗口绘制方法</a><div class="blog-slider__text">绘制可以观察基因内不同部位选择情况的滑动窗口图，基于 Kaks_calculator 计算。</div><a class="blog-slider__button" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt="">详情   </a></div></div></div><div class="blog-slider__pagination swiper-pagination-clickable swiper-pagination-bullets"></div></div></div>';
    console.log('已挂载butterfly_swiper')
    parent_div_git.insertAdjacentHTML("afterbegin",item_html)
    }
  var elist = 'undefined'.split(',');
  var cpage = location.pathname;
  var epage = 'all';
  var flag = 0;

  for (var i=0;i<elist.length;i++){
    if (cpage.includes(elist[i])){
      flag++;
    }
  }

  if ((epage ==='all')&&(flag == 0)){
    butterfly_swiper_injector_config();
  }
  else if (epage === cpage){
    butterfly_swiper_injector_config();
  }
  </script><script defer="" src="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper.min.js"></script><script defer="" data-pjax="" src="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper_init.js"></script><!-- hexo injector body_end end --></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="https://biojuse.com">Juse</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="https://biojuse.com">https://biojuse.com</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://biojuse.com" target="_blank">Juse's Blog</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/">生物信息学</a><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1/">生信</a><a class="post-meta__tags" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">转录组</a><a class="post-meta__tags" href="/tags/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">比较转录组学</a><a class="post-meta__tags" href="/tags/%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/">差异表达分析</a></div><div class="post_share"><div class="social-share" data-image="/pic/cover2.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer=""></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2022/12/19/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E5%85%AD%EF%BC%89%E2%80%94%E2%80%94%20GO%20%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/"><img class="prev-cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">比较转录组分析（六）—— GO 富集分析与可视化</div></div></a></div><div class="next-post pull-right"><a href="/2022/11/28/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E5%9B%9B%EF%BC%89%E2%80%94%E2%80%94%20%E7%BB%84%E8%A3%85%E7%9A%84%E6%B3%A8%E9%87%8A/"><img class="next-cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">比较转录组分析（四）—— 组装的 GO 及 KEGG 注释</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><div><a href="/2022/12/23/%E5%9F%BA%E4%BA%8E%20GraphPad%20Prism%20%E7%9A%84%E7%81%AB%E5%B1%B1%E5%9B%BE%E7%BB%98%E5%88%B6/" title="基于 GraphPad Prism 的火山图绘制"><img class="cover" src="/pic/cover.gif" data-original="/pic/prism17.png" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">基于 GraphPad Prism 的火山图绘制</div></div></a></div><div><a href="/2022/11/01/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%B8%80%EF%BC%89%E2%80%94%E2%80%94%20%E5%89%8D%E6%8F%90%E5%8F%8A%E6%A6%82%E8%A6%81/" title="比较转录组分析（一）—— 前提及概要"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-07-06</div><div class="title">比较转录组分析（一）—— 前提及概要</div></div></a></div><div><a href="/2022/11/21/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%B8%89%EF%BC%89%E2%80%94%E2%80%94%20%E7%BB%84%E8%A3%85%E7%9A%84%E8%B4%A8%E9%87%8F%E6%A3%80%E6%B5%8B%E4%B8%8E%E5%8E%BB%E5%86%97%E4%BD%99/" title="比较转录组分析（三）—— 组装的质量检测与去冗余"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-10-17</div><div class="title">比较转录组分析（三）—— 组装的质量检测与去冗余</div></div></a></div><div><a href="/2022/11/06/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%BA%8C%EF%BC%89%E2%80%94%E2%80%94%20%E5%8E%9F%E5%A7%8B%E6%95%B0%E6%8D%AE%E7%9A%84%E8%B4%A8%E6%8E%A7%E4%B8%8E%E8%BD%AC%E5%BD%95%E7%BB%84%E7%BB%84%E8%A3%85/" title="比较转录组分析（二）—— 原始数据的质控与转录组组装"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">比较转录组分析（二）—— 原始数据的质控与转录组组装</div></div></a></div><div><a href="/2022/12/19/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E5%85%AD%EF%BC%89%E2%80%94%E2%80%94%20GO%20%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/" title="比较转录组分析（六）—— GO 富集分析与可视化"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-07-25</div><div class="title">比较转录组分析（六）—— GO 富集分析与可视化</div></div></a></div><div><a href="/2022/11/28/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E5%9B%9B%EF%BC%89%E2%80%94%E2%80%94%20%E7%BB%84%E8%A3%85%E7%9A%84%E6%B3%A8%E9%87%8A/" title="比较转录组分析（四）—— 组装的 GO 及 KEGG 注释"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-10-17</div><div class="title">比较转录组分析（四）—— 组装的 GO 及 KEGG 注释</div></div></a></div></div></div><hr><div id="post-comment"><div class="comment-head"><div class="comment-headline"><i class="fas fa-comments fa-fw"></i><span> 评论</span></div></div><div class="comment-wrap"><div><div class="vcomment" id="vcomment"></div></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/pic/cover.gif" data-original="/pic/gitav.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"></div><div class="author-info__name">Juse</div><div class="author-info__description">偶尔徒步行进倒也不错</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">53</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">18</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">10</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/JuseTiZ"><i class="fab fa-github"></i><span>Follow Juse</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="/qq.html" target="_blank" title=""><i class="fab fa-qq"></i></a><a class="social-icon" href="https://space.bilibili.com/31269335" target="_blank" title=""><i class="fab fa-bilibili"></i></a><a class="social-icon" href="https://github.com/JuseTiZ" target="_blank" title=""><i class="fab fa-github"></i></a><a class="social-icon" href="https://steamcommunity.com/id/291195857/" target="_blank" title=""><i class="fab fa-steam"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content"><p align="center"><b>人生是旷野，不是轨道</b><br>⭐⭐🗞️ 新闻速递 📰⭐⭐<br>Juse 抽卡又保底了</p></div><div id="welcome-info"></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%89%8D%E8%A8%80"><span class="toc-number">1.</span> <span class="toc-text">前言</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%9B%B4%E6%96%B0%E6%97%A5%E5%BF%97"><span class="toc-number">1.1.</span> <span class="toc-text">更新日志</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%87%BA%E7%8E%B0%E7%9A%84%E4%B8%80%E4%BA%9B%E8%AF%8D"><span class="toc-number">1.2.</span> <span class="toc-text">出现的一些词</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%E8%BF%87%E7%A8%8B"><span class="toc-number">2.</span> <span class="toc-text">差异表达分析过程</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BC%B0%E8%AE%A1%E8%BD%AC%E5%BD%95%E6%9C%AC%E4%B8%B0%E5%BA%A6"><span class="toc-number">2.1.</span> <span class="toc-text">估计转录本丰度</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%A3%80%E6%9F%A5%E6%A0%B7%E6%9C%AC%E8%B4%A8%E9%87%8F"><span class="toc-number">2.2.</span> <span class="toc-text">检查样本质量</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90"><span class="toc-number">2.3.</span> <span class="toc-text">差异表达分析</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#DEG-%E6%8C%91%E9%80%89"><span class="toc-number">2.4.</span> <span class="toc-text">DEG 挑选</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%90%8E%E8%AE%B0"><span class="toc-number">3.</span> <span class="toc-text">后记</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E2%91%A0%E3%80%81%E5%85%B3%E4%BA%8E%E4%B8%89%E4%B8%AA%E8%BD%AC%E5%BD%95%E6%9C%AC%E4%B8%B0%E5%BA%A6%E4%BC%B0%E8%AE%A1%E8%BD%AF%E4%BB%B6%E7%9A%84%E5%B7%AE%E5%BC%82"><span class="toc-number">3.1.</span> <span class="toc-text">①、关于三个转录本丰度估计软件的差异</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E2%91%A1%E3%80%81%E4%B8%8D%E5%90%8C%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90%E8%BD%AF%E4%BB%B6%E7%9A%84%E5%B7%AE%E5%BC%82"><span class="toc-number">3.2.</span> <span class="toc-text">②、不同差异表达分析软件的差异</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E2%91%A2%E3%80%81%E5%85%B3%E4%BA%8E%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94"><span class="toc-number">3.3.</span> <span class="toc-text">③、关于批次效应</span></a></li></ol></li></ol></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div id="ft"><div class="ft-item-1"><div class="t-top"><div class="t-t-l"><p class="ft-t t-l-t">想说的话😉</p><div class="bg-ad"><div>Genes are not destiny. While they play a large role in shaping how we act, how we feel, and how we interact with others, our environment and personal choices have just as much impact on our behavior.</div><div class="btn-xz-box"><a class="btn-xz" target="_blank" rel="noopener" href="https://habitica.com/static/home">游戏人生不游戏 🎯</a></div></div></div><div class="t-t-r"><p class="ft-t t-l-t">仙人指路😝</p><ul class="ft-links"><li><a target="_blank" rel="noopener" href="https://www.fomal.cc/">建站参考</a><a target="_blank" rel="noopener" href="https://emojixd.com/">表情大全</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/talk/">看看生活</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/messageboard/">我要留言</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/myself/">关于博主</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/archives/">文章归档</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/categories/">文章分类</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/tags/">文章标签</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/about/">有的没的</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/new/">更新日志</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/link/">友情链接</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/404.html">404页面</a></li></ul></div></div></div><div class="ft-item-2"><p class="ft-t">友人链接🔗</p><div class="ft-img-group"><div class="img-group-item"><a target="_blank" rel="noopener" href="https://linctanny.github.io/"><img src="/pic/cover.gif" data-original="https://linctanny.github.io/img/p7.jpeg" alt=""></a></div><div class="img-group-item"><a target="_blank" rel="noopener" href="https://robust-vase.github.io/"><img src="/pic/cover.gif" data-original="https://robust-vase.github.io/img/John_King.webp" alt=""></a></div><div class="img-group-item"><a target="_blank" rel="noopener" href="https://blog.stariverfeel.eu.org/"><img src="/pic/cover.gif" data-original="https://avatars.githubusercontent.com/u/26035995?v=4" alt=""></a></div></div></div></div><div class="copyright">©2022 - 2024  @ Juse</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo </a><span class="footer-separator">|</span><span> 主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly </a><span class="footer-separator">|</span><span> 部署 </span><a target="_blank" rel="noopener" href="https://github.com/">GitPage </a><span class="footer-separator">|</span><span> 加速 </span><a target="_blank" rel="noopener" href="https://dash.cloudflare.com/">Cloudflare </a><span class="footer-separator">|</span><span> </span><a href="https://icp.gov.moe/?keyword=20232400" target="_blank">萌ICP备20232400号</a></div><div id="workboard"></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="translateLink" type="button" title="简繁转换">简</button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><a id="to_comment" href="#post-comment" title="直达评论"><i class="fas fa-comments"></i></a><button id="go-up" type="button" title="回到顶部"><i class="fas fa-arrow-up"></i></button></div></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"></div></div><hr><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div class="js-pjax" id="rightMenu"><div class="rightMenu-group rightMenu-small"><a class="rightMenu-item" href="javascript:window.history.back();"><i class="fa fa-arrow-left"></i></a><a class="rightMenu-item" href="javascript:window.history.forward();"><i class="fa fa-arrow-right"></i></a><a class="rightMenu-item" href="javascript:window.location.reload();"><i class="fa fa-refresh"></i></a><a class="rightMenu-item" href="/"><i class="fa fa-home"></i></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-text"><a class="rightMenu-item" href="javascript:rmf.copySelect();"><i class="fa fa-copy"></i><span>复制</span></a><a class="rightMenu-item" href="javascript:window.open(&quot;https://cn.bing.com/search?q=&quot;+window.getSelection().toString());window.location.reload();"><i class="fa fa-search"></i><span>必应搜索</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-too"><a class="rightMenu-item" href="javascript:window.open(window.getSelection().toString());window.location.reload();"><i class="fa fa-link"></i><span>转到链接</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-paste"><a class="rightMenu-item" href="javascript:rmf.paste()"><i class="fa fa-copy"></i><span>粘贴</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-post"><a class="rightMenu-item" href="#post-comment"><i class="fas fa-comment"></i><span>空降评论</span></a><a class="rightMenu-item" href="javascript:rmf.copyWordsLink()"><i class="fa fa-link"></i><span>复制本文地址</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-to"><a class="rightMenu-item" href="javascript:rmf.openWithNewTab()"><i class="fa fa-window-restore"></i><span>新窗口打开</span></a><a class="rightMenu-item" id="menu-too" href="javascript:rmf.open()"><i class="fa fa-link"></i><span>转到链接</span></a><a class="rightMenu-item" href="javascript:rmf.copyLink()"><i class="fa fa-copy"></i><span>复制链接</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-img"><a class="rightMenu-item" href="javascript:rmf.saveAs()"><i class="fa fa-download"></i><span>保存图片</span></a><a class="rightMenu-item" href="javascript:rmf.openWithNewTab()"><i class="fa fa-window-restore"></i><span>在新窗口打开</span></a><a class="rightMenu-item" href="javascript:rmf.copyLink()"><i class="fa fa-copy"></i><span>复制图片链接</span></a></div><div class="rightMenu-group rightMenu-line"><a class="rightMenu-item" href="/about/"><i class="fa fa-sitemap"></i><span>有的没的</span></a><a class="rightMenu-item" href="/new/"><i class="fa fa-bell"></i><span>更新日志</span></a><a class="rightMenu-item" href="javascript:rmf.fullScreen();"><i class="fas fa-expand"></i><span>切换全屏</span></a><a class="rightMenu-item" href="javascript:window.print();"><i class="fa-solid fa-print"></i><span>打印页面</span></a></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="/js/tw_cn.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><script src="/js/search/local-search.js"></script><script>var preloader = {
  endLoading: () => {
    document.body.style.overflow = 'auto';
    document.getElementById('loading-box').classList.add("loaded")
  },
  initLoading: () => {
    document.body.style.overflow = '';
    document.getElementById('loading-box').classList.remove("loaded")

  }
}
window.addEventListener('load',preloader.endLoading())</script><div class="js-pjax"><script>function loadValine () {
  function initValine () {
    const valine = new Valine(Object.assign({
      el: '#vcomment',
      appId: 'ORERCz6nkgV8OYWw8FovgBhQ-gzGzoHsz',
      appKey: 'BNxukaAtCZZ0K6TZldZpkY9e',
      avatar: 'mp',
      placeholder: '🐕温馨提示本站不具有邮件提示功能，Juse每隔一段时间会在后台检查评论并回复',
      serverURLs: '',
      emojiMaps: "",
      path: window.location.pathname,
      visitor: false
    }, null))
  }

  if (typeof Valine === 'function') initValine() 
  else getScript('https://cdn.jsdelivr.net/npm/valine/dist/Valine.min.js').then(initValine)
}

if ('Valine' === 'Valine' || !true) {
  if (true) btf.loadComment(document.getElementById('vcomment'),loadValine)
  else setTimeout(loadValine, 0)
} else {
  function loadOtherComment () {
    loadValine()
  }
}</script></div><script src="https://cdn.staticfile.org/jquery/3.6.3/jquery.min.js"></script><script type="text/javascript" src="/js/rightmenu.js"></script><script type="text/javascript" src="/js/txmap.js"></script><script type="text/javascript" src="/js/copy12.js"></script><script type="text/javascript" src="/js/ldex.js"></script><script type="text/javascript" src="/js/runtime.js"></script><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/dist/activate-power-mode.min.js"></script><script>POWERMODE.colorful = true;
POWERMODE.shake = false;
POWERMODE.mobile = true;
document.body.addEventListener('input', POWERMODE);
</script><script async="" data-pjax="" src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div>
        <style>
            [bg-lazy] {
                background-image: none !important;
                background-color: #eee !important;
            }
        </style>
        <script>
            window.imageLazyLoadSetting = {
                isSPA: false,
                preloadRatio: 1,
                processImages: null,
            };
        </script><script>window.addEventListener("load",function(){var t=/\.(gif|jpg|jpeg|tiff|png)$/i,r=/^data:image\/[a-z]+;base64,/;Array.prototype.slice.call(document.querySelectorAll("img[data-original]")).forEach(function(a){var e=a.parentNode;"A"===e.tagName&&(e.href.match(t)||e.href.match(r))&&(e.href=a.dataset.original)})});</script><script>!function(r){r.imageLazyLoadSetting.processImages=t;var e=r.imageLazyLoadSetting.isSPA,n=r.imageLazyLoadSetting.preloadRatio||1,c=a();function a(){var t=Array.prototype.slice.call(document.querySelectorAll("img[data-original]")),e=Array.prototype.slice.call(document.querySelectorAll("[bg-lazy]"));return t.concat(e)}function t(){e&&(c=a());for(var t,o=0;o<c.length;o++)0<=(t=(t=c[o]).getBoundingClientRect()).bottom&&0<=t.left&&t.top<=(r.innerHeight*n||document.documentElement.clientHeight*n)&&function(){var t,e,n,a,i=c[o];e=function(){c=c.filter(function(t){return i!==t}),r.imageLazyLoadSetting.onImageLoaded&&r.imageLazyLoadSetting.onImageLoaded(i)},(t=i).hasAttribute("bg-lazy")?(t.removeAttribute("bg-lazy"),e&&e()):(n=new Image,a=t.getAttribute("data-original"),n.onload=function(){t.src=a,t.removeAttribute("data-original"),e&&e()},t.src!==a&&(n.src=a))}()}function i(){clearTimeout(t.tId),t.tId=setTimeout(t,500)}t(),document.addEventListener("scroll",i),r.addEventListener("resize",i),r.addEventListener("orientationchange",i)}(this);</script><script type="text/javascript" charset="utf-8" src="/js/lazyload-plugin/lazyload.intersectionObserver.min.js"></script></body></html>