<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>比较转录组分析（六）—— GO 富集分析与可视化 | Juse's Blog</title><meta name="keywords" content="生物信息学,生信,转录组,比较转录组学,GO富集分析"><meta name="author" content="Juse"><meta name="copyright" content="Juse"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="对于无参转录组进行的 GO 富集分析，内含 pipeline 只需运行脚本便可完成一套分析。">
<meta property="og:type" content="article">
<meta property="og:title" content="比较转录组分析（六）—— GO 富集分析与可视化">
<meta property="og:url" content="https://biojuse.com/2022/12/19/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E5%85%AD%EF%BC%89%E2%80%94%E2%80%94%20GO%20%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/index.html">
<meta property="og:site_name" content="Juse's Blog">
<meta property="og:description" content="对于无参转录组进行的 GO 富集分析，内含 pipeline 只需运行脚本便可完成一套分析。">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://biojuse.com/pic/cover2.jpg">
<meta property="article:published_time" content="2022-12-19T07:45:00.000Z">
<meta property="article:modified_time" content="2024-07-25T07:03:28.403Z">
<meta property="article:author" content="Juse">
<meta property="article:tag" content="生物信息学">
<meta property="article:tag" content="生信">
<meta property="article:tag" content="转录组">
<meta property="article:tag" content="比较转录组学">
<meta property="article:tag" content="GO富集分析">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://biojuse.com/pic/cover2.jpg"><link rel="shortcut icon" href="/pic/webiron.png"><link rel="canonical" href="https://biojuse.com/2022/12/19/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E5%85%AD%EF%BC%89%E2%80%94%E2%80%94%20GO%20%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/"><link rel="preconnect" href="//cdn.jsdelivr.net"><link rel="preconnect" href="//busuanzi.ibruce.info"><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":false,"languages":{"hits_empty":"找不到您查询的内容：${query}"}},
  translate: {"defaultEncoding":1,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: true,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '比较转录组分析（六）—— GO 富集分析与可视化',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2024-07-25 15:03:28'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><link rel="stylesheet" href="/css/mycss.css"><link rel="stylesheet" href="/css/bg.css"><link rel="stylesheet" href="/css/hp.css"><script src="https://cdn1.tianli0.top/npm/vue@2.6.14/dist/vue.min.js"></script><script src="https://cdn1.tianli0.top/npm/element-ui@2.15.6/lib/index.js"></script><link rel="stylesheet" href="https://cdn1.tianli0.top/npm/element-ui@2.15.6/packages/theme-chalk/lib/index.css"><!-- hexo injector head_end start --><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Zfour/Butterfly-double-row-display@1.00/cardlistpost.min.css">
<style>#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags:before {content:"\A";
  white-space: pre;}#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags > .article-meta__separator{display:none}</style>
<link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiperstyle.css" media="print" onload="this.media='all'"><!-- hexo injector head_end end --><meta name="generator" content="Hexo 5.4.2"><link rel="alternate" href="/atom.xml" title="Juse's Blog" type="application/atom+xml">
</head><body><div id="loading-box"><div class="loading-left-bg"></div><div class="loading-right-bg"></div><div class="spinner-box"><div class="configure-border-1"><div class="configure-core"></div></div><div class="configure-border-2"><div class="configure-core"></div></div><div class="loading-word">加载中...</div></div></div><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/pic/cover.gif" data-original="/pic/gitav.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">51</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">18</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">10</div></a></div><hr><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fa fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> 文章</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fa fa-archive"></i><span> 归档</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fa fa-tags"></i><span> 标签</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fa fa-folder-open"></i><span> 分类</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/messageboard/"><i class="fa-fw fa fa-paper-plane"></i><span> 留言板</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fa fa-link"></i><span> 友链</span></a></div><div class="menus_item"><a class="site-page" href="/myself/"><i class="fa-fw fa fa-id-card"></i><span> 我自己</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fa fa-sitemap"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-shoe-prints"></i><span> 杂项</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/talk/"><i class="fa-fw fa fa-book"></i><span> 琐碎的日常</span></a></li><li><a class="site-page child" href="/new/"><i class="fa-fw fa fa-bell"></i><span> 更新日志</span></a></li><li><a class="site-page child" href="/paper/"><i class="fa-fw fa fa-sticky-note"></i><span> 文献笔记阁</span></a></li></ul></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/pic/cover2.jpg')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Juse's Blog</a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fa fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> 文章</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fa fa-archive"></i><span> 归档</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fa fa-tags"></i><span> 标签</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fa fa-folder-open"></i><span> 分类</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/messageboard/"><i class="fa-fw fa fa-paper-plane"></i><span> 留言板</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fa fa-link"></i><span> 友链</span></a></div><div class="menus_item"><a class="site-page" href="/myself/"><i class="fa-fw fa fa-id-card"></i><span> 我自己</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fa fa-sitemap"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-shoe-prints"></i><span> 杂项</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/talk/"><i class="fa-fw fa fa-book"></i><span> 琐碎的日常</span></a></li><li><a class="site-page child" href="/new/"><i class="fa-fw fa fa-bell"></i><span> 更新日志</span></a></li><li><a class="site-page child" href="/paper/"><i class="fa-fw fa fa-sticky-note"></i><span> 文献笔记阁</span></a></li></ul></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">比较转录组分析（六）—— GO 富集分析与可视化</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2022-12-19T07:45:00.000Z" title="发表于 2022-12-19 15:45:00">2022-12-19</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2024-07-25T07:03:28.403Z" title="更新于 2024-07-25 15:03:28">2024-07-25</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%AD%A6%E4%B9%A0/">学习</a><i class="fas fa-angle-right post-meta-separator"></i><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%AD%A6%E4%B9%A0/%E7%94%9F%E4%BF%A1/">生信</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">字数总计:</span><span class="word-count">4.2k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">阅读时长:</span><span>16分钟</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="比较转录组分析（六）—— GO 富集分析与可视化"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><hr>
<blockquote>
<div id="portalContainer"></div>
<script src="/js/ctlink.js"></script>
<script> generatePortalLinks(6);</script>
</blockquote>
<h2 id="前言"><a href="#前言" class="headerlink" title="前言"></a>前言</h2><p>GO 富集分析的用处非常广泛，在这里我们将用来做差异表达基因的分析，在其他情况下也有可能是做例如正选择基因的富集分析、高突变率基因的富集分析、基因家族的富集分析等等一系列基因群的功能分析，不过着重点不同。</p>
<h3 id="更新日志"><a href="#更新日志" class="headerlink" title="更新日志"></a>更新日志</h3><p>2022.12.22    更改了绘图脚本，新的气泡图以 GeneRatio 作为 x 轴标签（更改版本在 github 可见，文中未修改）。</p>
<p>2023.06.26    解封文章，简化了部分内容。</p>
<h3 id="出现的一些词"><a href="#出现的一些词" class="headerlink" title="出现的一些词"></a>出现的一些词</h3><p><strong>DAVID</strong>：全称 <strong>D</strong>atabase for <strong>A</strong>nnotation, <strong>V</strong>isualization and <strong>I</strong>ntegrated <strong>D</strong>iscovery，是一个广泛使用的富集分析平台，此外还可以在这里进行诸多 id 的转换。</p>
<p>链接：<a target="_blank" rel="noopener" href="https://david.ncifcrf.gov/">https://david.ncifcrf.gov/</a></p>
<h2 id="富集分析过程"><a href="#富集分析过程" class="headerlink" title="富集分析过程"></a>富集分析过程</h2><h3 id="需要明白的概念"><a href="#需要明白的概念" class="headerlink" title="需要明白的概念"></a>需要明白的概念</h3><p>什么是富集分析？关于这一点在网上的诸多文章中都有解释，但我想要以一个更为通俗易懂的说法去进行解释。</p>
<p>打个比方，我们调查了 5000 个人的兴趣爱好，其中有 200 个人都很喜欢玩电脑。</p>
<p>然后我们挑出了一个我们所感兴趣的人群，比方说我们从这 5000 个人中挑出来 100 个宅。</p>
<p>然后我们发现在这 100 个宅中，竟然有高达 50 个人喜欢玩电脑。</p>
<p>于是乎，在我们的调查的所有人（即背景集）中，喜欢玩电脑的只有 200/5000 = 1/25 = 4%。</p>
<p>但在我们所感兴趣的人群（即前景集）中，喜欢玩电脑的就已经达到 50/100 = 1/2 = 50%。</p>
<p>所以我们可以认为，我们感兴趣的人群 —— 宅，他们相较于所有人群普遍而言，会更喜欢玩电脑。</p>
<h3 id="需要准备的文件"><a href="#需要准备的文件" class="headerlink" title="需要准备的文件"></a>需要准备的文件</h3><p>无参转录组需要的文件有：</p>
<ul>
<li>前景基因集（在这里是<strong>差异表达基因列表</strong>）</li>
<li>背景基因注释集（在这里是<strong>组装的注释</strong>）</li>
</ul>
<p>如果后续需要使用 R 进行则还需要（本文的教程源自 <a target="_blank" rel="noopener" href="https://zhuanlan.zhihu.com/p/196761601">【生信小课堂】非模式物种或无参转录组GO注释与富集 from 知乎</a>）：</p>
<ul>
<li>GO term 的信息（下载地址：<a target="_blank" rel="noopener" href="http://geneontology.org/docs/download-ontology/#subsets">http://geneontology.org/docs/download-ontology/#subsets</a> 里的 <a target="_blank" rel="noopener" href="http://geneontology.org/docs/download-ontology/#go_basic">go-basic.obo</a>）</li>
</ul>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">wget http://current.geneontology.org/ontology/go-basic.obo</span></span><br></pre></td></tr></tbody></table></figure>

<p>然后使用脚本将其转化为 GO term list：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#get_go_term.py</span></span><br><span class="line"><span class="keyword">import</span> sys </span><br><span class="line">raw_file = <span class="built_in">open</span>(sys.argv[<span class="number">1</span>]).read()  </span><br><span class="line"><span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">"go_term.list"</span>,<span class="string">"w"</span>) <span class="keyword">as</span> output:</span><br><span class="line">     <span class="keyword">for</span> go_term <span class="keyword">in</span> raw_file.split(<span class="string">"[Term]"</span>):</span><br><span class="line">         go_id = <span class="string">''</span></span><br><span class="line">         name = <span class="string">''</span> </span><br><span class="line">         namespace = <span class="string">''</span></span><br><span class="line">         <span class="keyword">for</span> line <span class="keyword">in</span> go_term.split(<span class="string">"\n"</span>):</span><br><span class="line">             <span class="keyword">if</span> <span class="built_in">str</span>(line).startswith(<span class="string">"id"</span>) <span class="keyword">and</span> <span class="string">"GO:"</span> <span class="keyword">in</span> line:</span><br><span class="line">                 go_id = line.rstrip().split(<span class="string">" "</span>)[<span class="number">1</span>]</span><br><span class="line">             <span class="keyword">if</span> <span class="built_in">str</span>(line).startswith(<span class="string">"name:"</span>):</span><br><span class="line">                 name = line.rstrip().split(<span class="string">": "</span>)[<span class="number">1</span>]</span><br><span class="line">             <span class="keyword">if</span> <span class="built_in">str</span>(line).startswith(<span class="string">"namespace"</span>):</span><br><span class="line">                 namespace = line.rstrip().split(<span class="string">" "</span>)[<span class="number">1</span>]</span><br><span class="line">         term = go_id + <span class="string">'\t'</span> + name + <span class="string">'\t'</span> + namespace + <span class="string">'\n'</span></span><br><span class="line">         <span class="keyword">if</span> <span class="string">''</span> != go_id:</span><br><span class="line">             output.write(term)</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">python get_go_term.py go-basic.obo</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">会输出一个叫做 go_term.list 的文件</span></span><br></pre></td></tr></tbody></table></figure>

<p>然后就能得到 GO term list，或者你可以在我的 github 上直接下载（版本不是最新，慎重考虑），传送门点<a target="_blank" rel="noopener" href="https://github.com/JuseTiZ/PyScript-for-CT/blob/main/go_term.list">这里</a>。</p>
<p>此后我们可以通过 R 或者 DAVID 进行富集分析。</p>
<h3 id="通过-R-实现-——-clusterProfiler"><a href="#通过-R-实现-——-clusterProfiler" class="headerlink" title="通过 R 实现 —— clusterProfiler"></a>通过 R 实现 —— clusterProfiler</h3><p>主要是通过 R 的 ClusterProfiler 包进行，因此在进行分析前需要把这个包装好：</p>
<figure class="highlight r"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># R version 3.5以上版本:</span></span><br><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">"clusterProfiler"</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># R version 3.5以下版本:</span></span><br><span class="line">source<span class="punctuation">(</span><span class="string">"http://bioconductor.org/biocLite.R"</span><span class="punctuation">)</span></span><br><span class="line">biocLite<span class="punctuation">(</span><span class="string">"clusterProfiler"</span><span class="punctuation">)</span></span><br></pre></td></tr></tbody></table></figure>

<p>具体的安装教程可以 Bing 搜索，安装好后就可以在 R 上进行后续的分析，也可以在装了 R 的服务器上进行。</p>
<p>用自己电脑的 R 画图是完全可以的，但如果涉及到的要分析的物种比较多时就会显得比较麻烦，所以在这里我制作了一个 python pipeline 在服务器上运行直接得到想要的结果：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br><span class="line">229</span><br><span class="line">230</span><br><span class="line">231</span><br><span class="line">232</span><br><span class="line">233</span><br><span class="line">234</span><br><span class="line">235</span><br><span class="line">236</span><br><span class="line">237</span><br><span class="line">238</span><br><span class="line">239</span><br><span class="line">240</span><br><span class="line">241</span><br><span class="line">242</span><br><span class="line">243</span><br><span class="line">244</span><br><span class="line">245</span><br><span class="line">246</span><br><span class="line">247</span><br><span class="line">248</span><br><span class="line">249</span><br><span class="line">250</span><br><span class="line">251</span><br><span class="line">252</span><br><span class="line">253</span><br><span class="line">254</span><br><span class="line">255</span><br><span class="line">256</span><br><span class="line">257</span><br><span class="line">258</span><br><span class="line">259</span><br><span class="line">260</span><br><span class="line">261</span><br><span class="line">262</span><br><span class="line">263</span><br><span class="line">264</span><br><span class="line">265</span><br><span class="line">266</span><br><span class="line">267</span><br><span class="line">268</span><br><span class="line">269</span><br><span class="line">270</span><br><span class="line">271</span><br><span class="line">272</span><br><span class="line">273</span><br><span class="line">274</span><br><span class="line">275</span><br><span class="line">276</span><br><span class="line">277</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -*- coding: utf-8 -*-</span></span><br><span class="line"><span class="comment"># enrichment_plot.py</span></span><br><span class="line"><span class="comment"># Verson: 1.1</span></span><br><span class="line"><span class="comment"># Author: Juse</span></span><br><span class="line"><span class="comment"># Use R to plot enrichment dotplot.</span></span><br><span class="line"><span class="comment"># Usage: See github.</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> argparse</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> time</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_args</span>():</span><br><span class="line"></span><br><span class="line">	parser = argparse.ArgumentParser(description = <span class="string">"This is used to plot enrichment dotplot."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-gef"</span>, <span class="string">"--genefore"</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"The gene list as foregroud."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-geb"</span>, <span class="string">"--geneback"</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"The gene list and annotation as backgroud."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-go"</span>, <span class="string">"--golist"</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"The go_term.list file."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-w"</span>, <span class="string">"--width"</span>, default = <span class="string">'8'</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"Width of plot, default = 8."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-l"</span>, <span class="string">"--height"</span>, default = <span class="string">'8'</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"Height of plot, default = 8."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-y"</span>, <span class="string">"--yaxis"</span>, default = <span class="string">"Description"</span>, choices = [<span class="string">'Description'</span>,<span class="string">'ID'</span>], <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"The yaxis of plot, use GO Description as default."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-o"</span>, <span class="string">"--output"</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, required = <span class="literal">True</span>, <span class="built_in">help</span> = <span class="string">"The output directory, required."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-t"</span>, <span class="string">"--title"</span>, default = <span class="string">'NULL'</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"The title of plot, default = NULL."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-c"</span>, <span class="string">"--csv"</span>, default = <span class="string">'NULL'</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"Just plot with file given."</span>)</span><br><span class="line">	args = parser.parse_args()</span><br><span class="line">	<span class="keyword">return</span> args</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_path</span>(<span class="params">args</span>):</span><br><span class="line">	</span><br><span class="line">	<span class="keyword">if</span> args.csv == <span class="string">'NULL'</span>:</span><br><span class="line">		paths = {</span><br><span class="line">			<span class="string">'fore_gene'</span>: os.path.abspath(args.genefore),</span><br><span class="line">			<span class="string">'back_gene'</span>: os.path.abspath(args.geneback),</span><br><span class="line">			<span class="string">'go_list'</span>: os.path.abspath(args.golist),</span><br><span class="line">			<span class="string">'output_dir'</span>: os.path.abspath(args.output),</span><br><span class="line">		}</span><br><span class="line">	<span class="keyword">else</span>:</span><br><span class="line">		paths = {</span><br><span class="line">			<span class="string">'output_dir'</span>: os.path.abspath(args.output),</span><br><span class="line">			<span class="string">'csv_dir'</span>: os.path.abspath(args.csv)</span><br><span class="line">		}</span><br><span class="line">     </span><br><span class="line">	<span class="keyword">return</span> paths</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">R_code</span>(<span class="params">args, paths, run_time</span>):</span><br><span class="line"></span><br><span class="line">	rcode = <span class="string">'''</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 加载包</span></span><br><span class="line"><span class="string">library("clusterProfiler")</span></span><br><span class="line"><span class="string">library("ggplot2")</span></span><br><span class="line"><span class="string">library("DOSE")</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 加载各个参数</span></span><br><span class="line"><span class="string">goterm = "%juse%1"</span></span><br><span class="line"><span class="string">backgene = "%juse%2"</span></span><br><span class="line"><span class="string">foregene = "%juse%3"</span></span><br><span class="line"><span class="string">wid = %juse%4</span></span><br><span class="line"><span class="string">hei = %juse%5</span></span><br><span class="line"><span class="string">title = "%juse%7"</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 生成背景注释集</span></span><br><span class="line"><span class="string">go_class &lt;- read.delim(goterm, header=FALSE, stringsAsFactors =FALSE) </span></span><br><span class="line"><span class="string">names(go_class) &lt;- c('ID','Description','Ontology')  </span></span><br><span class="line"><span class="string">go_anno &lt;- read.delim(backgene, header=FALSE, stringsAsFactors =FALSE)</span></span><br><span class="line"><span class="string">names(go_anno) &lt;- c('gene_id','ID')</span></span><br><span class="line"><span class="string">go_anno &lt;-merge(go_anno, go_class, by = 'ID', all.x = TRUE)</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 生成前景基因集</span></span><br><span class="line"><span class="string">gene_list &lt;- read.delim(foregene,header=FALSE) </span></span><br><span class="line"><span class="string">names(gene_list) &lt;- c('gene_id') </span></span><br><span class="line"><span class="string">gene_select &lt;- gene_list$gene_id  </span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 进行富集分析</span></span><br><span class="line"><span class="string">go_rich &lt;- enricher(gene = gene_select,</span></span><br><span class="line"><span class="string">					TERM2GENE = go_anno[c('ID','gene_id')],</span></span><br><span class="line"><span class="string">					TERM2NAME = go_anno[c('ID','Description')],</span></span><br><span class="line"><span class="string">					pvalueCutoff = 0.05,</span></span><br><span class="line"><span class="string">					pAdjustMethod = 'BH',</span></span><br><span class="line"><span class="string">					qvalueCutoff = 0.05,</span></span><br><span class="line"><span class="string">					maxGSSize = 200) </span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 提取富集分析结果制成表格</span></span><br><span class="line"><span class="string">plot_data = cbind(go_rich$ID, go_rich$Description, go_rich$GeneRatio, go_rich$BgRatio, go_rich$Count, go_rich$p.adjust, go_rich$ID, go_rich$geneID)</span></span><br><span class="line"><span class="string">colnames(plot_data) = c("ID", "Description", "GeneRatio", "BgRatio", "Count","qvalue", "Ontology", "GeneID")</span></span><br><span class="line"><span class="string">plot_data = data.frame(plot_data)</span></span><br><span class="line"><span class="string">for(i in 1:length(plot_data$Ontology)){</span></span><br><span class="line"><span class="string">  if(is.na(plot_data$Description[i])){</span></span><br><span class="line"><span class="string">	plot_data$Ontology[i] = 'NA'</span></span><br><span class="line"><span class="string">  }else{</span></span><br><span class="line"><span class="string">	plot_data$Ontology[i] = go_class$Ontology[go_class$ID == plot_data$ID[i]]</span></span><br><span class="line"><span class="string">  }</span></span><br><span class="line"><span class="string">}</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 处理数据</span></span><br><span class="line"><span class="string">plot_data_noNA = plot_data[complete.cases(plot_data$Description),]</span></span><br><span class="line"><span class="string">plot_data_noNA = transform(plot_data_noNA, Count = as.numeric(Count),</span></span><br><span class="line"><span class="string">						   qvalue = as.numeric(qvalue))</span></span><br><span class="line"><span class="string">plot_data_noNA = plot_data_noNA[order(plot_data_noNA$qvalue,-plot_data_noNA$Count),]</span></span><br><span class="line"><span class="string">plot_data_noNA$Ontology[plot_data_noNA$Ontology == "cellular_component"] = "Cellular Component"</span></span><br><span class="line"><span class="string">plot_data_noNA$Ontology[plot_data_noNA$Ontology == "biological_process"] = "Biological Process"</span></span><br><span class="line"><span class="string">plot_data_noNA$Ontology[plot_data_noNA$Ontology == "molecular_function"] = "Molecular Function"</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">Top15 = c()</span></span><br><span class="line"><span class="string">CC = 0</span></span><br><span class="line"><span class="string">BP = 0</span></span><br><span class="line"><span class="string">MF = 0</span></span><br><span class="line"><span class="string">for(i in 1:length(plot_data_noNA$Ontology)){</span></span><br><span class="line"><span class="string">  if(plot_data_noNA$Ontology[i] == "Cellular Component"){</span></span><br><span class="line"><span class="string">	CC = CC + 1</span></span><br><span class="line"><span class="string">	if(CC &lt;= 5){</span></span><br><span class="line"><span class="string">	  Top15 = append(Top15,(i))</span></span><br><span class="line"><span class="string">	}</span></span><br><span class="line"><span class="string">  }</span></span><br><span class="line"><span class="string">  if(plot_data_noNA$Ontology[i] == "Biological Process"){</span></span><br><span class="line"><span class="string">	BP = BP + 1</span></span><br><span class="line"><span class="string">	if(BP &lt;= 5){</span></span><br><span class="line"><span class="string">	  Top15 = append(Top15,(i))</span></span><br><span class="line"><span class="string">	}</span></span><br><span class="line"><span class="string">  }</span></span><br><span class="line"><span class="string">  if(plot_data_noNA$Ontology[i] == "Molecular Function"){</span></span><br><span class="line"><span class="string">	MF = MF + 1</span></span><br><span class="line"><span class="string">	if(MF &lt;= 5){</span></span><br><span class="line"><span class="string">	  Top15 = append(Top15,(i))</span></span><br><span class="line"><span class="string">	}</span></span><br><span class="line"><span class="string">  }</span></span><br><span class="line"><span class="string">}</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 进行气泡图绘制</span></span><br><span class="line"><span class="string">enrichment_plot = ggplot(plot_data_noNA[Top15,],aes(x = parse_ratio(GeneRatio),y = reorder(%juse%6, Count))) + geom_point(aes(size=Count,color=qvalue,)) +</span></span><br><span class="line"><span class="string">  scale_color_gradient(low = "red", high = "blue") + theme_bw() + ylab(NULL) + xlab("GeneRatio") + </span></span><br><span class="line"><span class="string">  facet_wrap(~Ontology, scale="free",ncol = 1,strip.position = "right") + </span></span><br><span class="line"><span class="string">  theme(text = element_text(size = 15))</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">enrichment_plot_top10 = ggplot(plot_data_noNA[1:10,],aes(x = parse_ratio(GeneRatio),y = reorder(%juse%6, Count))) + geom_point(aes(size=Count,color=qvalue,)) +</span></span><br><span class="line"><span class="string">  scale_color_gradient(low = "red", high = "blue") + theme_bw() + ylab(NULL) + xlab("GeneRatio") + </span></span><br><span class="line"><span class="string">  theme(text = element_text(size = 15))</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">if(title != "NULL"){</span></span><br><span class="line"><span class="string">  enrichment_plot = enrichment_plot + ggtitle(title)</span></span><br><span class="line"><span class="string">  enrichment_plot_top10 = enrichment_plot_top10 + ggtitle(title)</span></span><br><span class="line"><span class="string">}</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 保存图及富集分析表格</span></span><br><span class="line"><span class="string">pdf("output%1%",width = wid,height = hei)</span></span><br><span class="line"><span class="string">enrichment_plot</span></span><br><span class="line"><span class="string">dev.off()</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">pdf("output%2%",width = wid,height = hei)</span></span><br><span class="line"><span class="string">enrichment_plot_top10</span></span><br><span class="line"><span class="string">dev.off()</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">write.csv(plot_data,file="output%3%")</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">'''</span></span><br><span class="line"></span><br><span class="line">	r_code_use = rcode.replace(<span class="string">"%juse%1"</span>, paths[<span class="string">'go_list'</span>])\</span><br><span class="line">		.replace(<span class="string">"%juse%2"</span>, paths[<span class="string">'back_gene'</span>])\</span><br><span class="line">		.replace(<span class="string">"%juse%3"</span>, paths[<span class="string">'fore_gene'</span>])\</span><br><span class="line">		.replace(<span class="string">"%juse%4"</span>, args.width)\</span><br><span class="line">		.replace(<span class="string">"%juse%5"</span>, args.height)\</span><br><span class="line">		.replace(<span class="string">"%juse%6"</span>, args.yaxis)\</span><br><span class="line">		.replace(<span class="string">"%juse%7"</span>, args.title)\</span><br><span class="line">		.replace(<span class="string">"output%1%"</span>, <span class="string">f"<span class="subst">{paths[<span class="string">'output_dir'</span>]}</span>/<span class="subst">{run_time}</span>/enrichment_plot.pdf"</span>)\</span><br><span class="line">		.replace(<span class="string">"output%2%"</span>, <span class="string">f"<span class="subst">{paths[<span class="string">'output_dir'</span>]}</span>/<span class="subst">{run_time}</span>/enrichment_plot_top10.pdf"</span>)\</span><br><span class="line">		.replace(<span class="string">"output%3%"</span>, <span class="string">f"<span class="subst">{paths[<span class="string">'output_dir'</span>]}</span>/<span class="subst">{run_time}</span>/enrichment_analysis.csv"</span>)\</span><br><span class="line">	</span><br><span class="line">	<span class="keyword">return</span> r_code_use</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">plot_R</span>(<span class="params">args, paths, run_time</span>):</span><br><span class="line"></span><br><span class="line">	rcode = <span class="string">'''</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 加载包</span></span><br><span class="line"><span class="string">library("DOSE")</span></span><br><span class="line"><span class="string">library("ggplot2")</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 加载数据集</span></span><br><span class="line"><span class="string">plot_data_noNA = read.csv("%juse%1", header=TRUE)</span></span><br><span class="line"><span class="string">wid = %juse%4</span></span><br><span class="line"><span class="string">hei = %juse%5</span></span><br><span class="line"><span class="string">title = "%juse%7"</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 处理数据</span></span><br><span class="line"><span class="string">plot_data_noNA = plot_data_noNA[complete.cases(plot_data_noNA$Description),]</span></span><br><span class="line"><span class="string">plot_data_noNA = transform(plot_data_noNA, Count = as.numeric(Count),</span></span><br><span class="line"><span class="string">						   qvalue = as.numeric(qvalue))</span></span><br><span class="line"><span class="string">plot_data_noNA = plot_data_noNA[order(plot_data_noNA$qvalue,-plot_data_noNA$Count),]</span></span><br><span class="line"><span class="string">plot_data_noNA$Ontology[plot_data_noNA$Ontology == "cellular_component"] = "Cellular Component"</span></span><br><span class="line"><span class="string">plot_data_noNA$Ontology[plot_data_noNA$Ontology == "biological_process"] = "Biological Process"</span></span><br><span class="line"><span class="string">plot_data_noNA$Ontology[plot_data_noNA$Ontology == "molecular_function"] = "Molecular Function"</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">Top15 = c()</span></span><br><span class="line"><span class="string">CC = 0</span></span><br><span class="line"><span class="string">BP = 0</span></span><br><span class="line"><span class="string">MF = 0</span></span><br><span class="line"><span class="string">for(i in 1:length(plot_data_noNA$Ontology)){</span></span><br><span class="line"><span class="string">  if(plot_data_noNA$Ontology[i] == "Cellular Component"){</span></span><br><span class="line"><span class="string">	CC = CC + 1</span></span><br><span class="line"><span class="string">	if(CC &lt;= 5){</span></span><br><span class="line"><span class="string">	  Top15 = append(Top15,(i))</span></span><br><span class="line"><span class="string">	}</span></span><br><span class="line"><span class="string">  }</span></span><br><span class="line"><span class="string">  if(plot_data_noNA$Ontology[i] == "Biological Process"){</span></span><br><span class="line"><span class="string">	BP = BP + 1</span></span><br><span class="line"><span class="string">	if(BP &lt;= 5){</span></span><br><span class="line"><span class="string">	  Top15 = append(Top15,(i))</span></span><br><span class="line"><span class="string">	}</span></span><br><span class="line"><span class="string">  }</span></span><br><span class="line"><span class="string">  if(plot_data_noNA$Ontology[i] == "Molecular Function"){</span></span><br><span class="line"><span class="string">	MF = MF + 1</span></span><br><span class="line"><span class="string">	if(MF &lt;= 5){</span></span><br><span class="line"><span class="string">	  Top15 = append(Top15,(i))</span></span><br><span class="line"><span class="string">	}</span></span><br><span class="line"><span class="string">  }</span></span><br><span class="line"><span class="string">}</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 进行气泡图绘制</span></span><br><span class="line"><span class="string">enrichment_plot = ggplot(plot_data_noNA[Top15,],aes(x = parse_ratio(GeneRatio),y = reorder(%juse%6, Count))) + geom_point(aes(size=Count,color=qvalue,)) +</span></span><br><span class="line"><span class="string">  scale_color_gradient(low = "red", high = "blue") + theme_bw() + ylab(NULL) + xlab("GeneRatio") + </span></span><br><span class="line"><span class="string">  facet_wrap(~Ontology, scale="free",ncol = 1,strip.position = "right") + </span></span><br><span class="line"><span class="string">  theme(text = element_text(size = 15))</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">enrichment_plot_top10 = ggplot(plot_data_noNA[1:10,],aes(x = parse_ratio(GeneRatio),y = reorder(%juse%6, Count))) + geom_point(aes(size=Count,color=qvalue,)) +</span></span><br><span class="line"><span class="string">  scale_color_gradient(low = "red", high = "blue") + theme_bw() + ylab(NULL) + xlab("GeneRatio") + </span></span><br><span class="line"><span class="string">  theme(text = element_text(size = 15))</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">if(title != "NULL"){</span></span><br><span class="line"><span class="string">  enrichment_plot = enrichment_plot + ggtitle(title)</span></span><br><span class="line"><span class="string">  enrichment_plot_top10 = enrichment_plot_top10 + ggtitle(title)</span></span><br><span class="line"><span class="string">}</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"># 保存图及富集分析表格</span></span><br><span class="line"><span class="string">pdf("output%1%",width = wid,height = hei)</span></span><br><span class="line"><span class="string">enrichment_plot</span></span><br><span class="line"><span class="string">dev.off()</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">pdf("output%2%",width = wid,height = hei)</span></span><br><span class="line"><span class="string">enrichment_plot_top10</span></span><br><span class="line"><span class="string">dev.off()</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">'''</span></span><br><span class="line"></span><br><span class="line">	r_code_use = rcode.replace(<span class="string">"%juse%1"</span>, paths[<span class="string">'csv_dir'</span>])\</span><br><span class="line">		.replace(<span class="string">"%juse%4"</span>, args.width)\</span><br><span class="line">		.replace(<span class="string">"%juse%5"</span>, args.height)\</span><br><span class="line">		.replace(<span class="string">"%juse%6"</span>, args.yaxis)\</span><br><span class="line">		.replace(<span class="string">"%juse%7"</span>, args.title)\</span><br><span class="line">		.replace(<span class="string">"output%1%"</span>, <span class="string">f"<span class="subst">{paths[<span class="string">'output_dir'</span>]}</span>/<span class="subst">{run_time}</span>/enrichment_plot.pdf"</span>)\</span><br><span class="line">		.replace(<span class="string">"output%2%"</span>, <span class="string">f"<span class="subst">{paths[<span class="string">'output_dir'</span>]}</span>/<span class="subst">{run_time}</span>/enrichment_plot_top10.pdf"</span>)\</span><br><span class="line">		</span><br><span class="line">	<span class="keyword">return</span> r_code_use</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">run_r_script</span>(<span class="params">r_code, paths, run_time</span>):</span><br><span class="line">	os.makedirs(<span class="string">f"<span class="subst">{paths[<span class="string">'output_dir'</span>]}</span>/<span class="subst">{run_time}</span>"</span>, exist_ok=<span class="literal">True</span>)</span><br><span class="line">	script_path = <span class="string">f"<span class="subst">{paths[<span class="string">'output_dir'</span>]}</span>/<span class="subst">{run_time}</span>/enrichment_plot.R"</span></span><br><span class="line">	<span class="keyword">with</span> <span class="built_in">open</span>(script_path, <span class="string">'w'</span>) <span class="keyword">as</span> f:</span><br><span class="line">		f.write(r_code)</span><br><span class="line">	os.system(<span class="string">f'Rscript <span class="subst">{script_path}</span>'</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">main</span>():</span><br><span class="line"></span><br><span class="line">	run_time = time.strftime(<span class="string">'%Y%m%d-%H%M%S'</span>)</span><br><span class="line">	args = get_args()</span><br><span class="line">	paths = get_path(args)</span><br><span class="line">	r_code = R_code(args, paths, run_time) <span class="keyword">if</span> args.csv == <span class="string">'NULL'</span> <span class="keyword">else</span> plot_R(args, paths, run_time)</span><br><span class="line">	run_r_script(r_code, paths, run_time)</span><br><span class="line">	<span class="built_in">print</span>(<span class="string">f'Finished! See output at <span class="subst">{paths[<span class="string">"output_dir"</span>]}</span>'</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">"__main__"</span>:</span><br><span class="line">	main()</span><br></pre></td></tr></tbody></table></figure>

<p>脚本的下载可以去 <a target="_blank" rel="noopener" href="https://github.com/JuseTiZ/PyScript-for-CT">github</a>，如果无法在服务器上使用 clusterProfiler 也可以把中间那一段 R 代码抄下来在自己的电脑上运行，只需要换掉一些特定的字符即可。</p>
<p><strong>运行例 ①：</strong></p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">python enrichment_plot.py -gef all_up.txt \</span></span><br><span class="line"><span class="language-bash">-geb onego.txt -go go_term.list \</span></span><br><span class="line"><span class="language-bash">-o outputdir \</span></span><br><span class="line"><span class="language-bash"><span class="comment">#-t Juse -l 8 -w 8</span></span></span><br></pre></td></tr></tbody></table></figure>

<p>使用所必须调用的参数：</p>
<p><code>-gef</code> 分析使用的前景基因集，例如上调或下调的差异表达基因。</p>
<p><code>-geb</code> 分析使用的背景基因注释集，即之前文章中所得到的一基因对一 GO term 文件。</p>
<p><code>-go</code> 在需要准备的文件中有提及到的 go_term.list。</p>
<p><code>-o</code> 输出文件夹。</p>
<p><strong>运行例 ②：</strong></p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">python enrichment_plot.py -c enrichment_analysis.csv \</span></span><br><span class="line"><span class="language-bash">-o outputdir \</span></span><br><span class="line"><span class="language-bash"><span class="comment">#-t Juse -l 8 -w 8</span></span></span><br></pre></td></tr></tbody></table></figure>

<p>当使用已有的 csv 进行分析时只需调用该 csv 文件即可（列名必须与运行例 ① 中输出的保持一致）。</p>
<p>在该情况下只会输出对应的图及原始 R 代码，不会再次输出 csv 文件。</p>
<p>可选参数：</p>
<p><code>-t</code> 图的标题，默认不设标题；<code>-l</code> <code>-w</code> 图的高和宽，默认都为 8；<code>-y</code> 当为 ID 时，作图的 y 轴标签会替换为 GO 号，默认为 Description 。</p>
<p>以<strong>运行例 ①</strong> 为标准，运行结束后，会在 outputdir 中出现一个以运行时间作为名字的文件夹，里面包括了以下几个文件：</p>
<p><img src="/pic/cover.gif" data-original="/pic/output.png"></p>
<p><code>enrichment_analysis.csv</code> 为富集分析的结果，其中已被废弃的 NA 项未被去除（运行例 ② 不会输出该文件）。</p>
<p><code>enrichment_plot.pdf</code> 为富集分析的气泡图，其中去除了 NA 项进行分析，取每个 Ontology 前五显著的 GO term 进行展示（不足五个则取全部）。</p>
<p><code>enrichment_plot_top10.pdf</code> 为富集分析的气泡图，其中去除了 NA 项进行分析，取最显著的十个 GO term 进行了展示（不分 Ontology）。</p>
<p><code>enrichment_plot.R</code> 为该次分析的代码。</p>
<p><img src="/pic/cover.gif" data-original="/pic/enrichment_plot.png"></p>
<p><img src="/pic/cover.gif" data-original="/pic/enrichment_csv.png"></p>
<p>另：曾在 <strong>比较转录组分析（四）—— 组装的 GO 注释</strong> 中提到过，富集分析里<strong>可以考虑</strong>把一些相关的 term 给删掉一部分，仅保留具有代表性的那个进行可视化，这里可以基于运行例 ① 输出的表格进行修改，删除一些相关 GO 项后再通过运行例 ② 的方法直接可视化。</p>
<h3 id="通过-DAVID-实现"><a href="#通过-DAVID-实现" class="headerlink" title="通过 DAVID 实现"></a>通过 DAVID 实现</h3><p>DAVID 的分析会严格很多，可能会导致我们得不到什么想要的结果，但是这种严格又未必不是件好事。</p>
<p>首先 DAVID 的 Document 可以看这里：<a target="_blank" rel="noopener" href="https://david.ncifcrf.gov/content.jsp?file=functional_annotation.html">DAVID Bioinformatics Resources (ncifcrf.gov)</a>。</p>
<p>为什么说 DAVID 严格？因为它在进行分析的时候首先<strong>把重复项删除了</strong>，其次还<strong>把所有原始 ID 中的冗余给去除了</strong>，于是乎很多原来在 clusterProfiler 的分析中具有富集现象的 term 在 DAVID 就失去了显著性。</p>
<p>这里我就以上面 clusterProfiler 中使用的数据为例，分享如何利用 DAVID 进行 GO 富集分析。</p>
<p>首先，我们要提取出原来我们 <code>blast against swissprot</code> 得到的注释信息（<code>blastx.outfmt6</code>）：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">cut</span> -f 2 blastx.outfmt6 &gt; back.txt</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash"><span class="built_in">cut</span> -f 2 blastp.outfmt6 &gt; back.txt</span></span><br><span class="line"><span class="meta prompt_"># </span><span class="language-bash">使用 blastp 的结果也行，不过需要改一下下面的代码。</span></span><br></pre></td></tr></tbody></table></figure>

<p>得到背景基因的注释集后，我们再通过上调基因和下调基因的列表，得到背景基因注释集：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">cat</span> all_up.txt | <span class="keyword">while</span> <span class="built_in">read</span> line;</span></span><br><span class="line">do</span><br><span class="line">grep -e "$line[[:space:]]" blastx.outfmt6 | cut -f 2 &gt;&gt; fore.txt</span><br><span class="line">done</span><br></pre></td></tr></tbody></table></figure>

<p>有了背景基因和前景基因注释信息后，就可以把它输入进 DAVID 进行分析。</p>
<p><img src="/pic/cover.gif" data-original="/pic/david1.png"></p>
<p><img src="/pic/cover.gif" data-original="/pic/david2.png"></p>
<p>Gene List 输入上调或下调的差异表达基因集，Background 输入背景基因注释集，它们是这样的（一行一注释）：</p>
<p><img src="/pic/cover.gif" data-original="/pic/david3.png"></p>
<p>在 Select Identifier 上选择 Uniprot ID。</p>
<p><img src="/pic/cover.gif" data-original="/pic/david4.png"></p>
<p>它会提示检测到基因集中有很多不同的物种，略过即可。</p>
<p>弄完以后点击右侧 Step2 下面的内容进入下一步分析，如果只需要做 GO 富集分析，则可以先按 clear all，然后勾选这三项：</p>
<p><img src="/pic/cover.gif" data-original="/pic/david5.png"></p>
<p>然后移到最下面有三个选项，其中 <strong>Functional Annotation Chart</strong> 就是 GO 富集结果了。</p>
<p>DAVID 采取的检验方法为费希尔精确检验，基于对应的 DAVID ID 进行计算 p 值和对应矫正值。以该文中涉及的数据为例，显著富集的 term 如果要以 <code>Benjamini</code> 矫正值为准，那么显著富集的 term 仅有三个，远远低于在 clusterProfiler 中识别到的 44 个。</p>
<p>后续的可视化步骤同样用 <code>R ggplot2</code> 实现，可以参考上文的代码进行。</p>
<h2 id="后记"><a href="#后记" class="headerlink" title="后记"></a>后记</h2><p><strong>下一篇预告</strong>：比较转录组分析（七） —— 系统发育分析。</p>
<p><strong>可能会出的</strong>：基于 Cytoscape 的富集网络图实现。</p>
<blockquote>
<p>如果有时间的话…… —— Juse</p>
</blockquote>
<h3 id="关于-BH-矫正"><a href="#关于-BH-矫正" class="headerlink" title="关于 BH 矫正"></a>关于 BH 矫正</h3><p>我不打算用过于科学的说法来解释做这个的动机，所以我找了一个很有意思的图片来展示：</p>
<p><img src="/pic/cover.gif" data-original="/pic/BH.jpg"></p>
<p>我再来说一段当初在生物统计学课上听老师讲的故事（内容可能有差异但大致相似）：</p>
<blockquote>
<p>有一只被称作神奇章鱼的章鱼，在进行拳击比赛时让它进行猜测谁会赢，结果在二十次比赛里有十九次它的触手都伸到了胜利方的牌子上，也就是说二十次它猜对了十九次，大家觉得这只章鱼是不是真的是一只具有魔力的、神奇的章鱼？</p>
<p>大家都会觉得比较神奇吧，19/20，95% 的正确率。</p>
<p>但是我再跟大家说一个事，事实上这个世界上有数千数万只章鱼正在进行这种猜测，但是只有几只章鱼脱颖而出，那么这时候大家还会觉得这是一只神奇的章鱼吗？</p>
</blockquote>
<p>这个故事精准地切中了多重比较中存在的问题，也就是子集和整体之间的问题，一个事件的发生概率再怎么样低，如果它重复的次数多了，那么它总会在一次试验中发生并在那次试验中展现出显著性，但实际上放在整体上看如果相信了这个显著性就会给我们带来错误的结论，这也就是我们要进行 p 值矫正的原因。</p>
<p>BH 矫正的办法就是：<strong>adjust.p = p * n / k</strong></p>
<p>其中，p 是指原来的 p 值，n 是指这次矫正中有 n 个 p 值，k 是指将 p 值从小到大排列以后该 p 的顺序数。</p>
<p>当排在前面的 p 值在矫正后比在后面的 p 值大时，排在前面的 p 值又会重新矫正为和后面 p 值相同的值，这也就是在结果中经常能看到好几个矫正后的 p 值都相同的原因。</p>
<h3 id="clusterProfiler-在-linux-R-上的安装"><a href="#clusterProfiler-在-linux-R-上的安装" class="headerlink" title="clusterProfiler 在 linux R 上的安装"></a>clusterProfiler 在 linux R 上的安装</h3><p>这里简单阐述下我自己是如何解决这个问题的：</p>
<p>首先，每个 R 包都可以通过 conda 安装，可以在 Bing 上搜索关键词 xxx conda。</p>
<p>例如可以先搜索： clusterProfiler conda</p>
<p><img src="/pic/cover.gif" data-original="/pic/conda1.png"></p>
<p>点进去，选择上方的 Files，在里面看到不同版本的以及适应不同平台的 R 包文件。</p>
<p>如果在 linux 系统上安装，则选择开头为 <code>noarch</code> 或者 <code>linux</code> 的文件下载。</p>
<p><img src="/pic/cover.gif" data-original="/pic/conda2.png"></p>
<p>此外，如上图红框所标注为该版本 R 包构建时所基于的 R 版本，因此还需要根据自己的 R 版本选择合适的包。</p>
<p>例如如果自己的 R 版本是 4.1 的，那么这里就可以选一个 4.0 的下载安装。</p>
<p>此后在 linux 中输入下列命令用 conda 安装 R 包：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">conda install xxx.tar.bz2</span></span><br></pre></td></tr></tbody></table></figure>

<p>这样做仅是把这个包给安装上了，但是众多的依赖项依然没有完备，因此在安装好后进入 R 并尝试加载时，系统会提示缺少什么包。</p>
<p>此后把缺少的包一个一个用上述方式搜索然后再用 conda 安装，直到依赖项全部弄好为止。</p>
<p>当然，也可以尝试直接用 R 的 <code>BiocManager::install</code> 安装，但并不总能成功。</p>
<p>此外，如果安装某个依赖项以后在加载 clusterProfiler 时依然显示缺失该依赖项，换一个依赖项的版本重新安装可能能够解决这个问题。</p>
<p>全部安装后以后，再进行一次就不会报错了：</p>
<p><img src="/pic/cover.gif" data-original="/pic/conda4.png"></p>
<p>之前我束手无策时选择了这条道路，虽然很笨，但是管用。</p>
<!-- hexo injector body_end start --> <script data-pjax="">if(document.getElementById('recent-posts') && (location.pathname ==='/'|| '/' ==='all')){
    var parent = document.getElementById('recent-posts');
    var child = '<div class="recent-post-item" style="width:100%;height: auto"><div id="catalog_magnet"><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/生信/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📚 Juseの生信分享 (22)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/绘图/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">👩‍💻 Juseの绘图经验 (4)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/杂项/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🤹 Juseの琐碎教程 (12)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/文献阅读/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📕 Juseの文献笔记 (2)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/软件开发/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🧰 Juseの软件开发 (8)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/博客/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📑 Juseの博客轻改 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><a class="magnet_link_more"  target="_blank" rel="noopener" href="https://jusetiz.github.io/categories" style="flex:1;text-align: center;margin-bottom: 10px;">查看更多...</a></div></div>';
    console.log('已挂载magnet')
    parent.insertAdjacentHTML("afterbegin",child)}
     </script><style>#catalog_magnet{flex-wrap: wrap;display: flex;width:100%;justify-content:space-between;padding: 10px 10px 0 10px;align-content: flex-start;}.magnet_item{flex-basis: calc(33.333333333333336% - 5px);background: #f2f2f2;margin-bottom: 10px;border-radius: 8px;transition: all 0.2s ease-in-out;}.magnet_item:hover{background: #b30070}.magnet_link_more{color:#555}.magnet_link{color:black}.magnet_link:hover{color:white}@media screen and (max-width: 600px) {.magnet_item {flex-basis: 100%;}}.magnet_link_context{display:flex;padding: 10px;font-size:16px;transition: all 0.2s ease-in-out;}.magnet_link_context:hover{padding: 10px 20px;}</style>
    <style></style><script data-pjax="">
  function butterfly_swiper_injector_config(){
    var parent_div_git = document.getElementById('recent-posts');
    var item_html = '<div class="recent-post-item" style="height: auto;width: 100%"><div class="blog-slider swiper-container-fade swiper-container-horizontal" id="swiper_container"><div class="blog-slider__wrp swiper-wrapper" style="transition-duration: 0ms;"><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-09-01</span><a class="blog-slider__title" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt="">JuseKit（八） —— 计算转录组组装指标</a><div class="blog-slider__text">本次更新新增了计算转录组组装指标的功能，经实测比 Trinity 快且结果一致。</div><a class="blog-slider__button" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-08-04</span><a class="blog-slider__title" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt="">JuseKit（七） —— 绘制 GO 富集分析气泡图</a><div class="blog-slider__text">本次更新新增 GO 富集气泡图绘制方法，只需文件一键执行！</div><a class="blog-slider__button" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-07-12</span><a class="blog-slider__title" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt="">DiscoVista 可视化系统发育不一致</a><div class="blog-slider__text">使用合适的工具将系统发育的不一致更加清晰地呈现出来。</div><a class="blog-slider__button" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/05/04/关于 PAML 的一二三事/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-05-04</span><a class="blog-slider__title" href="2023/05/04/关于 PAML 的一二三事/" alt="">关于 PAML 的一二三事</a><div class="blog-slider__text">PAML 进行正选择检测、选择压力分析和物种分歧时间估计的详细介绍，包括诸多需要注意的事项。</div><a class="blog-slider__button" href="2023/05/04/关于 PAML 的一二三事/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-07-04</span><a class="blog-slider__title" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt="">JuseKit（六） —— 绘制火山图</a><div class="blog-slider__text">此次更新新增了绘制火山图的功能，仅需输入对应格式的文件即可一键绘制火山图。</div><a class="blog-slider__button" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-05-14</span><a class="blog-slider__title" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt="">JuseKit（五） —— 用于系统发育分析的序列过滤</a><div class="blog-slider__text">本次更新带来序列过滤功能，可根据序列长度及物种序列数过滤比对。</div><a class="blog-slider__button" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/wxyd1.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-04-06</span><a class="blog-slider__title" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt="">系统发育转录组学是否靠谱？</a><div class="blog-slider__text">向来只闻 Phylogenomics，那么 Phylotranscriptomics 又如何呢？</div><a class="blog-slider__button" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/slwd.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-03-13</span><a class="blog-slider__title" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt="">单基因 dN、dS 和 ω 的滑动窗口绘制方法</a><div class="blog-slider__text">绘制可以观察基因内不同部位选择情况的滑动窗口图，基于 Kaks_calculator 计算。</div><a class="blog-slider__button" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt="">详情   </a></div></div></div><div class="blog-slider__pagination swiper-pagination-clickable swiper-pagination-bullets"></div></div></div>';
    console.log('已挂载butterfly_swiper')
    parent_div_git.insertAdjacentHTML("afterbegin",item_html)
    }
  var elist = 'undefined'.split(',');
  var cpage = location.pathname;
  var epage = 'all';
  var flag = 0;

  for (var i=0;i<elist.length;i++){
    if (cpage.includes(elist[i])){
      flag++;
    }
  }

  if ((epage ==='all')&&(flag == 0)){
    butterfly_swiper_injector_config();
  }
  else if (epage === cpage){
    butterfly_swiper_injector_config();
  }
  </script><script defer="" src="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper.min.js"></script><script defer="" data-pjax="" src="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper_init.js"></script><!-- hexo injector body_end end --></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="https://biojuse.com">Juse</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="https://biojuse.com">https://biojuse.com</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://biojuse.com" target="_blank">Juse's Blog</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/">生物信息学</a><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1/">生信</a><a class="post-meta__tags" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">转录组</a><a class="post-meta__tags" href="/tags/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">比较转录组学</a><a class="post-meta__tags" href="/tags/GO%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/">GO富集分析</a></div><div class="post_share"><div class="social-share" data-image="/pic/cover2.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer=""></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2022/12/23/%E5%9F%BA%E4%BA%8E%20GraphPad%20Prism%20%E7%9A%84%E7%81%AB%E5%B1%B1%E5%9B%BE%E7%BB%98%E5%88%B6/"><img class="prev-cover" src="/pic/cover.gif" data-original="/pic/prism17.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">基于 GraphPad Prism 的火山图绘制</div></div></a></div><div class="next-post pull-right"><a href="/2022/12/11/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%BA%94%EF%BC%89%E2%80%94%E2%80%94%20%E8%BD%AC%E5%BD%95%E6%9C%AC%E5%AE%9A%E9%87%8F%E4%B8%8E%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/"><img class="next-cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">比较转录组分析（五）—— 转录本定量与差异表达分析</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><div><a href="/2022/12/23/%E5%9F%BA%E4%BA%8E%20GraphPad%20Prism%20%E7%9A%84%E7%81%AB%E5%B1%B1%E5%9B%BE%E7%BB%98%E5%88%B6/" title="基于 GraphPad Prism 的火山图绘制"><img class="cover" src="/pic/cover.gif" data-original="/pic/prism17.png" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">基于 GraphPad Prism 的火山图绘制</div></div></a></div><div><a href="/2022/11/01/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%B8%80%EF%BC%89%E2%80%94%E2%80%94%20%E5%89%8D%E6%8F%90%E5%8F%8A%E6%A6%82%E8%A6%81/" title="比较转录组分析（一）—— 前提及概要"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-07-06</div><div class="title">比较转录组分析（一）—— 前提及概要</div></div></a></div><div><a href="/2022/11/21/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%B8%89%EF%BC%89%E2%80%94%E2%80%94%20%E7%BB%84%E8%A3%85%E7%9A%84%E8%B4%A8%E9%87%8F%E6%A3%80%E6%B5%8B%E4%B8%8E%E5%8E%BB%E5%86%97%E4%BD%99/" title="比较转录组分析（三）—— 组装的质量检测与去冗余"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-10</div><div class="title">比较转录组分析（三）—— 组装的质量检测与去冗余</div></div></a></div><div><a href="/2022/11/06/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%BA%8C%EF%BC%89%E2%80%94%E2%80%94%20%E5%8E%9F%E5%A7%8B%E6%95%B0%E6%8D%AE%E7%9A%84%E8%B4%A8%E6%8E%A7%E4%B8%8E%E8%BD%AC%E5%BD%95%E7%BB%84%E7%BB%84%E8%A3%85/" title="比较转录组分析（二）—— 原始数据的质控与转录组组装"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">比较转录组分析（二）—— 原始数据的质控与转录组组装</div></div></a></div><div><a href="/2022/12/11/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%BA%94%EF%BC%89%E2%80%94%E2%80%94%20%E8%BD%AC%E5%BD%95%E6%9C%AC%E5%AE%9A%E9%87%8F%E4%B8%8E%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/" title="比较转录组分析（五）—— 转录本定量与差异表达分析"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-08-22</div><div class="title">比较转录组分析（五）—— 转录本定量与差异表达分析</div></div></a></div><div><a href="/2022/11/28/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E5%9B%9B%EF%BC%89%E2%80%94%E2%80%94%20%E7%BB%84%E8%A3%85%E7%9A%84%E6%B3%A8%E9%87%8A/" title="比较转录组分析（四）—— 组装的 GO 及 KEGG 注释"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-08-15</div><div class="title">比较转录组分析（四）—— 组装的 GO 及 KEGG 注释</div></div></a></div></div></div><hr><div id="post-comment"><div class="comment-head"><div class="comment-headline"><i class="fas fa-comments fa-fw"></i><span> 评论</span></div></div><div class="comment-wrap"><div><div class="vcomment" id="vcomment"></div></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/pic/cover.gif" data-original="/pic/gitav.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"></div><div class="author-info__name">Juse</div><div class="author-info__description">偶尔徒步行进倒也不错</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">51</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">18</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">10</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/JuseTiZ"><i class="fab fa-github"></i><span>Follow Juse</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="/qq.html" target="_blank" title=""><i class="fab fa-qq"></i></a><a class="social-icon" href="https://space.bilibili.com/31269335" target="_blank" title=""><i class="fab fa-bilibili"></i></a><a class="social-icon" href="https://github.com/JuseTiZ" target="_blank" title=""><i class="fab fa-github"></i></a><a class="social-icon" href="https://steamcommunity.com/id/291195857/" target="_blank" title=""><i class="fab fa-steam"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content"><p align="center"><b>人生是旷野，不是轨道</b><br>⭐⭐🗞️ 新闻速递 📰⭐⭐<br>会萤吗？会萤的！💪</p></div><div id="welcome-info"></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%89%8D%E8%A8%80"><span class="toc-number">1.</span> <span class="toc-text">前言</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%9B%B4%E6%96%B0%E6%97%A5%E5%BF%97"><span class="toc-number">1.1.</span> <span class="toc-text">更新日志</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%87%BA%E7%8E%B0%E7%9A%84%E4%B8%80%E4%BA%9B%E8%AF%8D"><span class="toc-number">1.2.</span> <span class="toc-text">出现的一些词</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E8%BF%87%E7%A8%8B"><span class="toc-number">2.</span> <span class="toc-text">富集分析过程</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%9C%80%E8%A6%81%E6%98%8E%E7%99%BD%E7%9A%84%E6%A6%82%E5%BF%B5"><span class="toc-number">2.1.</span> <span class="toc-text">需要明白的概念</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%9C%80%E8%A6%81%E5%87%86%E5%A4%87%E7%9A%84%E6%96%87%E4%BB%B6"><span class="toc-number">2.2.</span> <span class="toc-text">需要准备的文件</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%80%9A%E8%BF%87-R-%E5%AE%9E%E7%8E%B0-%E2%80%94%E2%80%94-clusterProfiler"><span class="toc-number">2.3.</span> <span class="toc-text">通过 R 实现 —— clusterProfiler</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E9%80%9A%E8%BF%87-DAVID-%E5%AE%9E%E7%8E%B0"><span class="toc-number">2.4.</span> <span class="toc-text">通过 DAVID 实现</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%90%8E%E8%AE%B0"><span class="toc-number">3.</span> <span class="toc-text">后记</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%85%B3%E4%BA%8E-BH-%E7%9F%AB%E6%AD%A3"><span class="toc-number">3.1.</span> <span class="toc-text">关于 BH 矫正</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#clusterProfiler-%E5%9C%A8-linux-R-%E4%B8%8A%E7%9A%84%E5%AE%89%E8%A3%85"><span class="toc-number">3.2.</span> <span class="toc-text">clusterProfiler 在 linux R 上的安装</span></a></li></ol></li></ol></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div id="ft"><div class="ft-item-1"><div class="t-top"><div class="t-t-l"><p class="ft-t t-l-t">想说的话😉</p><div class="bg-ad"><div>Genes are not destiny. While they play a large role in shaping how we act, how we feel, and how we interact with others, our environment and personal choices have just as much impact on our behavior.</div><div class="btn-xz-box"><a class="btn-xz" target="_blank" rel="noopener" href="https://habitica.com/static/home">游戏人生不游戏 🎯</a></div></div></div><div class="t-t-r"><p class="ft-t t-l-t">仙人指路😝</p><ul class="ft-links"><li><a target="_blank" rel="noopener" href="https://www.fomal.cc/">建站参考</a><a target="_blank" rel="noopener" href="https://emojixd.com/">表情大全</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/talk/">看看生活</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/messageboard/">我要留言</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/myself/">关于博主</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/archives/">文章归档</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/categories/">文章分类</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/tags/">文章标签</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/about/">有的没的</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/new/">更新日志</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/link/">友情链接</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/404.html">404页面</a></li></ul></div></div></div><div class="ft-item-2"><p class="ft-t">友人链接🔗</p><div class="ft-img-group"><div class="img-group-item"><a target="_blank" rel="noopener" href="https://linctanny.github.io/"><img src="/pic/cover.gif" data-original="https://linctanny.github.io/img/p7.jpeg" alt=""></a></div><div class="img-group-item"><a target="_blank" rel="noopener" href="https://robust-vase.github.io/"><img src="/pic/cover.gif" data-original="https://robust-vase.github.io/img/John_King.webp" alt=""></a></div><div class="img-group-item"><a target="_blank" rel="noopener" href="https://blog.stariverfeel.eu.org/"><img src="/pic/cover.gif" data-original="https://avatars.githubusercontent.com/u/26035995?v=4" alt=""></a></div></div></div></div><div class="copyright">©2022 - 2024  @ Juse</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo </a><span class="footer-separator">|</span><span> 主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly </a><span class="footer-separator">|</span><span> 部署 </span><a target="_blank" rel="noopener" href="https://github.com/">GitPage </a><span class="footer-separator">|</span><span> 加速 </span><a target="_blank" rel="noopener" href="https://dash.cloudflare.com/">Cloudflare </a><span class="footer-separator">|</span><span> </span><a href="https://icp.gov.moe/?keyword=20232400" target="_blank">萌ICP备20232400号</a></div><div id="workboard"></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="translateLink" type="button" title="简繁转换">简</button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><a id="to_comment" href="#post-comment" title="直达评论"><i class="fas fa-comments"></i></a><button id="go-up" type="button" title="回到顶部"><i class="fas fa-arrow-up"></i></button></div></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"></div></div><hr><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div class="js-pjax" id="rightMenu"><div class="rightMenu-group rightMenu-small"><a class="rightMenu-item" href="javascript:window.history.back();"><i class="fa fa-arrow-left"></i></a><a class="rightMenu-item" href="javascript:window.history.forward();"><i class="fa fa-arrow-right"></i></a><a class="rightMenu-item" href="javascript:window.location.reload();"><i class="fa fa-refresh"></i></a><a class="rightMenu-item" href="/"><i class="fa fa-home"></i></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-text"><a class="rightMenu-item" href="javascript:rmf.copySelect();"><i class="fa fa-copy"></i><span>复制</span></a><a class="rightMenu-item" href="javascript:window.open(&quot;https://cn.bing.com/search?q=&quot;+window.getSelection().toString());window.location.reload();"><i class="fa fa-search"></i><span>必应搜索</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-too"><a class="rightMenu-item" href="javascript:window.open(window.getSelection().toString());window.location.reload();"><i class="fa fa-link"></i><span>转到链接</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-paste"><a class="rightMenu-item" href="javascript:rmf.paste()"><i class="fa fa-copy"></i><span>粘贴</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-post"><a class="rightMenu-item" href="#post-comment"><i class="fas fa-comment"></i><span>空降评论</span></a><a class="rightMenu-item" href="javascript:rmf.copyWordsLink()"><i class="fa fa-link"></i><span>复制本文地址</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-to"><a class="rightMenu-item" href="javascript:rmf.openWithNewTab()"><i class="fa fa-window-restore"></i><span>新窗口打开</span></a><a class="rightMenu-item" id="menu-too" href="javascript:rmf.open()"><i class="fa fa-link"></i><span>转到链接</span></a><a class="rightMenu-item" href="javascript:rmf.copyLink()"><i class="fa fa-copy"></i><span>复制链接</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-img"><a class="rightMenu-item" href="javascript:rmf.saveAs()"><i class="fa fa-download"></i><span>保存图片</span></a><a class="rightMenu-item" href="javascript:rmf.openWithNewTab()"><i class="fa fa-window-restore"></i><span>在新窗口打开</span></a><a class="rightMenu-item" href="javascript:rmf.copyLink()"><i class="fa fa-copy"></i><span>复制图片链接</span></a></div><div class="rightMenu-group rightMenu-line"><a class="rightMenu-item" href="/about/"><i class="fa fa-sitemap"></i><span>有的没的</span></a><a class="rightMenu-item" href="/new/"><i class="fa fa-bell"></i><span>更新日志</span></a><a class="rightMenu-item" href="javascript:rmf.fullScreen();"><i class="fas fa-expand"></i><span>切换全屏</span></a><a class="rightMenu-item" href="javascript:window.print();"><i class="fa-solid fa-print"></i><span>打印页面</span></a></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="/js/tw_cn.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><script src="/js/search/local-search.js"></script><script>var preloader = {
  endLoading: () => {
    document.body.style.overflow = 'auto';
    document.getElementById('loading-box').classList.add("loaded")
  },
  initLoading: () => {
    document.body.style.overflow = '';
    document.getElementById('loading-box').classList.remove("loaded")

  }
}
window.addEventListener('load',preloader.endLoading())</script><div class="js-pjax"><script>function loadValine () {
  function initValine () {
    const valine = new Valine(Object.assign({
      el: '#vcomment',
      appId: 'ORERCz6nkgV8OYWw8FovgBhQ-gzGzoHsz',
      appKey: 'BNxukaAtCZZ0K6TZldZpkY9e',
      avatar: 'mp',
      placeholder: '🐕温馨提示本站不具有邮件提示功能，Juse每隔一段时间会在后台检查评论并回复',
      serverURLs: '',
      emojiMaps: "",
      path: window.location.pathname,
      visitor: false
    }, null))
  }

  if (typeof Valine === 'function') initValine() 
  else getScript('https://cdn.jsdelivr.net/npm/valine/dist/Valine.min.js').then(initValine)
}

if ('Valine' === 'Valine' || !true) {
  if (true) btf.loadComment(document.getElementById('vcomment'),loadValine)
  else setTimeout(loadValine, 0)
} else {
  function loadOtherComment () {
    loadValine()
  }
}</script></div><script src="https://cdn.staticfile.org/jquery/3.6.3/jquery.min.js"></script><script type="text/javascript" src="/js/rightmenu.js"></script><script type="text/javascript" src="/js/txmap.js"></script><script type="text/javascript" src="/js/copy12.js"></script><script type="text/javascript" src="/js/ldex.js"></script><script type="text/javascript" src="/js/runtime.js"></script><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/dist/activate-power-mode.min.js"></script><script>POWERMODE.colorful = true;
POWERMODE.shake = false;
POWERMODE.mobile = true;
document.body.addEventListener('input', POWERMODE);
</script><script async="" data-pjax="" src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div>
        <style>
            [bg-lazy] {
                background-image: none !important;
                background-color: #eee !important;
            }
        </style>
        <script>
            window.imageLazyLoadSetting = {
                isSPA: false,
                preloadRatio: 1,
                processImages: null,
            };
        </script><script>window.addEventListener("load",function(){var t=/\.(gif|jpg|jpeg|tiff|png)$/i,r=/^data:image\/[a-z]+;base64,/;Array.prototype.slice.call(document.querySelectorAll("img[data-original]")).forEach(function(a){var e=a.parentNode;"A"===e.tagName&&(e.href.match(t)||e.href.match(r))&&(e.href=a.dataset.original)})});</script><script>!function(r){r.imageLazyLoadSetting.processImages=t;var e=r.imageLazyLoadSetting.isSPA,n=r.imageLazyLoadSetting.preloadRatio||1,c=a();function a(){var t=Array.prototype.slice.call(document.querySelectorAll("img[data-original]")),e=Array.prototype.slice.call(document.querySelectorAll("[bg-lazy]"));return t.concat(e)}function t(){e&&(c=a());for(var t,o=0;o<c.length;o++)0<=(t=(t=c[o]).getBoundingClientRect()).bottom&&0<=t.left&&t.top<=(r.innerHeight*n||document.documentElement.clientHeight*n)&&function(){var t,e,n,a,i=c[o];e=function(){c=c.filter(function(t){return i!==t}),r.imageLazyLoadSetting.onImageLoaded&&r.imageLazyLoadSetting.onImageLoaded(i)},(t=i).hasAttribute("bg-lazy")?(t.removeAttribute("bg-lazy"),e&&e()):(n=new Image,a=t.getAttribute("data-original"),n.onload=function(){t.src=a,t.removeAttribute("data-original"),e&&e()},t.src!==a&&(n.src=a))}()}function i(){clearTimeout(t.tId),t.tId=setTimeout(t,500)}t(),document.addEventListener("scroll",i),r.addEventListener("resize",i),r.addEventListener("orientationchange",i)}(this);</script><script type="text/javascript" charset="utf-8" src="/js/lazyload-plugin/lazyload.intersectionObserver.min.js"></script></body></html>