<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>比较转录组分析（七）—— 系统发育分析 | Juse's Blog</title><meta name="keywords" content="生物信息学,生信,系统发育分析,比较转录组学"><meta name="author" content="Juse"><meta name="copyright" content="Juse"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="基于转录组组装进行不同物种间的系统发育分析，严格意义上属于系统发育转录组学范畴。">
<meta property="og:type" content="article">
<meta property="og:title" content="比较转录组分析（七）—— 系统发育分析">
<meta property="og:url" content="https://biojuse.com/2023/02/17/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%B8%83%EF%BC%89%E2%80%94%E2%80%94%20%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/index.html">
<meta property="og:site_name" content="Juse's Blog">
<meta property="og:description" content="基于转录组组装进行不同物种间的系统发育分析，严格意义上属于系统发育转录组学范畴。">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://biojuse.com/pic/cover2.jpg">
<meta property="article:published_time" content="2023-02-17T12:30:00.000Z">
<meta property="article:modified_time" content="2023-10-10T01:54:36.000Z">
<meta property="article:author" content="Juse">
<meta property="article:tag" content="生物信息学">
<meta property="article:tag" content="生信">
<meta property="article:tag" content="系统发育分析">
<meta property="article:tag" content="比较转录组学">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://biojuse.com/pic/cover2.jpg"><link rel="shortcut icon" href="/pic/webiron.png"><link rel="canonical" href="https://biojuse.com/2023/02/17/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%B8%83%EF%BC%89%E2%80%94%E2%80%94%20%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/"><link rel="preconnect" href="//cdn.jsdelivr.net"><link rel="preconnect" href="//busuanzi.ibruce.info"><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":false,"languages":{"hits_empty":"找不到您查询的内容：${query}"}},
  translate: {"defaultEncoding":1,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: true,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '比较转录组分析（七）—— 系统发育分析',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2023-10-10 09:54:36'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><link rel="stylesheet" href="/css/mycss.css"><link rel="stylesheet" href="/css/bg.css"><link rel="stylesheet" href="/css/hp.css"><script src="https://cdn1.tianli0.top/npm/vue@2.6.14/dist/vue.min.js"></script><script src="https://cdn1.tianli0.top/npm/element-ui@2.15.6/lib/index.js"></script><link rel="stylesheet" href="https://cdn1.tianli0.top/npm/element-ui@2.15.6/packages/theme-chalk/lib/index.css"><!-- hexo injector head_end start --><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Zfour/Butterfly-double-row-display@1.00/cardlistpost.min.css">
<style>#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags:before {content:"\A";
  white-space: pre;}#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags > .article-meta__separator{display:none}</style>
<link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiperstyle.css" media="print" onload="this.media='all'"><!-- hexo injector head_end end --><meta name="generator" content="Hexo 5.4.2"><link rel="alternate" href="/atom.xml" title="Juse's Blog" type="application/atom+xml">
</head><body><div id="loading-box"><div class="loading-left-bg"></div><div class="loading-right-bg"></div><div class="spinner-box"><div class="configure-border-1"><div class="configure-core"></div></div><div class="configure-border-2"><div class="configure-core"></div></div><div class="loading-word">加载中...</div></div></div><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/pic/cover.gif" data-original="/pic/gitav.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">52</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">18</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">10</div></a></div><hr><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fa fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> 文章</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fa fa-archive"></i><span> 归档</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fa fa-tags"></i><span> 标签</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fa fa-folder-open"></i><span> 分类</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/messageboard/"><i class="fa-fw fa fa-paper-plane"></i><span> 留言板</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fa fa-link"></i><span> 友链</span></a></div><div class="menus_item"><a class="site-page" href="/myself/"><i class="fa-fw fa fa-id-card"></i><span> 我自己</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fa fa-sitemap"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-shoe-prints"></i><span> 杂项</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/talk/"><i class="fa-fw fa fa-book"></i><span> 琐碎的日常</span></a></li><li><a class="site-page child" href="/new/"><i class="fa-fw fa fa-bell"></i><span> 更新日志</span></a></li><li><a class="site-page child" href="/paper/"><i class="fa-fw fa fa-sticky-note"></i><span> 文献笔记阁</span></a></li></ul></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/pic/cover2.jpg')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Juse's Blog</a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fa fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> 文章</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fa fa-archive"></i><span> 归档</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fa fa-tags"></i><span> 标签</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fa fa-folder-open"></i><span> 分类</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/messageboard/"><i class="fa-fw fa fa-paper-plane"></i><span> 留言板</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fa fa-link"></i><span> 友链</span></a></div><div class="menus_item"><a class="site-page" href="/myself/"><i class="fa-fw fa fa-id-card"></i><span> 我自己</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fa fa-sitemap"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-shoe-prints"></i><span> 杂项</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/talk/"><i class="fa-fw fa fa-book"></i><span> 琐碎的日常</span></a></li><li><a class="site-page child" href="/new/"><i class="fa-fw fa fa-bell"></i><span> 更新日志</span></a></li><li><a class="site-page child" href="/paper/"><i class="fa-fw fa fa-sticky-note"></i><span> 文献笔记阁</span></a></li></ul></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">比较转录组分析（七）—— 系统发育分析</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2023-02-17T12:30:00.000Z" title="发表于 2023-02-17 20:30:00">2023-02-17</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2023-10-10T01:54:36.000Z" title="更新于 2023-10-10 09:54:36">2023-10-10</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%AD%A6%E4%B9%A0/">学习</a><i class="fas fa-angle-right post-meta-separator"></i><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%AD%A6%E4%B9%A0/%E7%94%9F%E4%BF%A1/">生信</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">字数总计:</span><span class="word-count">5.2k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">阅读时长:</span><span>19分钟</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="比较转录组分析（七）—— 系统发育分析"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><hr>
<blockquote>
<div id="portalContainer"></div>
<script src="/js/ctlink.js"></script>
<script> generatePortalLinks(7);</script>
</blockquote>
<h2 id="关于分类学和系统发育分析"><a href="#关于分类学和系统发育分析" class="headerlink" title="关于分类学和系统发育分析"></a>关于分类学和系统发育分析</h2><p>分类学，这个学科虽然听起来很简单，但从整个生物史的尺度出发都能算的上是一个非常具有深度且一直以来都争议颇多的学问。从几个世纪前，分类学家就已经致力于通过各种形态学差异或者生活史差异来描述不同物种之间的亲缘关系并且把它们分门别类成不同的类群。在那个时候，不同的学者也会根据 “个人之见” 进行颇具主观性的划分（而且某些近缘物种之间的形态学差异确实可以迥然不同），因此许多分类单元并不统一且一直以来都在不停变动，这一些现象在达尔文先生的《物种起源》中表现得淋漓尽致。</p>
<p>即使古早的分类方法并不完全准确，分类学家依然完成了非常瞩目的成就，就拿现在的物种分类情况来说，很多框架依然与以前一致，并且新的分类单元也是基于已有的框架延申出来的。</p>
<p>测序手段的发展算得上是分类学的一个里程碑，从分子层面的观察让我们对于物种的分类关系有了更为准确的把握，这催生了大量分类错误的纠正以及新物种的发现，并且由此诞生出来了分子系统发育学这一学科。目前很多生信软件都能帮助我们探讨不同物种间的系统发育关系。</p>
<p>这篇文将分享如何通过一些软件来进行系统发育分析。</p>
<h3 id="出现的一些词"><a href="#出现的一些词" class="headerlink" title="出现的一些词"></a>出现的一些词</h3><p><strong>直系同源物和旁系同源物</strong>：见 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/4d29d24883d2">生信基因功能分析工具：Orthofinder 使用教程</a>，更多详细信息见文章末尾题外话知识点。</p>
<p><strong>异同源（xenolog）</strong>：指通过基因水平转移，来源于共生或病毒侵染所产生的相似基因。也是一个进化学上比较热门的问题，在微生物、病毒及与这些类群密切相关的生物（如昆虫等）中较为常见。</p>
<h3 id="更新日志"><a href="#更新日志" class="headerlink" title="更新日志"></a>更新日志</h3><p>2023.06.24    解封文章，完善了关于 Phylopypruner 的说明，补充了各种信息。</p>
<p>2023.07.16    删去了部分内容，进行了一些补充。</p>
<p>2023.09.02    优化了脚本，所需要的时间大大降低了。</p>
<h2 id="分析过程"><a href="#分析过程" class="headerlink" title="分析过程"></a>分析过程</h2><h3 id="所需文件和软件"><a href="#所需文件和软件" class="headerlink" title="所需文件和软件"></a>所需文件和软件</h3><ul>
<li>不同物种的 pep 序列。</li>
<li>软件（相关的安装请自主查阅）：<ul>
<li>orthofinder</li>
<li>mafft</li>
<li>trimAl</li>
<li>iqtree (建议使用 v2 以上的版本，模型寻找更快)</li>
<li>phylopypruner</li>
</ul>
</li>
</ul>
<h3 id="使用-Orthofinder-寻找同源序列"><a href="#使用-Orthofinder-寻找同源序列" class="headerlink" title="使用 Orthofinder 寻找同源序列"></a>使用 Orthofinder 寻找同源序列</h3><p>其他可选的软件也有 OrthoMCL 等，不过从 Orthofinder 具有更突出的优势（见后记题外话）, 此外它涵盖了更多且更全面的功能，使用非常便捷，并不需要花过多的时间学习。</p>
<p>安装方法和使用教程见：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/16e0bbb2ba19">「基因组学」使用 OrthoFinder 进行直系同源基因分析</a>。</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">orthofinder -f ./ -t 16</span></span><br></pre></td></tr></tbody></table></figure>

<p>这里讲一些需要注意的点：</p>
<ul>
<li>最好指定合适的线程数（因为 orthofinder 默认使用的线程数是 128，不足 128 则使用最大的线程数）。</li>
<li>也可以输入 DNA 序列，需要使用参数 <code>-d</code>。</li>
<li>无需指定 <code>-M msa</code> ，这些工作可以自己做并且更具可控性。</li>
<li>Orhofinder 在找完同源群后还会为每个同源群进行建树，这会消耗更多的时间，而且它并不会计算自展值，因此如果想省时间在最开始运行命令时指定 <code>-og</code>。</li>
<li>跑前请一定确保所有的 pep 序列都已经打上了物种标识！不然在后续的分析中你将很难区分每条序列来自哪个物种。</li>
</ul>
<p>其他的参数使用默认值即可。理解输出的文件比运行软件更重要，因此这里着重分享 Orthofinder 得到的结果是什么以及有什么用：</p>
<p>有最详尽解释的 orthofinder 官网 <a target="_blank" rel="noopener" href="https://github.com/davidemms/OrthoFinder%E3%80%82">https://github.com/davidemms/OrthoFinder。</a></p>
<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">Comparative_Genomics_Statistics</summary><div class="toggle-content"><p>这个文件夹对于基于转录组分析的结果来说应该是没太大用处的。</p>
<p>它里面包括了一些基因复制事件的信息以及同源群的相关信息（比如说同源群的共享情况、同源群的大小信息等）。</p>
<p>Gene_Duplication_Events 同理，不过里面记录着更详细的基因复制事件信息。</p>
</div></details>

<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">Gene_Trees</summary><div class="toggle-content"><p>里面是关于同源群所对应的基因树。</p>
</div></details>

<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">Orthogroups &amp; Orthologues</summary><div class="toggle-content"><p>更详尽的同源群信息，比如每个同源群中包括了哪些物种的哪些基因。</p>
</div></details>

<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">Orthogroup_Sequences</summary><div class="toggle-content"><p>重要的文件夹，里面包括了所有同源群对应的序列信息，也就是每个 orthogroup 的 fasta 文件。</p>
</div></details>

<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">Single_Copy_Orthologue_Sequences</summary><div class="toggle-content"><p>在同源群中，<strong>每个物种只有一个基因</strong>的 orthogroup 序列，并不是严格意义上的单拷贝基因（而是推定得到的），如果数量足够可用于序列串联和建树。</p>
</div></details>

<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">Species_Tree</summary><div class="toggle-content"><p>为一份 orthofinder 推断的物种树，默认的方法为 STAG，原理是把包含了所有物种的 orthogroup 对应基因树挑出来，并计算两物种间最近的距离，最后根据每个 orthogroup 的距离矩阵计算得到最后的共识物种树，节点的支持率由这 orthogroup 中的真实情况进行统计计算。</p>
</div></details>

<p>另外一些文件涉及到异同源、HOG 等信息，详细可以前往官网查看。</p>
<h3 id="直接使用-Single-Copy-Orthologue-Sequences-中的序列（情况-①）"><a href="#直接使用-Single-Copy-Orthologue-Sequences-中的序列（情况-①）" class="headerlink" title="直接使用 Single_Copy_Orthologue_Sequences 中的序列（情况 ①）"></a>直接使用 Single_Copy_Orthologue_Sequences 中的序列（情况 ①）</h3><p>这种方法适用于 orthofinder 找到了很多 ‘single copy ortholog’ 的情况，不过这种一般只在分析的物种数较少时发生。</p>
<h3 id="对-Orthogroup（OG）-进行挑选分类（情况-②）"><a href="#对-Orthogroup（OG）-进行挑选分类（情况-②）" class="headerlink" title="对 Orthogroup（OG） 进行挑选分类（情况 ②）"></a>对 Orthogroup（OG） 进行挑选分类（情况 ②）</h3><p>进行分类的原因很简单，一些很大的 orthogroup 里肯定包含了很多 paralog，虽然网上对于 orthogroup 的翻译是直系同源群，但显然里面并不都是 ortholog。很多 Phylotranscriptomics 文章中，研究者也会首先挑选出那些较小且包含物种数较多的 orthogroup 进行后续分析。</p>
<p>此外，部分 OG 中包含的物种数量很少，这种 OG 我们也应当剔除分析之外，具体的挑选通过 Python 实现：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -*- coding: utf-8 -*-</span></span><br><span class="line"><span class="comment"># ortho_select</span></span><br><span class="line"><span class="comment"># Verson: 1.0</span></span><br><span class="line"><span class="comment"># This is used to select orthogroup from output of orthofinder.</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> argparse</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> time</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_args</span>():</span><br><span class="line"></span><br><span class="line">	<span class="keyword">global</span> args</span><br><span class="line">	parser = argparse.ArgumentParser(description = <span class="string">"This is used to select orthogroup from output of orthofinder."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-l"</span>, <span class="string">"--least"</span>, <span class="built_in">type</span> = <span class="built_in">int</span>, <span class="built_in">help</span> = <span class="string">"Least number of species concluding."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-b"</span>, <span class="string">"--big"</span>, default = <span class="number">20</span>, <span class="built_in">type</span> = <span class="built_in">int</span>, <span class="built_in">help</span> = <span class="string">"Orthogroup considered as a big ortho when a species have more than xx sequences. default = 20"</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-f"</span>, <span class="string">"--file"</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"The path to the orthofinder og file."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-o"</span>, <span class="string">"--output"</span>, <span class="built_in">type</span> = <span class="built_in">str</span>, <span class="built_in">help</span> = <span class="string">"The output path of dir."</span>)</span><br><span class="line">	args = parser.parse_args()</span><br><span class="line">	<span class="keyword">return</span> args</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">mkdir</span>():</span><br><span class="line"></span><br><span class="line">	<span class="keyword">global</span> outputDir</span><br><span class="line">	outputDir = os.path.abspath(args.output)</span><br><span class="line">	<span class="keyword">try</span>:</span><br><span class="line">		os.mkdir(<span class="string">f'<span class="subst">{outputDir}</span>/orthogroup_small'</span>)</span><br><span class="line">		os.mkdir(<span class="string">f'<span class="subst">{outputDir}</span>/orthogroup_big'</span>)</span><br><span class="line">	<span class="keyword">except</span>:</span><br><span class="line">		<span class="built_in">print</span>(<span class="string">"Please check the outputdir."</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">select_seq</span>(<span class="params">seq</span>):</span><br><span class="line"></span><br><span class="line">	species_seq = {}</span><br><span class="line">	<span class="keyword">with</span> <span class="built_in">open</span>(seq, <span class="string">"r"</span>) <span class="keyword">as</span> s:</span><br><span class="line">		<span class="keyword">for</span> line <span class="keyword">in</span> s:</span><br><span class="line">			<span class="keyword">if</span> line.startswith(<span class="string">"&gt;"</span>):</span><br><span class="line">				species = line.split(<span class="string">"@"</span>)[<span class="number">0</span>][<span class="number">1</span>:]</span><br><span class="line">				<span class="keyword">if</span> species <span class="keyword">in</span> species_seq.keys():</span><br><span class="line">					species_seq[species] += <span class="number">1</span></span><br><span class="line">				<span class="keyword">else</span>:</span><br><span class="line">					species_seq[species] = <span class="number">1</span></span><br><span class="line">	<span class="keyword">if</span> <span class="built_in">len</span>(species_seq.keys()) &lt; args.least:</span><br><span class="line">		<span class="built_in">print</span>(<span class="string">f"<span class="subst">{seq}</span>'s num of species is less than <span class="subst">{args.least}</span> and so abondoned."</span>)</span><br><span class="line">	<span class="keyword">else</span>:</span><br><span class="line">		<span class="keyword">if</span> <span class="built_in">all</span>(i &lt;= args.big <span class="keyword">for</span> i <span class="keyword">in</span> species_seq.values()):</span><br><span class="line">			os.system(<span class="string">f"cp <span class="subst">{seq}</span> <span class="subst">{outputDir}</span>/orthogroup_small/"</span>)</span><br><span class="line">		<span class="keyword">else</span>:</span><br><span class="line">			os.system(<span class="string">f"cp <span class="subst">{seq}</span> <span class="subst">{outputDir}</span>/orthogroup_big/"</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">run_select</span>():</span><br><span class="line"></span><br><span class="line">	inputdir = os.path.abspath(args.file)</span><br><span class="line">	<span class="keyword">for</span> i <span class="keyword">in</span> os.listdir(inputdir):</span><br><span class="line">		select_seq(<span class="string">f"<span class="subst">{inputdir}</span>/<span class="subst">{i}</span>"</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">main</span>():</span><br><span class="line"></span><br><span class="line">	arges = get_args()</span><br><span class="line">	mkdir()</span><br><span class="line">	run_select()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">"__main__"</span>:</span><br><span class="line"></span><br><span class="line">	t0 = time.time()</span><br><span class="line">	main()</span><br><span class="line">	<span class="built_in">print</span>(<span class="string">f'Total time used: <span class="subst">{time.time() - t0}</span>s\nFinished! See output at <span class="subst">{outputDir}</span>'</span>)</span><br></pre></td></tr></tbody></table></figure>

<p>如果物种标识以下划线区分则将脚本中的 <code>@</code> 改为 <code>_</code>，用法例：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">mkdir</span> ortho_seq</span></span><br><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">python ortho_select.py -l 30 -b 20 -f /pathto/OrthoFinder/Results_xxx/Orthogroup_Sequences -o ortho_seq</span></span><br></pre></td></tr></tbody></table></figure>

<p>其中：</p>
<ul>
<li><p><code>-l</code> 指定最少物种数</p>
</li>
<li><p><code>-b</code> 为当有物种序列数大于多少时该 OG 会被视为大 OG</p>
</li>
<li><p><code>-f</code> 为 orthofinder 跑出来的 Orthogroup_Sequences 路径</p>
</li>
<li><p><code>-o</code> 为输出的文件夹</p>
</li>
</ul>
<p>上述用例中，该命令将得到叫做 <code>ortho_seq</code> 的文件夹，其中包含 <code>orthogroup_big</code> 和 <code>orthogroup_small</code> 两个文件夹。</p>
<p>之后的操作将在 <code>orthogroup_small</code> 中进行（如果该文件夹处理后依然没得到足够的 msa 则需要对 <code>orthogroup_big</code> 中的序列再跑一次 orthofinder 以进行补充）。</p>
<h3 id="比对以及修剪"><a href="#比对以及修剪" class="headerlink" title="比对以及修剪"></a>比对以及修剪</h3><p>在对应的文件夹中输入以下命令：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="keyword">for</span> i <span class="keyword">in</span> *.fa;</span></span><br><span class="line">do</span><br><span class="line">id=${i%%.fa}</span><br><span class="line">mafft --thread 4 --auto $i &gt; $id.mafft.fasta</span><br><span class="line">done</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="keyword">for</span> i <span class="keyword">in</span> *mafft.fasta;</span></span><br><span class="line">do</span><br><span class="line">id=${i%%.mafft.fasta}</span><br><span class="line">trimal -in $i -out $id.trimal.fasta -gappyout</span><br><span class="line">done</span><br></pre></td></tr></tbody></table></figure>

<p><strong>mafft：</strong></p>
<p>mafft 的 <code>--thread</code> 对应使用的线程数。</p>
<p>此外 mafft 有多种比对模式以适用于不同的序列情况，<code>--auto</code> 开启后则是让 mafft 根据序列情况自动寻找适合的比对策略。</p>
<p><strong>trimal：</strong></p>
<p>trimal 也有很多种修剪策略，不同的修剪策略会选择不同的修剪阈值，目前的 trimal 主要包括三种策略 <em><code>gappyout</code><em>、</em><code>strict</code></em> 以及 *<code>strictplus</code>*。</p>
<p>第一种<strong>只考虑到比对的 gap 部分</strong>，会修剪掉比对中的 most gappy fraction。</p>
<p>后两种相较于第一种<strong>还考虑到了 gap 部分的相似性分数</strong>，其中 <em>strictplus</em> 比 <em>strict</em> 更加严格。</p>
<p>另外还有一种 <code>automated</code> 的方法，它会考虑到一些比对的信息比如说序列数量、序列间的一致性分数等然后基于此选择使用 <code>gappyout</code> or <code>strictplus</code>。</p>
<h3 id="二次筛选"><a href="#二次筛选" class="headerlink" title="二次筛选"></a>二次筛选</h3><p>在比对及修剪后部分 msa 的序列长度变短（可能修剪掉太多也有可能是 gap 太多），这些 msa 包含的信息量比较少，因此可以把它们筛掉选择其他那些 “合格” 的 msa 进行下一步的分析。</p>
<p>筛选的思路：选出长度高于一定阈值的 msa，其次删掉里面 gap 过多的序列，然后再根据删除后的物种数情况选择是否保留该 msa。</p>
<p>Python 实现：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -*- coding: utf-8 -*-</span></span><br><span class="line"><span class="comment"># trim_filter</span></span><br><span class="line"><span class="comment"># Verson: 1.0</span></span><br><span class="line"><span class="comment"># This is used to filter alignment to build genetree.</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">import</span> argparse</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_args</span>():</span><br><span class="line"></span><br><span class="line">	<span class="keyword">global</span> args</span><br><span class="line">	parser = argparse.ArgumentParser(description = <span class="string">"This is used to filter alignment to build genetree, run in the directory."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-s"</span>, <span class="string">"--seq"</span>, default = <span class="number">80</span>, <span class="built_in">type</span> = <span class="built_in">int</span>, <span class="built_in">help</span> = <span class="string">"Deletes original seqs shorter than this length, default = 80."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-a"</span>, <span class="string">"--align"</span>, default = <span class="number">80</span>, <span class="built_in">type</span> = <span class="built_in">int</span>, <span class="built_in">help</span> = <span class="string">"Minimum length of a trimmed alignment in amino acids, default = 80."</span>)</span><br><span class="line">	parser.add_argument(<span class="string">"-t"</span>, <span class="string">"--tax"</span>, <span class="built_in">type</span> = <span class="built_in">int</span>, <span class="built_in">help</span> = <span class="string">"Specify minimum number of taxa, namely Minimum number of OTUs to keep an OG. Required!"</span>)</span><br><span class="line">	args = parser.parse_args()</span><br><span class="line">	<span class="keyword">return</span> args</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">remove_n</span>(<span class="params">string, char</span>):</span><br><span class="line"></span><br><span class="line">	<span class="keyword">return</span> <span class="built_in">len</span>(re.sub(char, <span class="string">''</span>, string))</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">filter</span>():</span><br><span class="line"></span><br><span class="line">	min_tax_num = args.tax</span><br><span class="line">	min_len_file = args.seq</span><br><span class="line">	min_len_alig = args.align</span><br><span class="line">	<span class="keyword">for</span> trims <span class="keyword">in</span> os.listdir(<span class="string">"./"</span>):</span><br><span class="line">		os.system(<span class="string">f'echo "Processing file <span class="subst">{trims}</span>......" &gt;&gt; trim_filter.log'</span>)</span><br><span class="line">		<span class="comment"># 得到该比对的各个信息</span></span><br><span class="line">		id_seq = {}</span><br><span class="line">		<span class="keyword">with</span> <span class="built_in">open</span>(trims, <span class="string">'r'</span>) <span class="keyword">as</span> f:</span><br><span class="line">			<span class="keyword">for</span> line <span class="keyword">in</span> f:</span><br><span class="line">				<span class="keyword">if</span> line.startswith(<span class="string">"&gt;"</span>):</span><br><span class="line">					seq_name = line.strip()</span><br><span class="line">				<span class="keyword">else</span>:</span><br><span class="line">					<span class="comment"># 制作映射字典</span></span><br><span class="line">					<span class="keyword">try</span>:</span><br><span class="line">						id_seq[seq_name] += line</span><br><span class="line">					<span class="keyword">except</span>:</span><br><span class="line">						id_seq[seq_name] = line</span><br><span class="line"></span><br><span class="line">		os.system(<span class="string">f'rm <span class="subst">{trims}</span>'</span>)</span><br><span class="line">		<span class="comment"># 判断序列原始长度是否达标</span></span><br><span class="line">		<span class="keyword">if</span> remove_n(<span class="built_in">list</span>(id_seq.values())[<span class="number">0</span>], <span class="string">'\n'</span>) &lt; min_len_file:</span><br><span class="line">			os.system(<span class="string">f'echo "<span class="subst">{trims}</span> length is less than <span class="subst">{min_len_file}</span>AA and be removed." &gt;&gt; trim_filter.log'</span>)</span><br><span class="line">			<span class="keyword">continue</span></span><br><span class="line"></span><br><span class="line">		os.system(<span class="string">f'echo "<span class="subst">{trims}</span> length is enough for next filter." &gt;&gt; trim_filter.log'</span>)</span><br><span class="line">		<span class="comment"># 筛选合格序列</span></span><br><span class="line">		newid_seq = {}</span><br><span class="line">		<span class="keyword">for</span> i <span class="keyword">in</span> id_seq.keys():</span><br><span class="line">			<span class="keyword">if</span> remove_n(id_seq[i], <span class="string">'[-\n]'</span>) &gt;= min_len_alig:</span><br><span class="line">				newid_seq[i] = id_seq[i]</span><br><span class="line">			<span class="keyword">else</span>:</span><br><span class="line">				os.system(<span class="string">f'echo "<span class="subst">{trims}</span> <span class="subst">{i[<span class="number">1</span>:]}</span> is shorter than <span class="subst">{min_len_alig}</span>AA and be removed." &gt;&gt; trim_filter.log'</span>)</span><br><span class="line">		<span class="comment"># 判断物种数量是否合格</span></span><br><span class="line">		species_list = []</span><br><span class="line">		<span class="keyword">for</span> i <span class="keyword">in</span> newid_seq.keys():</span><br><span class="line">			species = i.split(<span class="string">'@'</span>)[<span class="number">0</span>][<span class="number">1</span>:]</span><br><span class="line">			<span class="comment"># 统计物种名单</span></span><br><span class="line">			<span class="keyword">if</span> species <span class="keyword">not</span> <span class="keyword">in</span> species_list:</span><br><span class="line">				species_list.append(species)</span><br><span class="line">		<span class="keyword">if</span> <span class="built_in">len</span>(species_list) &lt; min_tax_num:</span><br><span class="line">			os.system(<span class="string">f'echo "<span class="subst">{trims}</span> tax is less than <span class="subst">{min_tax_num}</span> and be removed." &gt;&gt; trim_filter.log'</span>)</span><br><span class="line">			<span class="keyword">continue</span></span><br><span class="line">		<span class="comment"># 重新再建比对</span></span><br><span class="line">		<span class="keyword">with</span> <span class="built_in">open</span>(trims, <span class="string">'w'</span>) <span class="keyword">as</span> f:</span><br><span class="line">			<span class="keyword">for</span> i <span class="keyword">in</span> newid_seq.keys():</span><br><span class="line">				f.write(i + <span class="string">'\n'</span>)</span><br><span class="line">				f.write(newid_seq[i])</span><br><span class="line">		os.system(<span class="string">f'echo "<span class="subst">{trims}</span> has been saved and filtered." &gt;&gt; trim_filter.log'</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">main</span>():</span><br><span class="line"></span><br><span class="line">	get_args()</span><br><span class="line">	<span class="keyword">try</span>:</span><br><span class="line">		<span class="built_in">filter</span>()</span><br><span class="line">	<span class="keyword">except</span>:</span><br><span class="line">		<span class="built_in">print</span>(<span class="string">"Please check the path."</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">"__main__"</span>:</span><br><span class="line"></span><br><span class="line">	<span class="built_in">print</span>(<span class="string">"Running..."</span>)</span><br><span class="line">	main()</span><br><span class="line">	<span class="built_in">print</span>(<span class="string">f'Finished! See running log in trim_filter.log'</span>)</span><br></pre></td></tr></tbody></table></figure>

<p>如果物种标识以下划线区分则将脚本中的 <code>@</code> 改为 <code>_</code>，用法例：</p>
<p>把修剪后的序列复制一份到新的文件夹中：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">mkdir</span> orthogroup_trim</span></span><br><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">cp</span> orthogroup_small/*trimal.fasta orthogroup_trim/</span></span><br><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">cd</span> orthogroup_trim</span></span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">python /pathto/trim_filter.py -t 30</span></span><br></pre></td></tr></tbody></table></figure>

<p>具体的参数含义可见脚本。</p>
<p>在这之后如果一开始使用的是 <code>Single_Copy_Orthologue_Sequences</code> 的序列，那么可以直接跳到序列串联一部分。</p>
<h3 id="构建基因树（情况-②）"><a href="#构建基因树（情况-②）" class="headerlink" title="构建基因树（情况 ②）"></a>构建基因树（情况 ②）</h3><figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="keyword">for</span> i <span class="keyword">in</span> `<span class="built_in">ls</span> *.trimal.fasta`;</span></span><br><span class="line">do</span><br><span class="line">id=${i%%.fasta}</span><br><span class="line">echo "iqtree2 -s $i -pre $id -bb 1000 -nt 4" &gt;&gt; iqtree_command</span><br><span class="line">done</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">cat</span> iqtree_command | parallel --no-notice -j 10</span></span><br></pre></td></tr></tbody></table></figure>

<p>并行命令使用 IQTREE 进行批量的基因树构建，请根据实际情况修改 <code>-nt</code> 及 <code>-j</code> 参数。</p>
<p>也可以先从小的 Orthogroup sequence 开始建树（更快）：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="keyword">for</span> i <span class="keyword">in</span> `<span class="built_in">ls</span> -r *.trimal.fasta`;</span></span><br><span class="line">do</span><br><span class="line">id=${i%%.fasta}</span><br><span class="line">echo "iqtree2 -s $i -pre $id -bb 1000 -nt 3" &gt;&gt; iqtree_command</span><br><span class="line">done</span><br><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">cat</span> iqtree_command | parallel --no-notice -j 20</span></span><br></pre></td></tr></tbody></table></figure>

<h3 id="运行-Phylopypruner（情况-②）"><a href="#运行-Phylopypruner（情况-②）" class="headerlink" title="运行 Phylopypruner（情况 ②）"></a>运行 Phylopypruner（情况 ②）</h3><p>Phylopypruner 是实现多对多到一对一的关键步骤，它的原理是根据基因树所提供的信息进行一系列过滤修剪从而得到最准确的一对一直系同源物（虽然这仍是推定 [putative] 的）。详尽的原理以及各个参数含义可见其 <a target="_blank" rel="noopener" href="https://gitlab.com/fethalen/phylopypruner/-/wikis/Home">gitlab</a>。</p>
<p><img src="/pic/cover.gif" data-original="/pic/proteomes2orthologs.png"></p>
<p>首先需要一个前置脚本 <code>OTU_replace.py</code> 用于替换基因树中的序列 id（iqtree 进行建树时会将 @ 替换成 _，但 Phylopypruner 只识别 @，当然也可以改源码让 Phylopypruner 可以识别 _）。</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -*- coding: utf-8 -*-</span></span><br><span class="line"><span class="comment"># OTU_replace</span></span><br><span class="line"><span class="comment"># Verson: 1.2</span></span><br><span class="line"><span class="comment"># This is used to replace OTU_ to OTU@ in fasta or treefile.</span></span><br><span class="line"><span class="comment"># Usage: python3 OTU_replace.py dir species_list</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">import</span> time</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">help_info</span>():</span><br><span class="line"></span><br><span class="line">	<span class="built_in">print</span>(<span class="string">"This is used to replace OTU_ to OTU@ in fasta or treefile.\n\tUsage: python3 OTU_replace.py dir species_list\n\tThe file can be indentified includes .fa .fasta .fas .tre .treefile"</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">getArgs</span>():</span><br><span class="line"></span><br><span class="line">	<span class="keyword">global</span> inputfiledir</span><br><span class="line">	<span class="keyword">global</span> species_list</span><br><span class="line"></span><br><span class="line">	args = sys.argv[<span class="number">1</span>:]</span><br><span class="line">	<span class="keyword">if</span> <span class="built_in">len</span>(args) == <span class="number">0</span>:</span><br><span class="line">		help_info()</span><br><span class="line">		sys.exit()</span><br><span class="line">	<span class="keyword">try</span>:</span><br><span class="line">		inputfiledir = os.path.abspath(args[<span class="number">0</span>])</span><br><span class="line">		species_list = os.path.abspath(args[<span class="number">1</span>])</span><br><span class="line">	<span class="keyword">except</span>:</span><br><span class="line">		<span class="built_in">print</span>(<span class="string">"Please check the dirpath."</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">spe_list_get</span>():</span><br><span class="line"></span><br><span class="line">	spe_list = []</span><br><span class="line">	<span class="keyword">with</span> <span class="built_in">open</span>(species_list, <span class="string">'r'</span>) <span class="keyword">as</span> s:</span><br><span class="line">		<span class="keyword">for</span> line <span class="keyword">in</span> s:</span><br><span class="line">			spe_list.append(line.strip())</span><br><span class="line">	<span class="keyword">return</span> spe_list</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">run_replace</span>(<span class="params">dir_path, spe_list</span>):</span><br><span class="line"></span><br><span class="line">    pattern = re.<span class="built_in">compile</span>(<span class="string">r'('</span> + <span class="string">'|'</span>.join(spe_list) + <span class="string">')_'</span>)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> filename <span class="keyword">in</span> os.listdir(dir_path):</span><br><span class="line">        <span class="keyword">if</span> filename.endswith((<span class="string">'.fa'</span>, <span class="string">'.fasta'</span>, <span class="string">'.fas'</span>, <span class="string">'.tre'</span>, <span class="string">'.treefile'</span>)):</span><br><span class="line">            <span class="keyword">with</span> <span class="built_in">open</span>(os.path.join(dir_path, filename), <span class="string">'r'</span>) <span class="keyword">as</span> f:</span><br><span class="line">                content = f.read()</span><br><span class="line"></span><br><span class="line">            new_content = pattern.sub(<span class="string">r'\1@'</span>, content)</span><br><span class="line"></span><br><span class="line">            <span class="keyword">with</span> <span class="built_in">open</span>(os.path.join(dir_path, filename), <span class="string">'w'</span>) <span class="keyword">as</span> f:</span><br><span class="line">                f.write(new_content)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">main</span>():</span><br><span class="line"></span><br><span class="line">	getArgs()</span><br><span class="line">	spe_list = spe_list_get()</span><br><span class="line">	run_replace(inputfiledir, spe_list)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">"__main__"</span>:</span><br><span class="line"></span><br><span class="line">	t0 = time.time()</span><br><span class="line">	main()</span><br><span class="line">	<span class="built_in">print</span>(<span class="string">f'Total time used: <span class="subst">{time.time() - t0}</span>s\nFinished!'</span>)</span><br></pre></td></tr></tbody></table></figure>

<p>species_list 文件中应当包含所有物种的对应物种标识：</p>
<p>例如如果 fasta 中的序列名为这样，则 species_list 应为其下的格式，其中一行对应一个物种名。</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">&gt;Human@xxxxxx</span><br><span class="line">&gt;Mouse@xxxxxxx</span><br><span class="line">&gt;Fish@xxxxxxxxx</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Human</span><br><span class="line">Mouse</span><br><span class="line">Fish</span><br></pre></td></tr></tbody></table></figure>

<p>此后使用 Phylopypruner 推定不同物种的一对一同源物。如果运行所用的数据集过大，在运行的过程中极有可能发生递归错误，因此可以考虑在 Phylopypruner 的 <code>__main__.py</code> 中设置更大的递归深度：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">phylopypruner --<span class="built_in">dir</span> ./ --min-len 100 --trim-lb 5 --threads 6 \</span></span><br><span class="line"><span class="language-bash">--min-support 0.75 --prune MI --min-taxa 6 --min-otu-occupancy 0.1 \</span></span><br><span class="line"><span class="language-bash">--min-gene-occupancy 0.1 --trim-freq-paralogs 4 --trim-divergent 1.25 \</span></span><br><span class="line"><span class="language-bash">--jackknife --min-pdist 1e-8</span></span><br></pre></td></tr></tbody></table></figure>

<p>上述各参数设置是直接照搬 Phylopypruner gitlab 官网所给的<strong>过滤标准最严格</strong>的示例参数，但是实际的参数使用应该视自身需求及已有数据集情况而定！</p>
<blockquote>
<p>PhyloPyPruner, when compared to pre-existing orthology inference programs, have many more parameters that you could potentially tweak so finding a good starting point might be difficult. These are some suggested settings to get you started, but don’t forget to adjust the settings according to your own needs and the overall structure of the dataset that you are working with.</p>
</blockquote>
<p>这里附上 Phylopypruner 各参数含义：</p>
<ul>
<li><code>--dir</code> 多序列比对和树文件所在文件夹。</li>
<li><code>--min-len</code> 比对的最小长度，可以根据之前选择的过滤阈值修改。</li>
<li><code>--trim-lb</code> 删除长度超过所有分支标准差 n 倍的分支。</li>
<li><code>--threads</code> 使用的线程数。</li>
<li><code>--min-support</code> 将支持率低于 n 的节点归并成多分枝。</li>
<li><code>--prune</code> 使用 xx 方法进行剪枝。</li>
<li><code>--min-taxa</code> 不输出物种数小于 n 的比对，可以根据之前选择的过滤阈值修改。</li>
<li><code>--min-otu-occupancy</code> 如果某个物种的比对出现率在 n 以下，则删除该物种。</li>
<li><code>--min-gene-occupancy</code> 删除包含 1-n 以上 gap 的序列。</li>
<li><code>--trim-freq-paralogs</code> &amp; <code>--trim-divergent</code> &amp; <code>--min-pdist</code>：控制旁系同源过滤和直系同源质量的参数。</li>
</ul>
<blockquote>
<p><code>--trim-freq-paralogs 4</code> will get rid of taxa with more than 4 times the standard deviation of the total number of paralogs divided by the number of alignments in which each taxon is present in</p>
<p><code>--trim-divergent 1.25</code> will remove all sequences from a taxon on a per-alignment-basis, if the relative distance from the taxons own sequences, when compared to all other sequences, is than 125%</p>
<p><code>--min-pdist 1e-8</code> will remove any sequence pair which belong to two different species and have a distance that is less than 1e-8</p>
</blockquote>
<ul>
<li><code>--jackknife</code> 逐个删除物种，返回 <em><strong>删除某个物种后将恢复多少序列和比对</strong></em> 的信息。</li>
</ul>
<p>如果运行终止并返回 <code>RecursionError</code> 错误，那么你可以在 Phylopypruner 根目录下的 <code>phylopypruner/__main__.py</code> 中对 <code>main()</code> 部分进行修改：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sys.setrecursionlimit(<span class="number">100000</span>) <span class="comment"># 添加该行，如果还不行就将该值提高为更高的值，例如 200000</span></span><br></pre></td></tr></tbody></table></figure>

<p>对于已安装 conda 并使用 pip 安装的情况，该文件将位于 <code>anaconda3/lib/pythonx.x/site-packages/phylopypruner</code> 中。</p>
<p>最后所得到的伪一对一直系同源基因序列就可以在 <code>phylopypruner_output/output_alignments</code> 中查看，其中每个物种最多存在一条序列，需要注意的是，某些 orthogroup 可能会分离得到多个 <code>xxx_pruned_x.fasta</code>：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">OGxxxxxxx.trimal_pruned_1.fasta</span><br><span class="line">OGxxxxxxx.trimal_pruned_2.fasta</span><br></pre></td></tr></tbody></table></figure>

<h3 id="串联序列，Iqtree-建树"><a href="#串联序列，Iqtree-建树" class="headerlink" title="串联序列，Iqtree 建树"></a>串联序列，Iqtree 建树</h3><p><strong>此处以最大似然法建树软件 IQTREE 为例</strong>，当然也可以视自身要求选择其他方法和对应软件。</p>
<p>串联序列这一部分已在之前的文章中提到过，请走：<a href="https://biojuse.com/2023/01/29/%E7%94%A8%E4%BA%8E%20IQtree%20%E7%9A%84%E5%BA%8F%E5%88%97%E4%B8%B2%E8%81%94%E6%96%B9%E6%B3%95/">传送门</a>。</p>
<p>串联后运行下面的命令进行建树：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">iqtree2 -s concatenation_ortho.fasta \</span></span><br><span class="line"><span class="language-bash">-spp IQ_partition.txt \</span></span><br><span class="line"><span class="language-bash">-m MFP+MERGE -bb 1000 -quiet -nt 20 -pre allOG</span></span><br></pre></td></tr></tbody></table></figure>

<p>各个参数的含义：</p>
<p><code>-s</code> 是输入的串联序列文件，<code>-spp</code> 则对应前者的分区信息，<code>-nt</code> 是使用的线程数（可以使用 <code>-nt AUTO</code> 让 iqtree 自己选择合适的线程数）。</p>
<p><code>-m</code> 是 iqtree 的运行模式，选定 <code>MFP+MERGE</code> 时 iqtree 会自动寻找最佳分区方案和替换模型。</p>
<p><code>-bb</code> 是 iqtree 的一种自展值计算方法 <em>ultrafast bootstrap</em> 的重复次数，一般设置 1000 次。</p>
<p>注意，<em>ultrafast bootstrap</em> 虽然相较于传统的计算方法只要求更低的计算资源，但是仅当支持率超过 95 时才能说明这个分支是可信的。</p>
<p><code>-quiet</code> 开启静默模式，此时 iqtree 不会有任何的屏幕输出。</p>
<p>更多相关的信息请见 iqtree 的 <a target="_blank" rel="noopener" href="http://www.iqtree.org/doc/">documentation</a>。</p>
<p>建完树以后会得到一系列文件，以下一些文件比较关键：</p>
<p><code>xxx.treefile</code> 树文件，为多序列联合建树的结果。</p>
<p><code>xxx.best_scheme.nex</code> 和 <code>xxx.best_model.nex</code> 中有具体的分区信息和模型选择。</p>
<p>而这次得到的树文件，我们就可以把它当作一个揭示了物种间系统发育关系的<strong>物种树</strong>。</p>
<h3 id="其他方法建树"><a href="#其他方法建树" class="headerlink" title="其他方法建树"></a>其他方法建树</h3><p>可见其他文章，例：</p>
<p><a href="https://biojuse.com/2023/07/06/%E8%B4%9D%E5%8F%B6%E6%96%AF%E5%BB%BA%E6%A0%91%E4%B9%8B%20Mrbayes%20%E7%AF%87/">贝叶斯建树之 Mrbayes 篇</a></p>
<p><a href="https://biojuse.com/2023/06/23/Astral%20%E5%BB%BA%E6%A0%91%E6%8C%87%E5%8D%97/">Astral 建树指南</a></p>
<h2 id="题外话"><a href="#题外话" class="headerlink" title="题外话"></a>题外话</h2><p>提及到的所有脚本都放在了 <a target="_blank" rel="noopener" href="https://github.com/JuseTiZ/pythonscript">github</a> 上，要搬去自用的话需要根据自身情况进行一定的修改等。</p>
<p>本文提出的仅是系统发育转录组学中一种可行的方法套路，请谨慎参考。<strong>所有相关脚本及代码的<em>具体参数选择</em>都<em>应依据实际需求进行更正</em>。</strong></p>
<p>如果搞不明白脚本怎么运行想省时间，可以试着看一看博客中的 <em><strong>JuseKit</strong></em> 系列，下载软件并根据教程进行操作（当然更建议学一学 Python）。</p>
<h3 id="Orthofinder-相比之-OrthoMCL-优势在哪"><a href="#Orthofinder-相比之-OrthoMCL-优势在哪" class="headerlink" title="Orthofinder 相比之 OrthoMCL 优势在哪"></a>Orthofinder 相比之 OrthoMCL 优势在哪</h3><p>简单来说，OrthoMCL 划分同源群时使用的是 <code>reciprocal BLAST best-hit method（RBH）</code>，这种方法有个坏处在于它会受到序列长度的干扰。</p>
<p>基于此，Orthofinder 使用了序列长度进行矫正，提高了同源群划分的精度，这种矫正后的方法叫做 <code>reciprocal best normalised hit（RBNH）</code>。</p>
<p><img src="/pic/cover.gif" data-original="/pic/ortho.png"></p>
<blockquote>
<p>OrthoFinder: solving fundamental biases in whole genome comparisons dramatically improves orthogroup inference accuracy</p>
<p>from <em>Genome Biology</em></p>
</blockquote>
<p>不过这并不代表绝对准确，某些 Orthogroup 中可能仍带有许多的 Paralog。</p>
<h3 id="准确推出-Ortholog-与-Paralog-的必要性"><a href="#准确推出-Ortholog-与-Paralog-的必要性" class="headerlink" title="准确推出 Ortholog 与 Paralog 的必要性"></a>准确推出 Ortholog 与 Paralog 的必要性</h3><p>这个问题很简单，以下图为例：</p>
<p><img src="/pic/cover.gif" data-original="/pic/ortho2.png"></p>
<p>当我们使用直系同源基因进行推断时，不管使用 α 还是 β 都能得到正确的结论，即这三个物种中鼠和人具有更近的亲缘关系。</p>
<p>但如果选错了基因，比如说挑选了青蛙和人的 α 基因，以及鼠的 β 基因，虽然挑取的都为同源基因，但是直系同源和旁系同源的区别会使得我们推断出青蛙和人的亲缘关系更近这一错误结论。</p>
<p>因此，<strong>直系同源的正确划分是推断系统发育关系的重点</strong>。</p>
<h3 id="蛋白序列建树和核苷酸序列建树"><a href="#蛋白序列建树和核苷酸序列建树" class="headerlink" title="蛋白序列建树和核苷酸序列建树"></a>蛋白序列建树和核苷酸序列建树</h3><p>使用核苷酸序列和蛋白序列建树得到的结果大多时候都是一致的，但也有可能存在不同。</p>
<p>原因是核苷酸实际上是一种更 “精” 的序列，它相较蛋白序列而言包含更多的信息，比如说某些同义替换等。</p>
<p>因此当存在出入时一般需要再进行考证，大多数情况下蛋白序列和核苷酸序列建树结果的不一致是出现在某些支持率较低的 node 上，这时再添加一些 Ortholog 分析可能可以解决问题。</p>
<!-- hexo injector body_end start --> <script data-pjax="">if(document.getElementById('recent-posts') && (location.pathname ==='/'|| '/' ==='all')){
    var parent = document.getElementById('recent-posts');
    var child = '<div class="recent-post-item" style="width:100%;height: auto"><div id="catalog_magnet"><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/生信/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📚 Juseの生信分享 (23)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/绘图/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">👩‍💻 Juseの绘图经验 (4)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/杂项/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🤹 Juseの琐碎教程 (12)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/文献阅读/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📕 Juseの文献笔记 (2)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/软件开发/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🧰 Juseの软件开发 (8)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/博客/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📑 Juseの博客轻改 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><a class="magnet_link_more"  target="_blank" rel="noopener" href="https://jusetiz.github.io/categories" style="flex:1;text-align: center;margin-bottom: 10px;">查看更多...</a></div></div>';
    console.log('已挂载magnet')
    parent.insertAdjacentHTML("afterbegin",child)}
     </script><style>#catalog_magnet{flex-wrap: wrap;display: flex;width:100%;justify-content:space-between;padding: 10px 10px 0 10px;align-content: flex-start;}.magnet_item{flex-basis: calc(33.333333333333336% - 5px);background: #f2f2f2;margin-bottom: 10px;border-radius: 8px;transition: all 0.2s ease-in-out;}.magnet_item:hover{background: #b30070}.magnet_link_more{color:#555}.magnet_link{color:black}.magnet_link:hover{color:white}@media screen and (max-width: 600px) {.magnet_item {flex-basis: 100%;}}.magnet_link_context{display:flex;padding: 10px;font-size:16px;transition: all 0.2s ease-in-out;}.magnet_link_context:hover{padding: 10px 20px;}</style>
    <style></style><script data-pjax="">
  function butterfly_swiper_injector_config(){
    var parent_div_git = document.getElementById('recent-posts');
    var item_html = '<div class="recent-post-item" style="height: auto;width: 100%"><div class="blog-slider swiper-container-fade swiper-container-horizontal" id="swiper_container"><div class="blog-slider__wrp swiper-wrapper" style="transition-duration: 0ms;"><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-09-01</span><a class="blog-slider__title" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt="">JuseKit（八） —— 计算转录组组装指标</a><div class="blog-slider__text">本次更新新增了计算转录组组装指标的功能，经实测比 Trinity 快且结果一致。</div><a class="blog-slider__button" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-08-04</span><a class="blog-slider__title" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt="">JuseKit（七） —— 绘制 GO 富集分析气泡图</a><div class="blog-slider__text">本次更新新增 GO 富集气泡图绘制方法，只需文件一键执行！</div><a class="blog-slider__button" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-07-12</span><a class="blog-slider__title" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt="">DiscoVista 可视化系统发育不一致</a><div class="blog-slider__text">使用合适的工具将系统发育的不一致更加清晰地呈现出来。</div><a class="blog-slider__button" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/05/04/关于 PAML 的一二三事/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-05-04</span><a class="blog-slider__title" href="2023/05/04/关于 PAML 的一二三事/" alt="">关于 PAML 的一二三事</a><div class="blog-slider__text">PAML 进行正选择检测、选择压力分析和物种分歧时间估计的详细介绍，包括诸多需要注意的事项。</div><a class="blog-slider__button" href="2023/05/04/关于 PAML 的一二三事/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-07-04</span><a class="blog-slider__title" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt="">JuseKit（六） —— 绘制火山图</a><div class="blog-slider__text">此次更新新增了绘制火山图的功能，仅需输入对应格式的文件即可一键绘制火山图。</div><a class="blog-slider__button" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-05-14</span><a class="blog-slider__title" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt="">JuseKit（五） —— 用于系统发育分析的序列过滤</a><div class="blog-slider__text">本次更新带来序列过滤功能，可根据序列长度及物种序列数过滤比对。</div><a class="blog-slider__button" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/wxyd1.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-04-06</span><a class="blog-slider__title" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt="">系统发育转录组学是否靠谱？</a><div class="blog-slider__text">向来只闻 Phylogenomics，那么 Phylotranscriptomics 又如何呢？</div><a class="blog-slider__button" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/slwd.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-03-13</span><a class="blog-slider__title" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt="">单基因 dN、dS 和 ω 的滑动窗口绘制方法</a><div class="blog-slider__text">绘制可以观察基因内不同部位选择情况的滑动窗口图，基于 Kaks_calculator 计算。</div><a class="blog-slider__button" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt="">详情   </a></div></div></div><div class="blog-slider__pagination swiper-pagination-clickable swiper-pagination-bullets"></div></div></div>';
    console.log('已挂载butterfly_swiper')
    parent_div_git.insertAdjacentHTML("afterbegin",item_html)
    }
  var elist = 'undefined'.split(',');
  var cpage = location.pathname;
  var epage = 'all';
  var flag = 0;

  for (var i=0;i<elist.length;i++){
    if (cpage.includes(elist[i])){
      flag++;
    }
  }

  if ((epage ==='all')&&(flag == 0)){
    butterfly_swiper_injector_config();
  }
  else if (epage === cpage){
    butterfly_swiper_injector_config();
  }
  </script><script defer="" src="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper.min.js"></script><script defer="" data-pjax="" src="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper_init.js"></script><!-- hexo injector body_end end --></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="https://biojuse.com">Juse</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="https://biojuse.com">https://biojuse.com</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://biojuse.com" target="_blank">Juse's Blog</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/">生物信息学</a><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1/">生信</a><a class="post-meta__tags" href="/tags/%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90/">系统发育分析</a><a class="post-meta__tags" href="/tags/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%AD%A6/">比较转录组学</a></div><div class="post_share"><div class="social-share" data-image="/pic/cover2.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer=""></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2023/03/02/%E5%9F%BA%E4%BA%8E%20Cytoscape%20%E7%9A%84%E5%AF%8C%E9%9B%86%E7%BD%91%E7%BB%9C%E5%9B%BE%E7%BB%98%E5%88%B6/"><img class="prev-cover" src="/pic/cover.gif" data-original="/pic/cyto.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">基于 Cytoscape 的富集网络图绘制</div></div></a></div><div class="next-post pull-right"><a href="/2023/01/29/%E7%94%A8%E4%BA%8E%20IQtree%20%E7%9A%84%E5%BA%8F%E5%88%97%E4%B8%B2%E8%81%94%E6%96%B9%E6%B3%95/"><img class="next-cover" src="/pic/cover.gif" data-original="/pic/cover4.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">用于 IQtree 的序列串联方法</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><div><a href="/2023/07/12/DiscoVista%20%E5%8F%AF%E8%A7%86%E5%8C%96%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E4%B8%8D%E4%B8%80%E8%87%B4/" title="DiscoVista 可视化系统发育不一致"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">DiscoVista 可视化系统发育不一致</div></div></a></div><div><a href="/2023/06/23/Astral%20%E5%BB%BA%E6%A0%91%E6%8C%87%E5%8D%97/" title="Astral 建树指南"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-12-13</div><div class="title">Astral 建树指南</div></div></a></div><div><a href="/2022/12/23/%E5%9F%BA%E4%BA%8E%20GraphPad%20Prism%20%E7%9A%84%E7%81%AB%E5%B1%B1%E5%9B%BE%E7%BB%98%E5%88%B6/" title="基于 GraphPad Prism 的火山图绘制"><img class="cover" src="/pic/cover.gif" data-original="/pic/prism17.png" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">基于 GraphPad Prism 的火山图绘制</div></div></a></div><div><a href="/2022/11/01/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%B8%80%EF%BC%89%E2%80%94%E2%80%94%20%E5%89%8D%E6%8F%90%E5%8F%8A%E6%A6%82%E8%A6%81/" title="比较转录组分析（一）—— 前提及概要"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2024-07-06</div><div class="title">比较转录组分析（一）—— 前提及概要</div></div></a></div><div><a href="/2022/11/21/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%B8%89%EF%BC%89%E2%80%94%E2%80%94%20%E7%BB%84%E8%A3%85%E7%9A%84%E8%B4%A8%E9%87%8F%E6%A3%80%E6%B5%8B%E4%B8%8E%E5%8E%BB%E5%86%97%E4%BD%99/" title="比较转录组分析（三）—— 组装的质量检测与去冗余"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-10</div><div class="title">比较转录组分析（三）—— 组装的质量检测与去冗余</div></div></a></div><div><a href="/2022/11/06/%E6%AF%94%E8%BE%83%E8%BD%AC%E5%BD%95%E7%BB%84%E5%88%86%E6%9E%90%EF%BC%88%E4%BA%8C%EF%BC%89%E2%80%94%E2%80%94%20%E5%8E%9F%E5%A7%8B%E6%95%B0%E6%8D%AE%E7%9A%84%E8%B4%A8%E6%8E%A7%E4%B8%8E%E8%BD%AC%E5%BD%95%E7%BB%84%E7%BB%84%E8%A3%85/" title="比较转录组分析（二）—— 原始数据的质控与转录组组装"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover2.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">比较转录组分析（二）—— 原始数据的质控与转录组组装</div></div></a></div></div></div><hr><div id="post-comment"><div class="comment-head"><div class="comment-headline"><i class="fas fa-comments fa-fw"></i><span> 评论</span></div></div><div class="comment-wrap"><div><div class="vcomment" id="vcomment"></div></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/pic/cover.gif" data-original="/pic/gitav.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"></div><div class="author-info__name">Juse</div><div class="author-info__description">偶尔徒步行进倒也不错</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">52</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">18</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">10</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/JuseTiZ"><i class="fab fa-github"></i><span>Follow Juse</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="/qq.html" target="_blank" title=""><i class="fab fa-qq"></i></a><a class="social-icon" href="https://space.bilibili.com/31269335" target="_blank" title=""><i class="fab fa-bilibili"></i></a><a class="social-icon" href="https://github.com/JuseTiZ" target="_blank" title=""><i class="fab fa-github"></i></a><a class="social-icon" href="https://steamcommunity.com/id/291195857/" target="_blank" title=""><i class="fab fa-steam"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content"><p align="center"><b>人生是旷野，不是轨道</b><br>⭐⭐🗞️ 新闻速递 📰⭐⭐<br>会萤吗？会萤的！💪</p></div><div id="welcome-info"></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%85%B3%E4%BA%8E%E5%88%86%E7%B1%BB%E5%AD%A6%E5%92%8C%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90"><span class="toc-number">1.</span> <span class="toc-text">关于分类学和系统发育分析</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%87%BA%E7%8E%B0%E7%9A%84%E4%B8%80%E4%BA%9B%E8%AF%8D"><span class="toc-number">1.1.</span> <span class="toc-text">出现的一些词</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%9B%B4%E6%96%B0%E6%97%A5%E5%BF%97"><span class="toc-number">1.2.</span> <span class="toc-text">更新日志</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%88%86%E6%9E%90%E8%BF%87%E7%A8%8B"><span class="toc-number">2.</span> <span class="toc-text">分析过程</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%89%80%E9%9C%80%E6%96%87%E4%BB%B6%E5%92%8C%E8%BD%AF%E4%BB%B6"><span class="toc-number">2.1.</span> <span class="toc-text">所需文件和软件</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8-Orthofinder-%E5%AF%BB%E6%89%BE%E5%90%8C%E6%BA%90%E5%BA%8F%E5%88%97"><span class="toc-number">2.2.</span> <span class="toc-text">使用 Orthofinder 寻找同源序列</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%9B%B4%E6%8E%A5%E4%BD%BF%E7%94%A8-Single-Copy-Orthologue-Sequences-%E4%B8%AD%E7%9A%84%E5%BA%8F%E5%88%97%EF%BC%88%E6%83%85%E5%86%B5-%E2%91%A0%EF%BC%89"><span class="toc-number">2.3.</span> <span class="toc-text">直接使用 Single_Copy_Orthologue_Sequences 中的序列（情况 ①）</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%AF%B9-Orthogroup%EF%BC%88OG%EF%BC%89-%E8%BF%9B%E8%A1%8C%E6%8C%91%E9%80%89%E5%88%86%E7%B1%BB%EF%BC%88%E6%83%85%E5%86%B5-%E2%91%A1%EF%BC%89"><span class="toc-number">2.4.</span> <span class="toc-text">对 Orthogroup（OG） 进行挑选分类（情况 ②）</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%AF%94%E5%AF%B9%E4%BB%A5%E5%8F%8A%E4%BF%AE%E5%89%AA"><span class="toc-number">2.5.</span> <span class="toc-text">比对以及修剪</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BA%8C%E6%AC%A1%E7%AD%9B%E9%80%89"><span class="toc-number">2.6.</span> <span class="toc-text">二次筛选</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%9E%84%E5%BB%BA%E5%9F%BA%E5%9B%A0%E6%A0%91%EF%BC%88%E6%83%85%E5%86%B5-%E2%91%A1%EF%BC%89"><span class="toc-number">2.7.</span> <span class="toc-text">构建基因树（情况 ②）</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%BF%90%E8%A1%8C-Phylopypruner%EF%BC%88%E6%83%85%E5%86%B5-%E2%91%A1%EF%BC%89"><span class="toc-number">2.8.</span> <span class="toc-text">运行 Phylopypruner（情况 ②）</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%B8%B2%E8%81%94%E5%BA%8F%E5%88%97%EF%BC%8CIqtree-%E5%BB%BA%E6%A0%91"><span class="toc-number">2.9.</span> <span class="toc-text">串联序列，Iqtree 建树</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%85%B6%E4%BB%96%E6%96%B9%E6%B3%95%E5%BB%BA%E6%A0%91"><span class="toc-number">2.10.</span> <span class="toc-text">其他方法建树</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E9%A2%98%E5%A4%96%E8%AF%9D"><span class="toc-number">3.</span> <span class="toc-text">题外话</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#Orthofinder-%E7%9B%B8%E6%AF%94%E4%B9%8B-OrthoMCL-%E4%BC%98%E5%8A%BF%E5%9C%A8%E5%93%AA"><span class="toc-number">3.1.</span> <span class="toc-text">Orthofinder 相比之 OrthoMCL 优势在哪</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%87%86%E7%A1%AE%E6%8E%A8%E5%87%BA-Ortholog-%E4%B8%8E-Paralog-%E7%9A%84%E5%BF%85%E8%A6%81%E6%80%A7"><span class="toc-number">3.2.</span> <span class="toc-text">准确推出 Ortholog 与 Paralog 的必要性</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%9B%8B%E7%99%BD%E5%BA%8F%E5%88%97%E5%BB%BA%E6%A0%91%E5%92%8C%E6%A0%B8%E8%8B%B7%E9%85%B8%E5%BA%8F%E5%88%97%E5%BB%BA%E6%A0%91"><span class="toc-number">3.3.</span> <span class="toc-text">蛋白序列建树和核苷酸序列建树</span></a></li></ol></li></ol></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div id="ft"><div class="ft-item-1"><div class="t-top"><div class="t-t-l"><p class="ft-t t-l-t">想说的话😉</p><div class="bg-ad"><div>Genes are not destiny. While they play a large role in shaping how we act, how we feel, and how we interact with others, our environment and personal choices have just as much impact on our behavior.</div><div class="btn-xz-box"><a class="btn-xz" target="_blank" rel="noopener" href="https://habitica.com/static/home">游戏人生不游戏 🎯</a></div></div></div><div class="t-t-r"><p class="ft-t t-l-t">仙人指路😝</p><ul class="ft-links"><li><a target="_blank" rel="noopener" href="https://www.fomal.cc/">建站参考</a><a target="_blank" rel="noopener" href="https://emojixd.com/">表情大全</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/talk/">看看生活</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/messageboard/">我要留言</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/myself/">关于博主</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/archives/">文章归档</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/categories/">文章分类</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/tags/">文章标签</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/about/">有的没的</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/new/">更新日志</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/link/">友情链接</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/404.html">404页面</a></li></ul></div></div></div><div class="ft-item-2"><p class="ft-t">友人链接🔗</p><div class="ft-img-group"><div class="img-group-item"><a target="_blank" rel="noopener" href="https://linctanny.github.io/"><img src="/pic/cover.gif" data-original="https://linctanny.github.io/img/p7.jpeg" alt=""></a></div><div class="img-group-item"><a target="_blank" rel="noopener" href="https://robust-vase.github.io/"><img src="/pic/cover.gif" data-original="https://robust-vase.github.io/img/John_King.webp" alt=""></a></div><div class="img-group-item"><a target="_blank" rel="noopener" href="https://blog.stariverfeel.eu.org/"><img src="/pic/cover.gif" data-original="https://avatars.githubusercontent.com/u/26035995?v=4" alt=""></a></div></div></div></div><div class="copyright">©2022 - 2024  @ Juse</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo </a><span class="footer-separator">|</span><span> 主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly </a><span class="footer-separator">|</span><span> 部署 </span><a target="_blank" rel="noopener" href="https://github.com/">GitPage </a><span class="footer-separator">|</span><span> 加速 </span><a target="_blank" rel="noopener" href="https://dash.cloudflare.com/">Cloudflare </a><span class="footer-separator">|</span><span> </span><a href="https://icp.gov.moe/?keyword=20232400" target="_blank">萌ICP备20232400号</a></div><div id="workboard"></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="translateLink" type="button" title="简繁转换">简</button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><a id="to_comment" href="#post-comment" title="直达评论"><i class="fas fa-comments"></i></a><button id="go-up" type="button" title="回到顶部"><i class="fas fa-arrow-up"></i></button></div></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"></div></div><hr><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div class="js-pjax" id="rightMenu"><div class="rightMenu-group rightMenu-small"><a class="rightMenu-item" href="javascript:window.history.back();"><i class="fa fa-arrow-left"></i></a><a class="rightMenu-item" href="javascript:window.history.forward();"><i class="fa fa-arrow-right"></i></a><a class="rightMenu-item" href="javascript:window.location.reload();"><i class="fa fa-refresh"></i></a><a class="rightMenu-item" href="/"><i class="fa fa-home"></i></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-text"><a class="rightMenu-item" href="javascript:rmf.copySelect();"><i class="fa fa-copy"></i><span>复制</span></a><a class="rightMenu-item" href="javascript:window.open(&quot;https://cn.bing.com/search?q=&quot;+window.getSelection().toString());window.location.reload();"><i class="fa fa-search"></i><span>必应搜索</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-too"><a class="rightMenu-item" href="javascript:window.open(window.getSelection().toString());window.location.reload();"><i class="fa fa-link"></i><span>转到链接</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-paste"><a class="rightMenu-item" href="javascript:rmf.paste()"><i class="fa fa-copy"></i><span>粘贴</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-post"><a class="rightMenu-item" href="#post-comment"><i class="fas fa-comment"></i><span>空降评论</span></a><a class="rightMenu-item" href="javascript:rmf.copyWordsLink()"><i class="fa fa-link"></i><span>复制本文地址</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-to"><a class="rightMenu-item" href="javascript:rmf.openWithNewTab()"><i class="fa fa-window-restore"></i><span>新窗口打开</span></a><a class="rightMenu-item" id="menu-too" href="javascript:rmf.open()"><i class="fa fa-link"></i><span>转到链接</span></a><a class="rightMenu-item" href="javascript:rmf.copyLink()"><i class="fa fa-copy"></i><span>复制链接</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-img"><a class="rightMenu-item" href="javascript:rmf.saveAs()"><i class="fa fa-download"></i><span>保存图片</span></a><a class="rightMenu-item" href="javascript:rmf.openWithNewTab()"><i class="fa fa-window-restore"></i><span>在新窗口打开</span></a><a class="rightMenu-item" href="javascript:rmf.copyLink()"><i class="fa fa-copy"></i><span>复制图片链接</span></a></div><div class="rightMenu-group rightMenu-line"><a class="rightMenu-item" href="/about/"><i class="fa fa-sitemap"></i><span>有的没的</span></a><a class="rightMenu-item" href="/new/"><i class="fa fa-bell"></i><span>更新日志</span></a><a class="rightMenu-item" href="javascript:rmf.fullScreen();"><i class="fas fa-expand"></i><span>切换全屏</span></a><a class="rightMenu-item" href="javascript:window.print();"><i class="fa-solid fa-print"></i><span>打印页面</span></a></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="/js/tw_cn.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><script src="/js/search/local-search.js"></script><script>var preloader = {
  endLoading: () => {
    document.body.style.overflow = 'auto';
    document.getElementById('loading-box').classList.add("loaded")
  },
  initLoading: () => {
    document.body.style.overflow = '';
    document.getElementById('loading-box').classList.remove("loaded")

  }
}
window.addEventListener('load',preloader.endLoading())</script><div class="js-pjax"><script>function loadValine () {
  function initValine () {
    const valine = new Valine(Object.assign({
      el: '#vcomment',
      appId: 'ORERCz6nkgV8OYWw8FovgBhQ-gzGzoHsz',
      appKey: 'BNxukaAtCZZ0K6TZldZpkY9e',
      avatar: 'mp',
      placeholder: '🐕温馨提示本站不具有邮件提示功能，Juse每隔一段时间会在后台检查评论并回复',
      serverURLs: '',
      emojiMaps: "",
      path: window.location.pathname,
      visitor: false
    }, null))
  }

  if (typeof Valine === 'function') initValine() 
  else getScript('https://cdn.jsdelivr.net/npm/valine/dist/Valine.min.js').then(initValine)
}

if ('Valine' === 'Valine' || !true) {
  if (true) btf.loadComment(document.getElementById('vcomment'),loadValine)
  else setTimeout(loadValine, 0)
} else {
  function loadOtherComment () {
    loadValine()
  }
}</script></div><script src="https://cdn.staticfile.org/jquery/3.6.3/jquery.min.js"></script><script type="text/javascript" src="/js/rightmenu.js"></script><script type="text/javascript" src="/js/txmap.js"></script><script type="text/javascript" src="/js/copy12.js"></script><script type="text/javascript" src="/js/ldex.js"></script><script type="text/javascript" src="/js/runtime.js"></script><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/dist/activate-power-mode.min.js"></script><script>POWERMODE.colorful = true;
POWERMODE.shake = false;
POWERMODE.mobile = true;
document.body.addEventListener('input', POWERMODE);
</script><script async="" data-pjax="" src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div>
        <style>
            [bg-lazy] {
                background-image: none !important;
                background-color: #eee !important;
            }
        </style>
        <script>
            window.imageLazyLoadSetting = {
                isSPA: false,
                preloadRatio: 1,
                processImages: null,
            };
        </script><script>window.addEventListener("load",function(){var t=/\.(gif|jpg|jpeg|tiff|png)$/i,r=/^data:image\/[a-z]+;base64,/;Array.prototype.slice.call(document.querySelectorAll("img[data-original]")).forEach(function(a){var e=a.parentNode;"A"===e.tagName&&(e.href.match(t)||e.href.match(r))&&(e.href=a.dataset.original)})});</script><script>!function(r){r.imageLazyLoadSetting.processImages=t;var e=r.imageLazyLoadSetting.isSPA,n=r.imageLazyLoadSetting.preloadRatio||1,c=a();function a(){var t=Array.prototype.slice.call(document.querySelectorAll("img[data-original]")),e=Array.prototype.slice.call(document.querySelectorAll("[bg-lazy]"));return t.concat(e)}function t(){e&&(c=a());for(var t,o=0;o<c.length;o++)0<=(t=(t=c[o]).getBoundingClientRect()).bottom&&0<=t.left&&t.top<=(r.innerHeight*n||document.documentElement.clientHeight*n)&&function(){var t,e,n,a,i=c[o];e=function(){c=c.filter(function(t){return i!==t}),r.imageLazyLoadSetting.onImageLoaded&&r.imageLazyLoadSetting.onImageLoaded(i)},(t=i).hasAttribute("bg-lazy")?(t.removeAttribute("bg-lazy"),e&&e()):(n=new Image,a=t.getAttribute("data-original"),n.onload=function(){t.src=a,t.removeAttribute("data-original"),e&&e()},t.src!==a&&(n.src=a))}()}function i(){clearTimeout(t.tId),t.tId=setTimeout(t,500)}t(),document.addEventListener("scroll",i),r.addEventListener("resize",i),r.addEventListener("orientationchange",i)}(this);</script><script type="text/javascript" charset="utf-8" src="/js/lazyload-plugin/lazyload.intersectionObserver.min.js"></script></body></html>