<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"><title>关于 PAML 的一二三事 | Juse's Blog</title><meta name="keywords" content="生物信息学,生信"><meta name="author" content="Juse"><meta name="copyright" content="Juse"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="PAML 进行正选择检测、选择压力分析和物种分歧时间估计的详细介绍，包括诸多需要注意的事项。">
<meta property="og:type" content="article">
<meta property="og:title" content="关于 PAML 的一二三事">
<meta property="og:url" content="https://biojuse.com/2023/05/04/%E5%85%B3%E4%BA%8E%20PAML%20%E7%9A%84%E4%B8%80%E4%BA%8C%E4%B8%89%E4%BA%8B/index.html">
<meta property="og:site_name" content="Juse's Blog">
<meta property="og:description" content="PAML 进行正选择检测、选择压力分析和物种分歧时间估计的详细介绍，包括诸多需要注意的事项。">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://biojuse.com/pic/cover4.jpg">
<meta property="article:published_time" content="2023-05-04T08:00:00.000Z">
<meta property="article:modified_time" content="2024-07-25T07:08:45.388Z">
<meta property="article:author" content="Juse">
<meta property="article:tag" content="生物信息学">
<meta property="article:tag" content="生信">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://biojuse.com/pic/cover4.jpg"><link rel="shortcut icon" href="/pic/webiron.png"><link rel="canonical" href="https://biojuse.com/2023/05/04/%E5%85%B3%E4%BA%8E%20PAML%20%E7%9A%84%E4%B8%80%E4%BA%8C%E4%B8%89%E4%BA%8B/"><link rel="preconnect" href="//cdn.jsdelivr.net"><link rel="preconnect" href="//busuanzi.ibruce.info"><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":false,"languages":{"hits_empty":"找不到您查询的内容：${query}"}},
  translate: {"defaultEncoding":1,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简"},
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: true,
    post: false
  },
  runtime: '',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '关于 PAML 的一二三事',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2024-07-25 15:08:45'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><link rel="stylesheet" href="/css/mycss.css"><link rel="stylesheet" href="/css/bg.css"><link rel="stylesheet" href="/css/hp.css"><script src="https://cdn1.tianli0.top/npm/vue@2.6.14/dist/vue.min.js"></script><script src="https://cdn1.tianli0.top/npm/element-ui@2.15.6/lib/index.js"></script><link rel="stylesheet" href="https://cdn1.tianli0.top/npm/element-ui@2.15.6/packages/theme-chalk/lib/index.css"><!-- hexo injector head_end start --><link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Zfour/Butterfly-double-row-display@1.00/cardlistpost.min.css">
<style>#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags:before {content:"\A";
  white-space: pre;}#recent-posts > .recent-post-item >.recent-post-info > .article-meta-wrap > .tags > .article-meta__separator{display:none}</style>
<link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiperstyle.css" media="print" onload="this.media='all'"><!-- hexo injector head_end end --><meta name="generator" content="Hexo 5.4.2"><link rel="alternate" href="/atom.xml" title="Juse's Blog" type="application/atom+xml">
</head><body><div id="loading-box"><div class="loading-left-bg"></div><div class="loading-right-bg"></div><div class="spinner-box"><div class="configure-border-1"><div class="configure-core"></div></div><div class="configure-border-2"><div class="configure-core"></div></div><div class="loading-word">加载中...</div></div></div><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/pic/cover.gif" data-original="/pic/gitav.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">57</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">18</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">10</div></a></div><hr><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fa fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> 文章</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fa fa-archive"></i><span> 归档</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fa fa-tags"></i><span> 标签</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fa fa-folder-open"></i><span> 分类</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/messageboard/"><i class="fa-fw fa fa-paper-plane"></i><span> 留言板</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fa fa-link"></i><span> 友链</span></a></div><div class="menus_item"><a class="site-page" href="/myself/"><i class="fa-fw fa fa-id-card"></i><span> 我自己</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fa fa-sitemap"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-shoe-prints"></i><span> 杂项</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/talk/"><i class="fa-fw fa fa-book"></i><span> 琐碎的日常</span></a></li><li><a class="site-page child" href="/new/"><i class="fa-fw fa fa-bell"></i><span> 更新日志</span></a></li><li><a class="site-page child" href="/paper/"><i class="fa-fw fa fa-sticky-note"></i><span> 文献笔记阁</span></a></li></ul></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url('/pic/cover4.jpg')"><nav id="nav"><span id="blog_name"><a id="site-name" href="/">Juse's Blog</a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> 搜索</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fa fa-home"></i><span> 首页</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-list"></i><span> 文章</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fa fa-archive"></i><span> 归档</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fa fa-tags"></i><span> 标签</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fa fa-folder-open"></i><span> 分类</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/messageboard/"><i class="fa-fw fa fa-paper-plane"></i><span> 留言板</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fa fa-link"></i><span> 友链</span></a></div><div class="menus_item"><a class="site-page" href="/myself/"><i class="fa-fw fa fa-id-card"></i><span> 我自己</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fa fa-sitemap"></i><span> 关于</span></a></div><div class="menus_item"><a class="site-page group hide" href="javascript:void(0);"><i class="fa-fw fas fa-shoe-prints"></i><span> 杂项</span><i class="fas fa-chevron-down"></i></a><ul class="menus_item_child"><li><a class="site-page child" href="/talk/"><i class="fa-fw fa fa-book"></i><span> 琐碎的日常</span></a></li><li><a class="site-page child" href="/new/"><i class="fa-fw fa fa-bell"></i><span> 更新日志</span></a></li><li><a class="site-page child" href="/paper/"><i class="fa-fw fa fa-sticky-note"></i><span> 文献笔记阁</span></a></li></ul></div></div><div id="toggle-menu"><a class="site-page"><i class="fas fa-bars fa-fw"></i></a></div></div></nav><div id="post-info"><h1 class="post-title">关于 PAML 的一二三事</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" datetime="2023-05-04T08:00:00.000Z" title="发表于 2023-05-04 16:00:00">2023-05-04</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" datetime="2024-07-25T07:08:45.388Z" title="更新于 2024-07-25 15:08:45">2024-07-25</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%AD%A6%E4%B9%A0/">学习</a><i class="fas fa-angle-right post-meta-separator"></i><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E5%AD%A6%E4%B9%A0/%E7%94%9F%E4%BF%A1/">生信</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">字数总计:</span><span class="word-count">9k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">阅读时长:</span><span>34分钟</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="关于 PAML 的一二三事"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">阅读量:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container"><hr>
<h2 id="PAML-简介"><a href="#PAML-简介" class="headerlink" title="PAML 简介"></a>PAML 简介</h2><p>PAML，全称 Phylogenetic Analysis by Maximum Likelihood，由国人大佬杨子恒教授开发，是一款用于进行各种系统发育分析的软件包。里面的诸多模型可以说是系统发育学文章里的常客，其所涵盖的功能包括但不限于：</p>
<ul>
<li>重建系统发育树。</li>
<li>检测选择压力。</li>
<li>估计进化速率。</li>
<li>估计物种或基因的分歧时间。</li>
</ul>
<p>想要最详细地了解这个软件，可直接下载官方 tutorial 进行啃读：</p>
<blockquote>
<p>PAML Manual：<a target="_blank" rel="noopener" href="http://abacus.gene.ucl.ac.uk/software/pamlDOC.pdf">http://abacus.gene.ucl.ac.uk/software/pamlDOC.pdf</a></p>
<p>PAML FAQs：<a target="_blank" rel="noopener" href="http://abacus.gene.ucl.ac.uk/software/pamlFAQs.pdf">http://abacus.gene.ucl.ac.uk/software/pamlFAQs.pdf</a></p>
</blockquote>
<p>也有其他博主写的精细介绍，详见：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/152a6216c794">PAML-discussion-group from 简书</a></p>
<p>如果有些报错问题或者运行问题在网上找不到解释，也许可以看看 <a target="_blank" rel="noopener" href="https://groups.google.com/g/pamlsoftware">PAML 的谷歌论坛</a>。</p>
<p>提示，本文将基于 <em><strong>linux</strong></em> 系统的操作进行讲解。</p>
<h3 id="更新日志"><a href="#更新日志" class="headerlink" title="更新日志"></a>更新日志</h3><p>2023.05.20    补充了支位点模型的局限性。</p>
<p>2023.06.21    补充了自由比模型的相关内容。</p>
<p>2023.06.24    补充了自由比模型的批量运行方法。</p>
<p>2023.07.03    补充了 <code>mcmctree</code> 的运行介绍，精简部分内容。</p>
<p>2023.07.19    补充了 <code>mcmctree</code> 的相关内容，修改了后记内容。</p>
<p>2023.08.19    补充了正选择检测（支位点模型）的内容。</p>
<h2 id="正选择检测（支位点模型）"><a href="#正选择检测（支位点模型）" class="headerlink" title="正选择检测（支位点模型）"></a>正选择检测（支位点模型）</h2><p>如果想要系统地了解正选择检测的各种方法及其局限性等相关内容，可以参见官方出版论文教程：<a target="_blank" rel="noopener" href="https://academic.oup.com/mbe/article/40/4/msad041/7140562?searchresult=1&amp;login=false#403497381">初学者的 PAML 正选择检测指南</a>。</p>
<p>该部分侧重于检测<strong>特定分支</strong>在<strong>特定位点</strong>上的正选择（因此被称作支位点模型 <strong>Branch-Site Model</strong>）。其他的还有：</p>
<ul>
<li>支模型（Branch model），检测作用于<strong>特定谱系</strong>的正选择，<code>m0 vs branch model</code>。</li>
<li>位点模型（Site model），检测作用于基因中<strong>位点</strong>的正选择，可选 <code>M0 vs M1a</code> <code>M1a vs M2a</code> 及 <code>M7 vs M8</code>。</li>
</ul>
<h3 id="原理及运行"><a href="#原理及运行" class="headerlink" title="原理及运行"></a>原理及运行</h3><p>这里首先详细讲一讲它的原理，<strong>PAML tutorial</strong> 中说法：</p>
<blockquote>
<p>The <em>branch-site</em> model A (see the section Codon substitution models above) is specified by using both variables model and NSsites. </p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Model A: model = 2, NSsites = 2, fix_omega = 0</span><br></pre></td></tr></tbody></table></figure>

<p>This is the alternative model for the branch-site test of positive selection. The null model is also the branch-site model A but with <em>ω</em>2 = 1 fixed, specified by </p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Model A1: model = 2, NSsites = 2, fix_omega = 1, omega = 1</span><br></pre></td></tr></tbody></table></figure>
</blockquote>
<p>简单来说，正选择检测的原理就是拿出两个模型（一个认为有正选择，一个认为没有）的最大似然对数值，通过似然比检验计算 p 值。</p>
<p>具体说法（可以跳过，无伤大雅）：</p>
<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">正选择检测原理</summary><div class="toggle-content"><p>支位点模型的基本原理是允许树的不同分支和序列的不同部位之间 ω 发生变化，该模型将树的分支分为前景分支（树文件中用特殊符号标记）和背景分支。</p>
<p>在前景分支上，某些位点可能发生正选择，而在背景分支上，这些位点可能没有受到正选择的影响。因此，支位点模型允许我们在指定的前景分支上检测正选择信号：</p>
<blockquote>
<p>The branch test (<a href="javascript:;">Yang 1998</a>) assigns two different <em>ω</em> parameters (<em>ω</em>F and <em>ω</em>B) to the foreground and background branches on the tree. The null hypothesis of the test is <em>H</em>0: <em>ω</em>F = 1, whereas the alternative is <em>H</em>1: <em>ω</em>F ≥ 1, with <em>ω</em>B to be a free parameter under both hypotheses.</p>
</blockquote>
<p>在实际应用中，支位点模型通常包含两个子模型：一个所有分支都中性进化（ω = 1）的零模型（A1）和一个认为前景枝存在正选择（ω &gt; 1）的备择模型（A2）。通过计算两个模型的最大似然值并进行似然比检验，评估正选择的统计显著性。</p>
<p>似然比检验的大致原理是：比较两个模型的最大似然值，从而确定哪个模型更能解释观测数据。步骤大致有：</p>
<ol>
<li>分别为零模型和备择模型计算最大似然值（给定模型下观测数据的最大概率）。</li>
<li>计算似然比统计量（两模型最大似然值之差的两倍）。</li>
<li>似然比统计量服从卡方分布，其自由度等于两个模型之间的参数差。根据卡方分布的临界值和实际计算得到的似然比统计量，确定拒绝还是接受零假设。</li>
</ol>
</div></details>

<p>对一个基因进行正选择检测需要运行两次 PAML，并根据输出文件中的最大似然值和自由度进行卡方检验，判断其是否存在显著正选择：</p>
<p>Model A1：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">seqfile = 这里要填你的序列文件路径</span><br><span class="line">treefile = 这里要填你的树文件路径，要检测是否有正选择的分支记得标记</span><br><span class="line">outfile = 这里要填你的输出文件，PAML会将结果输出到此处</span><br><span class="line">noisy = 9   </span><br><span class="line">verbose = 0   </span><br><span class="line">runmode = 0  </span><br><span class="line">seqtype = 1   </span><br><span class="line">CodonFreq = 2   </span><br><span class="line">clock = 0</span><br><span class="line">aaDist = 0</span><br><span class="line">model = 2</span><br><span class="line">NSsites = 2   </span><br><span class="line">icode = 0   </span><br><span class="line">Mgene = 0</span><br><span class="line">fix_kappa = 0   </span><br><span class="line">kappa = 2   </span><br><span class="line">fix_omega = 1   </span><br><span class="line">omega = 1   </span><br><span class="line">fix_alpha = 1   </span><br><span class="line">alpha = .0  </span><br><span class="line">Malpha = 0   </span><br><span class="line">ncatG = 3   </span><br><span class="line">getSE = 0   </span><br><span class="line">RateAncestor = 0 </span><br><span class="line">fix_blength = 0  </span><br><span class="line">method = 0   </span><br><span class="line">Small_Diff = .45e-6</span><br><span class="line">cleandata = 1</span><br></pre></td></tr></tbody></table></figure>

<p>这些参数的含义（可以跳过，无伤大雅）：</p>
<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">各参数具体含义</summary><div class="toggle-content"><p><code>noisy</code> ：信息打印级别。</p>
<p><code>verbose</code> ：输出信息详细程度。</p>
<p><code>runmode</code> ：设置为 0，PAML 将使用你提供的进化树。</p>
<p><code>seqtype</code> ：设置为 1，指定核苷酸序列。</p>
<p><code>CodonFreq</code> ：密码子频率模型，设置为 2 时可以捕获第三个密码子位置的碱基偏好，而这通常与选择压力较低的密码子部位有关。</p>
<p><code>clock</code> ：分子钟模型，支位点模型不接受其他选择。</p>
<p><code>aaDist </code>： 氨基酸替换模型，该处不使用。</p>
<p><code>model </code> 和 <code>NSsites</code>，设置为 2 代表支位点模型。</p>
<p><code>icode</code> ：遗传密码表，0 是标准密码表，需根据自身情况修改。</p>
<p><code>Mgene</code> ：多基因模型，0 表示使用单一模型。针对所有基因使用相同的模型。</p>
<p><code>fix_kappa</code>：是否固定kappa值，0 表示不固定。这允许程序自动估计转换/颠换比率。</p>
<p><code>kappa</code>：初始 kappa 值。程序会自动调整 kappa 值以找到最佳估计。</p>
<p><code>fix_omega</code>：是否固定 omega 值，1 表示固定。</p>
<p><code>omega</code>：初始 omega 值。</p>
<p><code>fix_alpha</code>：是否固定 alpha 值，1 表示固定。意味着不考虑不同位点之间的替换速率变异。</p>
<p><code>alpha</code>：初始 alpha 值。表示不使用 Gamma 分布模型</p>
<p><code>Malpha</code>：混合 Gamma 分布，0 表示不使用。</p>
<p><code>ncatG</code>：Gamma 分布类别数。没有使用 Gamma 分布模型，所以类别数设置为 1 即可。</p>
<p><code>getSE</code>：是否计算参数的标准误差，0 表示不计算。</p>
<p><code>RateAncestor</code>：是否计算祖先节点的替换速率，0 不计算，如果想顺便重建祖先序列的话也可以调 1，就是慢了一些。<strong>注：调 1 方便后续比较祖先序列确定平行替换</strong>。</p>
<p><code>fix_blength</code>：是否固定分支长度，0 表示不固定。这允许程序自动估计分支长度。</p>
<p><code>method</code>：计算最大似然值的方法，表示使用一阶梯度法。这是 codeml 中常用的方法，适用于大多数情况。</p>
<p><code>Small_Diff</code>：用于收敛判断的阈值。EasyCodeml 同款配置。</p>
<p><code>cleandata</code> ：是否删除缺失数据和停止密码子，1 表示要删除，具体如何选择看研究需求。</p>
</div></details>

<p>Model A2：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">seqfile = 这里要填你的序列文件路径</span><br><span class="line">treefile = 这里要填你的树文件路径，要检测是否有正选择的分支记得标记</span><br><span class="line">outfile = 这里要填你的输出文件，PAML 会将结果输出到此处</span><br><span class="line">noisy = 9   </span><br><span class="line">verbose = 0   </span><br><span class="line">runmode = 0  </span><br><span class="line">seqtype = 1   </span><br><span class="line">CodonFreq = 2   </span><br><span class="line">clock = 0</span><br><span class="line">aaDist = 0</span><br><span class="line">model = 2</span><br><span class="line">NSsites = 2   </span><br><span class="line">icode = 0   </span><br><span class="line">Mgene = 0</span><br><span class="line">fix_kappa = 0   </span><br><span class="line">kappa = 2   </span><br><span class="line">fix_omega = 0 </span><br><span class="line">omega = 1.5   </span><br><span class="line">fix_alpha = 1   </span><br><span class="line">alpha = .0  </span><br><span class="line">Malpha = 0   </span><br><span class="line">ncatG = 3   </span><br><span class="line">getSE = 0   </span><br><span class="line">RateAncestor = 0 </span><br><span class="line">fix_blength = 0  </span><br><span class="line">method = 0   </span><br><span class="line">Small_Diff = .45e-6</span><br><span class="line">cleandata = 1</span><br></pre></td></tr></tbody></table></figure>

<p>相较于 A1，它的变动有：</p>
<ul>
<li><code>fix_omega</code> 变成 0，这允许 ω 自由估计，这时 <code>omega</code> 的值将作为优化过程的起点。</li>
<li><code>omega</code> 变成 1.5，表示在优化过程开始时，ω 的初始值为 1.5。</li>
</ul>
<p>在不同的 paper 中或许会看到不同的 omega 值（常是 1.5 或 2），但这一般无关紧要，因为在后续的优化过程中，该值会变化以最大化似然。这个值只是一个起始点，并不代表最终 ω。选择一个合适的初始值只是有助于加速优化过程并提高收敛的稳定性。但大多数情况下，不同的初始值通常会收敛到相似的结果。</p>
<p>运行（上面两个模型分别操作，控制文件中就是上面两个模型的具体参数）：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">codeml xxx(你的控制文件)</span></span><br></pre></td></tr></tbody></table></figure>

<p>不同模型跑出的结果（outfile）的似然值一般都是这种样子：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">lnL(ntime:  7  np: 11):  -1225.318164</span><br></pre></td></tr></tbody></table></figure>

<p>查找 lnl 就能找到，两个文件里都有，记录下两个文件中的 <code>np</code> 和 <code>lnL</code> （上例中为 11 和 -1225.318164）。计算 D 和自由度：</p>
<ul>
<li>D = 2 | lnL1 - lnL0 |</li>
<li>df = np1 - np0</li>
</ul>
<blockquote>
<p>关于此处有个问题，从数学角度出发，使用似然比检验时更高的似然值意味着模型更能适应或描述数据，因此正选择模型（Model A2）的似然值显著高于中性模型（Model A1）时才应该能够成为数据中存在正选择的证据。</p>
<p>但是不管是网上的各个教程还是 EasyCodeml 中，对于似然值之差都取绝对值进行似然比检验，这导致有些时候正选择模型的似然值就算更低也能得到统计显著性（这种情况虽然少但确实会发生），这又给我带来了两个疑问：</p>
<ul>
<li>Model A2 包含更多的参数，这提供了更多的自由度来捕捉数据中的模式，因此它的似然值理应更高（或至少一样）。</li>
<li>为什么两个模型存在显著差异（不管哪个拟合的更好）就能说明存在显著正选择？</li>
</ul>
<p>目前我并不确定具体结论，现已在 PAML discussion group 上发贴求助，等得到具体回答以后会对文章内容进行修改。</p>
</blockquote>
<p>然后使用卡方检验，这里可以用 python scipy 包中的 <code>scipy.stats.chi2</code> 进行。</p>
<p>以上就是整个正选择检测的流程，接下来放一个脚本，该脚本可以用于批量进行正选择检测，该脚本原作者是一个<a href="">简书</a>博主，这里对他的脚本进行了一定的更改以适应合理的批量处理：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># callCodeml.py</span></span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"><span class="keyword">from</span> scipy.stats <span class="keyword">import</span> chi2</span><br><span class="line"><span class="keyword">import</span> time</span><br><span class="line"><span class="keyword">from</span> typing <span class="keyword">import</span> <span class="type">List</span>, <span class="type">Tuple</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">help_info</span>() -&gt; <span class="literal">None</span>:</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">"This script is used to call codeML.\nUsage: python3 callCodeml.py Dir treeFile cpuNum\n"</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_args</span>() -&gt; <span class="type">Tuple</span>[<span class="built_in">str</span>, <span class="type">List</span>[<span class="built_in">str</span>], <span class="built_in">str</span>, <span class="built_in">int</span>]:</span><br><span class="line">    args = sys.argv[<span class="number">1</span>:]</span><br><span class="line">    <span class="keyword">if</span> <span class="built_in">len</span>(args) &lt; <span class="number">3</span>:</span><br><span class="line">        help_info()</span><br><span class="line">        sys.exit(<span class="number">1</span>)</span><br><span class="line">    </span><br><span class="line">    dir_path = os.path.abspath(args[<span class="number">0</span>])</span><br><span class="line">    <span class="keyword">if</span> <span class="keyword">not</span> os.path.isdir(dir_path):</span><br><span class="line">        help_info()</span><br><span class="line">        sys.exit(<span class="number">1</span>)</span><br><span class="line">    </span><br><span class="line">    tree_file = os.path.abspath(args[<span class="number">1</span>])</span><br><span class="line">    <span class="keyword">try</span>:</span><br><span class="line">        cpu_num = <span class="built_in">int</span>(args[<span class="number">2</span>])</span><br><span class="line">    <span class="keyword">except</span> ValueError:</span><br><span class="line">        help_info()</span><br><span class="line">        sys.exit(<span class="number">1</span>)</span><br><span class="line">    </span><br><span class="line">    seqs = [x <span class="keyword">for</span> x <span class="keyword">in</span> os.listdir(dir_path) <span class="keyword">if</span> <span class="keyword">not</span> x.startswith(<span class="string">'.'</span>)]</span><br><span class="line">    <span class="keyword">return</span> dir_path, seqs, tree_file, cpu_num</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">create_dir</span>(<span class="params">base_dir: <span class="built_in">str</span>, name: <span class="built_in">str</span></span>) -&gt; <span class="literal">None</span>:</span><br><span class="line">    dirs = [</span><br><span class="line">        <span class="string">f'<span class="subst">{base_dir}</span>/'</span>,</span><br><span class="line">        <span class="string">f'<span class="subst">{base_dir}</span>/<span class="subst">{name}</span>'</span>,</span><br><span class="line">        <span class="string">f'<span class="subst">{base_dir}</span>/<span class="subst">{name}</span>/null'</span>,</span><br><span class="line">        <span class="string">f'<span class="subst">{base_dir}</span>/<span class="subst">{name}</span>/alte'</span></span><br><span class="line">    ]</span><br><span class="line">    <span class="keyword">for</span> dir_path <span class="keyword">in</span> dirs:</span><br><span class="line">        os.makedirs(dir_path, exist_ok=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">create_ctl</span>(<span class="params">base_dir: <span class="built_in">str</span>, dir_path: <span class="built_in">str</span>, name: <span class="built_in">str</span>, tree_file: <span class="built_in">str</span></span>) -&gt; <span class="literal">None</span>:</span><br><span class="line">    ctl_null = <span class="string">f'''</span></span><br><span class="line"><span class="string">    seqfile = <span class="subst">{dir_path}</span>/<span class="subst">{name}</span></span></span><br><span class="line"><span class="string">    treefile = <span class="subst">{tree_file}</span></span></span><br><span class="line"><span class="string">    outfile = <span class="subst">{base_dir}</span>/<span class="subst">{name}</span>/null/<span class="subst">{name}</span>_null.res</span></span><br><span class="line"><span class="string">    noisy = 9</span></span><br><span class="line"><span class="string">    verbose = 0</span></span><br><span class="line"><span class="string">    runmode = 0</span></span><br><span class="line"><span class="string">    seqtype = 1</span></span><br><span class="line"><span class="string">    CodonFreq = 2</span></span><br><span class="line"><span class="string">    clock = 0</span></span><br><span class="line"><span class="string">    aaDist = 0</span></span><br><span class="line"><span class="string">    model = 2</span></span><br><span class="line"><span class="string">    NSsites = 2</span></span><br><span class="line"><span class="string">    icode = 0</span></span><br><span class="line"><span class="string">    Mgene = 0</span></span><br><span class="line"><span class="string">    fix_kappa = 0</span></span><br><span class="line"><span class="string">    kappa = 2</span></span><br><span class="line"><span class="string">    fix_omega = 1</span></span><br><span class="line"><span class="string">    omega = 1</span></span><br><span class="line"><span class="string">    fix_alpha = 1</span></span><br><span class="line"><span class="string">    alpha = .0</span></span><br><span class="line"><span class="string">    Malpha = 0</span></span><br><span class="line"><span class="string">    ncatG = 3</span></span><br><span class="line"><span class="string">    getSE = 0</span></span><br><span class="line"><span class="string">    RateAncestor = 0</span></span><br><span class="line"><span class="string">    fix_blength = 0</span></span><br><span class="line"><span class="string">    method = 0</span></span><br><span class="line"><span class="string">    Small_Diff = .45e-6</span></span><br><span class="line"><span class="string">    cleandata = 1</span></span><br><span class="line"><span class="string">    '''</span></span><br><span class="line">    ctl_alte = ctl_null.replace(<span class="string">f'<span class="subst">{base_dir}</span>/<span class="subst">{name}</span>/null/<span class="subst">{name}</span>_null.res'</span>, <span class="string">f'<span class="subst">{base_dir}</span>/<span class="subst">{name}</span>/alte/<span class="subst">{name}</span>_alte.res'</span>)\</span><br><span class="line">                       .replace(<span class="string">'fix_omega = 1'</span>, <span class="string">'fix_omega = 0'</span>)\</span><br><span class="line">                       .replace(<span class="string">'omega = 1'</span>, <span class="string">'omega = 1.5'</span>)</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">f'<span class="subst">{base_dir}</span>/<span class="subst">{name}</span>/null/null_profile.ctl'</span>, <span class="string">'w'</span>) <span class="keyword">as</span> f:</span><br><span class="line">        f.write(ctl_null)</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">f'<span class="subst">{base_dir}</span>/<span class="subst">{name}</span>/alte/alte_profile.ctl'</span>, <span class="string">'w'</span>) <span class="keyword">as</span> f:</span><br><span class="line">        f.write(ctl_alte)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">run_create</span>(<span class="params">base_dir: <span class="built_in">str</span>, dir_path: <span class="built_in">str</span>, seqs: <span class="type">List</span>[<span class="built_in">str</span>], tree_file: <span class="built_in">str</span></span>) -&gt; <span class="literal">None</span>:</span><br><span class="line">    <span class="keyword">for</span> name <span class="keyword">in</span> seqs:</span><br><span class="line">        create_dir(base_dir, name)</span><br><span class="line">        create_ctl(base_dir, dir_path, name, tree_file)</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">f'<span class="subst">{base_dir}</span>/codeml.sh'</span>, <span class="string">'w'</span>) <span class="keyword">as</span> f:</span><br><span class="line">        f.write(<span class="string">'''</span></span><br><span class="line"><span class="string">#!/bin/bash</span></span><br><span class="line"><span class="string">list=$(ls -1)</span></span><br><span class="line"><span class="string">count=$(ls -1 | wc -l)</span></span><br><span class="line"><span class="string">echo "Total number: $count"</span></span><br><span class="line"><span class="string">dir=$(pwd)</span></span><br><span class="line"><span class="string">echo "Start Now..."</span></span><br><span class="line"><span class="string">for i in $list</span></span><br><span class="line"><span class="string">do</span></span><br><span class="line"><span class="string">    {</span></span><br><span class="line"><span class="string">    echo "cd $dir/$i/null; codeml ./null_profile.ctl &gt; log.txt; cd $dir/$i/alte; codeml ./alte_profile.ctl &gt; log.txt" &gt;&gt; paml_command</span></span><br><span class="line"><span class="string">    }</span></span><br><span class="line"><span class="string">done</span></span><br><span class="line"><span class="string">wait</span></span><br><span class="line"><span class="string">'''</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f"<span class="subst">{time.strftime(<span class="string">'%Y-%m-%d %H:%M:%S'</span>)}</span> Files created.\n"</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">run_codeml</span>(<span class="params">base_dir: <span class="built_in">str</span>, num: <span class="built_in">int</span></span>) -&gt; <span class="literal">None</span>:</span><br><span class="line">    os.chdir(base_dir)</span><br><span class="line">    os.system(<span class="string">'bash codeml.sh'</span>)</span><br><span class="line">    os.system(<span class="string">f'cat paml_command | parallel --no-notice -j <span class="subst">{num}</span>'</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f"\n<span class="subst">{time.strftime(<span class="string">'%Y-%m-%d %H:%M:%S'</span>)}</span> Codeml Done...\n"</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">get_res</span>(<span class="params">path: <span class="built_in">str</span></span>) -&gt; <span class="type">Tuple</span>[<span class="built_in">float</span>, <span class="built_in">int</span>, <span class="built_in">float</span>]:</span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(path, <span class="string">"r"</span>) <span class="keyword">as</span> f:</span><br><span class="line">        t = f.read()</span><br><span class="line">        kappa = <span class="built_in">float</span>(re.findall(<span class="string">r"kappa\s*=\s*([\d\.]+)"</span>, t)[<span class="number">0</span>])</span><br><span class="line">        lnL = <span class="built_in">float</span>(re.findall(<span class="string">r'lnL\s*\(.*\)\s*=\s*([-\d\.]+)'</span>, t)[<span class="number">0</span>])</span><br><span class="line">        np = <span class="built_in">int</span>(re.findall(<span class="string">r'lnL\s*\(.*\)\s*=\s*[-\d\.]+\s*\(\s*(\d+)'</span>, t)[<span class="number">0</span>])</span><br><span class="line">        <span class="keyword">return</span> lnL, np, kappa</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">run_stat</span>(<span class="params">base_dir: <span class="built_in">str</span>, name: <span class="built_in">str</span></span>) -&gt; <span class="literal">None</span>:</span><br><span class="line">    <span class="keyword">try</span>:</span><br><span class="line">        path_alte = <span class="string">f'<span class="subst">{base_dir}</span>/<span class="subst">{name}</span>/alte/<span class="subst">{name}</span>_alte.res'</span></span><br><span class="line">        path_null = <span class="string">f'<span class="subst">{base_dir}</span>/<span class="subst">{name}</span>/null/<span class="subst">{name}</span>_null.res'</span></span><br><span class="line">        </span><br><span class="line">        lnL0, np0, _ = get_res(path_null)</span><br><span class="line">        lnL1, np1, kappa = get_res(path_alte)</span><br><span class="line">        </span><br><span class="line">        lnl = <span class="built_in">abs</span>(lnL0 - lnL1) * <span class="number">2</span></span><br><span class="line">        np = <span class="built_in">abs</span>(np0 - np1)</span><br><span class="line">        pvalue = <span class="number">1</span> - chi2.cdf(lnl, np)</span><br><span class="line">        </span><br><span class="line">        <span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">f'<span class="subst">{base_dir}</span>/result.tsv'</span>, <span class="string">'a'</span>) <span class="keyword">as</span> f:</span><br><span class="line">            f.write(<span class="string">f"<span class="subst">{name}</span>\t<span class="subst">{lnL0}</span>\t<span class="subst">{lnL1}</span>\t<span class="subst">{np0}</span>\t<span class="subst">{np1}</span>\t<span class="subst">{kappa}</span>\t<span class="subst">{pvalue}</span>\n"</span>)</span><br><span class="line">    <span class="keyword">except</span> Exception <span class="keyword">as</span> e:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">f"<span class="subst">{time.strftime(<span class="string">'%Y-%m-%d %H:%M:%S'</span>)}</span> <span class="subst">{name}</span> failed to stats: <span class="subst">{e}</span>"</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">main</span>() -&gt; <span class="literal">None</span>:</span><br><span class="line">    start_time = time.strftime(<span class="string">'%Y%m%d-%H%M%S'</span>)</span><br><span class="line">    base_dir = <span class="string">f"<span class="subst">{os.getcwd()}</span>/WorkingDir_<span class="subst">{start_time}</span>"</span></span><br><span class="line">    </span><br><span class="line">    dir_path, seqs, tree_file, cpu_num = get_args()</span><br><span class="line">    run_create(base_dir, dir_path, seqs, tree_file)</span><br><span class="line">    run_codeml(base_dir, cpu_num)</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">for</span> name <span class="keyword">in</span> seqs:</span><br><span class="line">        run_stat(base_dir, name)</span><br><span class="line">    </span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f'Total time used: <span class="subst">{time.time() - start_time:<span class="number">.2</span>f}</span> seconds\n'</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">"__main__"</span>:</span><br><span class="line">    t0 = time.time()</span><br><span class="line">    main()</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">f'Total time used: <span class="subst">{time.time() - t0:<span class="number">.2</span>f}</span> seconds\n'</span>)</span><br></pre></td></tr></tbody></table></figure>

<p>相较于原来的版本，变动有：</p>
<ul>
<li>使用了 <code>parallel</code> 来控制并行命令的数量，因为一开始会对所有的序列文件同时开始处理，但如果用到了很多（比如几千个）序列的话这样做就会把 CPU 挤爆，所以加上一些控制的手段保证能几十个几十个来运行。</li>
<li>把文件名识别去除了（之前貌似有限制以特定字符开头或结尾啥的），过程中可能会有部分报错但不会影响结果。</li>
</ul>
<p>运行方式：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">python callCodeml.py Dir treeFile numofcpu</span></span><br></pre></td></tr></tbody></table></figure>

<ul>
<li><code>Dir</code> 指定存放 PAML 格式序列文件的文件夹。</li>
<li><code>treefile</code> 指定物种树。</li>
<li><code>numofcpu</code> 指定预计使用的线程数。</li>
</ul>
<p>所有结果文件会输出在当前目录下以 <code>WorkingDir</code> 开头的文件夹中，所有结果的统计将出现在该文件夹下的 <code>result.tsv</code> 中。</p>
<p>运行前可能需要注意的事项：</p>
<ul>
<li>输入的树文件要为<strong>无根树</strong>，输入有根树会导致 PAML 估计错误。</li>
<li>输入的序列文件要是 PAML 需要的格式（非 fasta）。</li>
<li>确保前景枝已被标记，具体的标记方法可以看之前推荐的教程。</li>
</ul>
<p>你可能会感兴趣的问题：</p>
<ul>
<li>正选择检测时，应当输入<strong>每个基因的基因树</strong>作为树文件还是输入<strong>物种树</strong>作为树文件这一个问题，事实上在很多论文里都是有分歧的，两种方法都有人做，不过也有文章已经指出过，当基因树和物种树拓扑结构不一致时，将物种树作为输入树将会错误地估计枝长，导致更高的假阳性率。</li>
</ul>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://doi.org/10.1093/sysbio/syw018">Gene tree discordance causes apparent substitution rate variation.</a></p>
<p>from <em>Systematic Biology</em></p>
<p>Discussion group 里也有人提过这个问题，杨子恒教授的说法：</p>
<p>The best tree to use would be one that accurately describes the true evolutionary history of the aligned sequences.  </p>
<p>If the gene and species trees disagree, you’ll need to decide which is a better approximation of the true history.  </p>
<p> – a gene tree may be a poor indicator of true history simply because of limited sample size, or because of unusual patterns of sequence evolution (say, if there was strong convergent evolution) </p>
<p> – a species tree may be a poor indicator of true history if gene duplications occurred or if speciation events were tightly clustered (say, if there was incomplete lineage sorting)</p>
<p>Regardless of which tree you select, reviewers may ask that you conduct some/all of the analyses again on the other tree to ensure the results are robust. </p>
</blockquote>
<ul>
<li>经过卡方检验发现部分基因存在显著正选择，但是并没有正选择位点。这可能是因为整个基因表现出了正选择但是<strong>单独位点的信号</strong>太弱。</li>
</ul>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://groups.google.com/g/pamlsoftware/c/j8VKor3t7ws">Can we say a gene is under positve selcetion if the LRT tests results is significant but BEB detects no positively selective sites?</a></p>
</blockquote>
<ul>
<li>是否应当去除外群进行分析？杨子恒教授认为，如果外群离得不是很远，那么将其纳入正选择检测可能是有益的。</li>
</ul>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://groups.google.com/g/pamlsoftware/c/dNypX5VoOJ8/m/foPjr1wrAwAJ">Question on branch labeling with two species (multiple strains)</a></p>
</blockquote>
<ul>
<li>不同版本的 PAML 跑出来的结果可能会不同，请做好版本控制工作。</li>
<li>是否要删除比对中的 gap ？杨子恒团队建议删除主要是 gap 或难以对齐的区域之后，使用 <code>cleandata = 0</code> 保留数据中的信息。</li>
</ul>
<p>还有很多很好的问题有博主整理出来了，可见 <a target="_blank" rel="noopener" href="https://www.jianshu.com/p/152a6216c794">PAML-discussion-group from 简书</a>。这里我就只放一些我在分析时遇到的几个问题还有对应的解释出来。</p>
<h3 id="局限性"><a href="#局限性" class="headerlink" title="局限性"></a>局限性</h3><p>虽然说支位点模型是最广为应用的正选择检测方法之一，但是它的很多假设也在一定程度上削弱了它的可靠性。</p>
<ul>
<li>支位点模型假设所有分支所有位点的进化是<strong>独立</strong>的，在很多情况下这并不成立（你可能想知道 <a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E9%81%97%E4%BC%A0%E8%BF%9E%E9%94%81">遗传连锁</a> 或 <a target="_blank" rel="noopener" href="https://zh.wikipedia.org/wiki/%E5%85%B1%E5%90%8C%E6%BC%94%E5%8C%96">协同进化</a>）。</li>
<li>支位点模型先验地将所有分支分为了存在正选择的前景枝和负选择或中性选择的背景枝两类，当背景枝的情况不符合这个假设时会导致非常高的假阳性率或假阴性率（<a target="_blank" rel="noopener" href="https://academic.oup.com/mbe/article/28/11/3033/1044451">Sergei et al.</a>）。</li>
<li>支位点模型假设所有的分支在所有的位点中共享相同的 dS，在很多情况下这并不成立。</li>
</ul>
<p>此外，也要考虑到一些可能存在的客观不足，比如说：</p>
<ul>
<li>样本量过小（小样本偏差）。如果分析中涉及到的物种太少，那么物种的特异性就会被过度放大（考虑少数物种时以为存在正选择信号，但是加入更多物种后发现实际上并没有），而且参数估计可能会有很大的不确定性，导致结果偏向某一方向。</li>
<li>某些正选择信号可能是遗传漂变造成的（轻微有害突变被漂变固定），这在有效种群大小较低的物种中更有可能发生。</li>
</ul>
<p>所以，如果想要得到更准确的正选择结果，可以考虑使用 <code>Hyphy</code> 等其他生信软件（或模型）对结果进行补充，同时注意采取严格的假阳性控制手段（例如 BH 矫正等）。</p>
<p><code>Hyphy</code> 的模型有时候会表现得更好，现在越来越多的相关文章都是同时使用 <code>PAML</code> 和 <code>Hyphy</code> 进行分析，然后将两者的结果一起展示。</p>
<h2 id="dN-dS-计算（自由比模型）"><a href="#dN-dS-计算（自由比模型）" class="headerlink" title="dN / dS 计算（自由比模型）"></a>dN / dS 计算（自由比模型）</h2><h3 id="原理及运行-1"><a href="#原理及运行-1" class="headerlink" title="原理及运行"></a>原理及运行</h3><p>这一部分也可以叫做<em><strong>选择压力分析</strong></em>，使用的是 PAML Branch Model 中的 <code>free-ratio model</code>。</p>
<p>与之前讲到的支位点模型不同，自由比模型允许 ω(dN/dS) 在所有分支中变化，它对所有分支的 ω 进行独立统计（包括末端分支和祖先分支）。</p>
<p>PAML 的控制文件中，使用 <code>model = 1</code> 来指定自由比模型。</p>
<p>具体如下：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">seqfile = 这里要填你的序列文件路径</span><br><span class="line">treefile = 这里要填你的树文件路径，要检测是否有正选择的分支记得标记</span><br><span class="line">outfile = 这里要填你的输出文件，PAML 会将结果输出到此处</span><br><span class="line"></span><br><span class="line">noisy = 9</span><br><span class="line">verbose = 1</span><br><span class="line">runmode = 0</span><br><span class="line"></span><br><span class="line">seqtype = 1</span><br><span class="line">CodonFreq = 2</span><br><span class="line">model = 1</span><br><span class="line">NSsites = 0</span><br><span class="line">icode = 0</span><br><span class="line">Mgene = 0</span><br><span class="line"></span><br><span class="line">fix_kappa = 0</span><br><span class="line">kappa = 2</span><br><span class="line"></span><br><span class="line">fix_omega = 0</span><br><span class="line">omega = 1</span><br><span class="line"></span><br><span class="line">fix_alpha = 1</span><br><span class="line">ncatG = 8</span><br><span class="line"></span><br><span class="line">getSE = 0</span><br><span class="line">RateAncestor = 0</span><br><span class="line">Small_Diff = .5e-6</span><br><span class="line">cleandata = 1</span><br></pre></td></tr></tbody></table></figure>

<p>部分参数并没有详细指定，PAML 会选择默认选项（默认情况可以参见 <code>pamlDOC.pdf</code>）</p>
<p>然后直接运行即可：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">codeml xxx(你的控制文件)</span></span><br></pre></td></tr></tbody></table></figure>

<p>得到的结果可在 <code>outfile</code> 中查看，一般而言各个分支的 ω 值都在最后一行（以井号 <code>#</code> 作为标识），就像这样：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">(A #0.1 , B #0.2 , C #0.3)</span><br></pre></td></tr></tbody></table></figure>

<p>和支位点模型相同，这里列出批量运行自由比模型的 Python 脚本：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -*- coding: utf-8 -*-</span></span><br><span class="line"><span class="comment"># free-ratio-calcu</span></span><br><span class="line"><span class="comment"># Author: Juse</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"></span><br><span class="line">args = sys.argv[<span class="number">1</span>:]</span><br><span class="line"><span class="keyword">if</span> <span class="built_in">len</span>(args) == <span class="number">0</span>:</span><br><span class="line">	<span class="built_in">print</span>(<span class="string">"Usage: python free-ratio_calcu.py treefile seqdir outputDir"</span>)</span><br><span class="line">	sys.exit()</span><br><span class="line"><span class="keyword">try</span>:</span><br><span class="line">	treefile = os.path.abspath(args[<span class="number">0</span>])</span><br><span class="line">	seqdir = os.path.abspath(args[<span class="number">1</span>])</span><br><span class="line">	outputDir = os.path.abspath(args[<span class="number">2</span>])</span><br><span class="line"><span class="keyword">except</span>:</span><br><span class="line">	<span class="built_in">print</span>(<span class="string">"Please check your command."</span>)</span><br><span class="line">	sys.exit()</span><br><span class="line"></span><br><span class="line">paml_seq = []</span><br><span class="line"><span class="keyword">for</span> paml <span class="keyword">in</span> os.listdir(seqdir):</span><br><span class="line">	paml_seq.append(<span class="string">f"<span class="subst">{seqdir}</span>/<span class="subst">{paml}</span>"</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> seqid <span class="keyword">in</span> paml_seq:</span><br><span class="line">	outid = <span class="string">f"<span class="subst">{seqid.split(<span class="string">'/'</span>)[-<span class="number">1</span>]}</span>.mlc"</span></span><br><span class="line">	<span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">"codeml.ctl"</span>, <span class="string">"w"</span>) <span class="keyword">as</span> f:</span><br><span class="line">		f.write(<span class="string">f'''</span></span><br><span class="line"><span class="string">seqfile = <span class="subst">{seqid}</span></span></span><br><span class="line"><span class="string">treefile = <span class="subst">{treefile}</span></span></span><br><span class="line"><span class="string">outfile = <span class="subst">{outputDir}</span>/<span class="subst">{outid}</span></span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">noisy = 9</span></span><br><span class="line"><span class="string">verbose = 1</span></span><br><span class="line"><span class="string">runmode = 0</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">seqtype = 1</span></span><br><span class="line"><span class="string">CodonFreq = 2</span></span><br><span class="line"><span class="string">model = 1</span></span><br><span class="line"><span class="string">NSsites = 0</span></span><br><span class="line"><span class="string">icode = 0</span></span><br><span class="line"><span class="string">Mgene = 0</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">fix_kappa = 0</span></span><br><span class="line"><span class="string">kappa = 2</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">fix_omega = 0</span></span><br><span class="line"><span class="string">omega = 1</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">fix_alpha = 1</span></span><br><span class="line"><span class="string">ncatG = 8</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">getSE = 0</span></span><br><span class="line"><span class="string">RateAncestor = 0</span></span><br><span class="line"><span class="string">Small_Diff = .5e-6</span></span><br><span class="line"><span class="string">cleandata = 1</span></span><br><span class="line"><span class="string">			'''</span>)</span><br><span class="line">	os.system(<span class="string">"codeml"</span>)</span><br></pre></td></tr></tbody></table></figure>

<p>具体脚本已经上传到 GitHub 上（<a target="_blank" rel="noopener" href="https://github.com/JuseTiZ/PyScript-for-CT/blob/main/free-ratio-calcu.py">链接</a>），运行：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">python free-ratio_calcu.py treefile seqdir outputDir</span></span><br></pre></td></tr></tbody></table></figure>

<ul>
<li><code>treefile</code> 指定物种树文件。</li>
<li><code>seqdir</code> 指定序列文件夹，应将所有要运行自由比模型的序列放置在其中<strong>并且不包括任何其他文件</strong>，序列应为 PAML 格式。</li>
<li><code>outputDir</code> 指定输出文件夹，所有的结果文件将在该处产生，建议指定一个空文件夹以衔接后续批量提取。</li>
</ul>
<p>在这里我并没有像支位点模型那个脚本一样设置多线程操作，所以如果数据较多的友友可以尝试自行修改脚本以提高效率。此外序列的格式要为 PAML 所需格式而非传统 fasta 格式，需要进行序列格式转换的可以使用 Phylosuite 或者 JuseKit。</p>
<h3 id="批量提取结果"><a href="#批量提取结果" class="headerlink" title="批量提取结果"></a>批量提取结果</h3><p>一般来说我们关注的都是末端分支 ω 值，然后取出进行统计比较，如果想要快捷提取出不同物种的 ω 值可以考虑使用 shell 的 <code>grep</code> 或者 Python 的 <code>re</code>。</p>
<p>在这里给出我的 Python 脚本实例（<a target="_blank" rel="noopener" href="https://github.com/JuseTiZ/PyScript-for-CT/blob/main/free-ratio-omega.py">下载链接</a>）：</p>
<figure class="highlight python"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># free-ratio-omega</span></span><br><span class="line"><span class="comment"># Author: Juse</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">import</span> os</span><br><span class="line"></span><br><span class="line">spefile = sys.argv[<span class="number">1</span>]</span><br><span class="line">mlcfiledir = sys.argv[<span class="number">2</span>]</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">species = []</span><br><span class="line"><span class="keyword">with</span> <span class="built_in">open</span>(spefile, <span class="string">'r'</span>) <span class="keyword">as</span> f:</span><br><span class="line">    <span class="keyword">for</span> line <span class="keyword">in</span> f:</span><br><span class="line">        <span class="keyword">if</span> line.strip() != <span class="string">''</span>:</span><br><span class="line">             species.append(line.strip())</span><br><span class="line">species[:] = <span class="built_in">sorted</span>(species)</span><br><span class="line"></span><br><span class="line"><span class="keyword">with</span> <span class="built_in">open</span>(<span class="string">'free-ratio.result'</span>, <span class="string">'w'</span>) <span class="keyword">as</span> op:</span><br><span class="line"></span><br><span class="line">    head = <span class="string">'\t'</span>.join(species)</span><br><span class="line">    op.write(<span class="string">f'File\t<span class="subst">{head}</span>\n'</span>)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">for</span> mlc <span class="keyword">in</span> os.listdir(mlcfiledir):</span><br><span class="line">        op.write(mlc+<span class="string">'\t'</span>)</span><br><span class="line">        mlcpath = <span class="string">f'<span class="subst">{mlcfiledir}</span>/<span class="subst">{mlc}</span>'</span></span><br><span class="line">        <span class="keyword">with</span> <span class="built_in">open</span>(mlcpath, <span class="string">'r'</span>) <span class="keyword">as</span> m:</span><br><span class="line">            file_content = m.read()</span><br><span class="line">            <span class="keyword">for</span> spe <span class="keyword">in</span> species:</span><br><span class="line">                search = <span class="string">fr'<span class="subst">{spe}</span> #(\d+(?:\.\d+)?)'</span></span><br><span class="line">                omega = re.search(search, file_content).group(<span class="number">1</span>)</span><br><span class="line">                op.write(omega + <span class="string">'\t'</span>)</span><br><span class="line">        op.write(<span class="string">'\n'</span>)</span><br></pre></td></tr></tbody></table></figure>

<p>要求的输入有两个：</p>
<ul>
<li>物种名文件，格式如下（一行一个物种名，填需要统计 ω 值的物种）：</li>
</ul>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">SpeA</span><br><span class="line">SpeB</span><br><span class="line">SpeC</span><br></pre></td></tr></tbody></table></figure>

<ul>
<li>PAML 运行结果文件（也就是上面的 <code>outfile</code>）。</li>
</ul>
<p>脚本运行方式：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">python free-ratio-omega.py spename_file path_of_outfile</span></span><br></pre></td></tr></tbody></table></figure>

<p>把 <code>spename_file</code> <code>path_of_outfile</code> 两项换成对应文件和路径即可。</p>
<p>统计结果会出现在运行命令的路径中（<code>free-ratio.result</code>），结果文件实例：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">File	A	B	C</span><br><span class="line">1	0.00733407	0.0001	0.00574998</span><br><span class="line">2	0.119151	0.0917734	0.0648513</span><br><span class="line">3	0.0653332	0.0288158	0.0296085</span><br></pre></td></tr></tbody></table></figure>

<p>需要注意的点：</p>
<ul>
<li>在一些基因中，可能会存在有些分支几乎没有变化的情况，这时很容易出现极端值（<code>999</code> 或 <code>0.0001</code>），因此在分析之前最好排除这些基因。关于这一点，部分文献中也有讲到一定的过滤方法，例如根据 <code>N*dN</code>  <code>S*dS</code> 的大小进行过滤。</li>
<li>本文章提供的脚本仅支持获取末端分支的 ω 值，如有需要提取某一祖先分支 ω 值，可以结合 Python 的 <code>ete3</code> 包完善脚本进行。</li>
</ul>
<p>然后就可以使用 R 或其他软件根据得到的结果进行绘图：</p>
<p><img src="/pic/cover.gif" data-original="/pic2/paml1.png"></p>
<h2 id="估计物种分歧时间"><a href="#估计物种分歧时间" class="headerlink" title="估计物种分歧时间"></a>估计物种分歧时间</h2><p>这一功能通过 PAML 的 <code>mcmctree</code> 实现，虽然该方法较本文中已经描述的几种方法要更复杂得多（涉及到许多参数的设置），但是由于其拥有专门的 tutorial 教程，因此这里我只挑出其中一些我个人认为比较重要的地方进行展示。</p>
<p>首先，请过目 <a target="_blank" rel="noopener" href="http://abacus.gene.ucl.ac.uk/software/MCMCtree.Tutorials.pdf">mcmctree 的 tutorial</a>，这对于理解程序的运行和各个参数而言还是非常重要的，其中还展示了相关示例和实践，可以跟着做一遍。此外，也可以看看简书博主总结的教程：<a target="_blank" rel="noopener" href="https://www.jianshu.com/p/46b28829b078">mcmctree估算物种分歧时间</a>。</p>
<p>如果不嫌看英文麻烦，也可以看看这篇教程：<a target="_blank" rel="noopener" href="https://fish-evol.org/mcmctreeExampleVert6/text1Eng.html">Example of mcmctree analysis</a></p>
<p>如果想系统地了解这部分的理论背景，可以看这篇综述文章：<a target="_blank" rel="noopener" href="https://www.nature.com/articles/nrg.2015.8">https://www.nature.com/articles/nrg.2015.8</a></p>
<p>这里讲一些比较关键的点：</p>
<p>首先是估计物种分歧时间所需要的条件：一个<strong>有根树（无枝长信息）</strong>以及相关的<strong>化石矫准</strong>。也可以用其他文献中估计的分歧时间作为校准点（这种情况被称为二次校准），不过相对来说没那么严谨；或者说以某个地质历史事件发生时间作为校准点，例如某个岛屿分裂导致地理隔离并随之引起两物种的分化。在该分析中，校准信息越准确，分析结果越可靠。</p>
<p>参数设置的相关注意事项：</p>
<ul>
<li>在松弛钟模型（<code>clock = 2 or 3</code>）下，如果树根部没有校准点，那么就需要通过 <code>RootAge</code> 指定一个分歧时间上限（所有物种的最近共同祖先形成时间）。</li>
</ul>
<blockquote>
<p>Juse 注：松弛钟模型允许不同分支进化速率不同，这与严格钟模型相反。</p>
</blockquote>
<ul>
<li>选择时间单位时应使节点的分歧时间大致在 <code>0.01-10</code> 的范围内（超过 10 应该也没有关系但不要低于 0.01），比方说如果分歧时间在大约 100-1000MY，那么可以将 100MY 视作一个时间单位。此外，一些参数的设置需根据你对时间尺度的选择进行调整。</li>
</ul>
<blockquote>
<p>Juse 注：具体的例子在 mcmctree tutorial 的 11 页可见。</p>
<p>在 <code>clock = 2</code> 时，<strong>只有</strong> <code>rgene_gamma</code> 要随之变动。而 <code>clock = 3</code> 时，<code>sigma2_gamma</code> <strong>也要</strong>变化。</p>
</blockquote>
<ul>
<li>指定先验替换率，具体怎么设置取决于对替换率有多大的把握，如果并无太大把握就将 αμ 设置成 <code>1</code> <code>1.5</code> 或 <code>2</code>，然后调整 βμ 至其合理。</li>
</ul>
<blockquote>
<p>Juse 注：有些生物类群的替换率可能能从文献中得到，例如：<a target="_blank" rel="noopener" href="https://pubmed.ncbi.nlm.nih.gov/26603774/">例子</a>。</p>
<p>虽然 PAML 现在使用的 Dirichlet 先验能一定程度上减小错误替代率带来的影响，但关于替换率的参数设定仍需要谨慎，因为该值会直接影响分歧时间的后验估计。</p>
<p>如果没有相关信息，可以通过指定 <code>clock=1</code> 运行 baseml 来获得大致的替换率估计并用于先验设定（详见之前提供的英文教程或后文补充部分）。</p>
</blockquote>
<ul>
<li>最好进行多次分析，确保结果相似（要使用不同的 <code>seed</code> 或直接将 <code>seed</code> 设置为 <code>-1</code>）。对于分析是否已经收敛，可以用 tutorial 中提到的方法去判断，也可以看一看<a target="_blank" rel="noopener" href="https://link.springer.com/protocol/10.1007/978-1-0716-2691-7_1">这篇文章</a>中阐述的方法（基于 R 实现）。若结果不相似，说明分析可能还未收敛，此时可能需要对 <code>burnin</code> <code>sampfreq</code> <code>nsample</code> 三个参数进行一定的修改。</li>
<li>此处涉及了一个新的参数为 <code>ndata</code>，指定的是序列的分区数量，tutorial 中所涉及到的序列就为分区数据（分三个部分，分别是密码子的第一、第二和第三个碱基），如果没有分区的话指定 <code>ndata = 1</code> 即可。</li>
<li>关于选择软边界还是硬边界，文章指出使用软边界能够更加灵活并且减弱错误化石校准的影响。</li>
<li>关于出生率、死亡率和抽样分数的先验设置，文章发现其对于后验时间估计只有很小的影响，但依然建议尝试改变这些值以观察后验估计的鲁棒性。</li>
</ul>
<blockquote>
<p><a target="_blank" rel="noopener" href="https://academic.oup.com/mbe/article/23/1/212/1193630"><strong>Bayesian Estimation of Species Divergence Times Under a Molecular Clock Using Multiple Fossil Calibrations with Soft Bounds</strong></a></p>
<p>from <em>Molecular Biology and Evolution</em></p>
</blockquote>
<ul>
<li>关于怎么选择模型，你可能会对以下信息感兴趣：</li>
</ul>
<blockquote>
<p>不同 Model 指定的具体模型：</p>
<p>0:JC69，1:K80，2:F81，3:F84，4:HKY85</p>
<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">ChatGPT3.5 的回答</summary><div class="toggle-content"><blockquote>
<p>在这些模型中，JC69（Jukes-Cantor 69）是最简单的模型，它假设所有碱基或氨基酸的变异率相等，并且没有考虑碱基或氨基酸之间的相似性。因此，JC69模型通常只用于较近的亲缘关系分析。</p>
<p>K80（Kimura 2参数）模型是比JC69更复杂的模型，它考虑了不同碱基或氨基酸之间的相似性，并引入了转换和转换/颠换的不同变异率。</p>
<p>F81（Felsenstein 81）模型是另一个比JC69更复杂的模型，它考虑了碱基或氨基酸之间的不同变异率，并假设平稳状态的分布是相等的。</p>
<p>F84（Felsenstein 84）模型是F81模型的改进版本，它引入了不同的碱基或氨基酸转换和转换/颠换的变异率。</p>
<p>而HKY85（Hasegawa-Kishino-Yano 85）模型是这些模型中最复杂的一个。它在F84模型的基础上引入了碱基或氨基酸的酶位点偏好，并考虑了碱基或氨基酸的相对频率。因此，HKY85模型在描述分子演化过程中更准确和灵活，具有更高的泛用性。</p>
</blockquote>
</div></details>

<details class="toggle" style="border: 1px solid  "><summary class="toggle-button" style="background-color:  ;color:  \#49b1f5">ChatGPT4.0 的回答</summary><div class="toggle-content"><blockquote>
<ol>
<li>**JC69 (Jukes-Cantor 1969)**：是最简单的模型之一。它假定四种碱基（A, T, C, G）替换的概率是均等的，即任何两种碱基之间的替换率是相等的。</li>
<li>**K80 (Kimura 2-parameter)**：比JC69稍微复杂一些。它引入了两个参数，一个是转换的替换率（purine之间或pyrimidine之间的替换），另一个是转变的替换率（purine和pyrimidine之间的替换）。</li>
<li>**F81 (Felsenstein 1981)**：在JC69的基础上增加了考虑站点频率不均的特点，即每种碱基出现的概率可以不同。</li>
<li>**HKY85 (Hasegawa, Kishino and Yano 1985)**：这个模型进一步扩展了K80，允许碱基频率不等，并且区分了转换和转变的替换率。</li>
<li>**F84 (Felsenstein 1984)**：类似于HKY85，但在处理转换和转变的替换率上略有不同。</li>
</ol>
<p>综合来看，<strong>HKY85</strong>和<strong>F84</strong>是你所列举的模型中较为复杂的，它们在处理碱基替换时考虑了更多的参数。由于HKY85和F84在处理替换率和频率方面都较为灵活，因此它们通常具有较高的泛用性。在许多情况下，HKY85模型是一种非常受欢迎的选择。</p>
</blockquote>
</div></details>

<p>在文章的具体示例中，分别使用了 <code>JC</code> 模型和 <code>HKY+G</code> 模型进行对比，<code>JC</code> 在部分情况下会显得不可靠。</p>
<p>所以如果不缺计算资源和时间的话， <code>Model = 4</code> 或许会是一个万金油选择（同时设置 <code>alpha = 0.5</code> &amp; <code>ncatG = 5</code> 以使用 <code>HKY85+G5 Model</code>）。</p>
</blockquote>
<h3 id="实战演练"><a href="#实战演练" class="headerlink" title="实战演练"></a>实战演练</h3><p>由于该部分 tutorial 已经提供了相关的示例，因此这里为了我将直接示范其通过近似似然节省运行时间的方法。</p>
<p>首先，为什么要使用近似似然 —— 因为近似似然能大幅节约时间，这是 tutorial 的原文：</p>
<blockquote>
<p>For large alignments, calculation of the likelihood function during the MCMC is computationally expensive, and estimation of divergence times is very slow. Thorne et al.[8] suggested using an approximate method to calculate the likelihood that improves the speed of the MCMC dramatically.</p>
</blockquote>
<p>这是<a target="_blank" rel="noopener" href="https://link.springer.com/protocol/10.1007/978-1-0716-2691-7_1">论文</a>片段：</p>
<blockquote>
<p>Because Bayesian molecular-clock dating relies on expensive MCMC sampling, the time savings obtained by using the approximation can be quite dramatic, up to 1000× depending on the dataset.</p>
</blockquote>
<p>本次示例使用的文件均在 PAML 根目录的 <code>examples/DatingSoftBound</code> 下，且均在 tutorial 中可见。</p>
<p>首先跳转到该文件夹，该文件夹中有一控制文件为 <code>mcmctree.ctl</code>，将其内容改为：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">        seed = -1</span><br><span class="line">     seqfile = mtCDNApri123.txt</span><br><span class="line">    treefile = mtCDNApri.trees</span><br><span class="line">    mcmcfile = mcmc.txt</span><br><span class="line">     outfile = out.txt</span><br><span class="line"></span><br><span class="line">       ndata = 3</span><br><span class="line">     seqtype = 0    * 0: nucleotides; 1:codons; 2:AAs</span><br><span class="line">     usedata = 3    * 0: no data; 1:seq like; 2:normal approximation; 3:out.BV (in.BV)</span><br><span class="line">       clock = 2    * 1: global clock; 2: independent rates; 3: correlated rates</span><br><span class="line">     RootAge = '&lt;1.0'  * safe constraint on root age, used if no fossil for root.</span><br><span class="line"></span><br><span class="line">       model = 4    * 0:JC69, 1:K80, 2:F81, 3:F84, 4:HKY85</span><br><span class="line">       alpha = 0.5    * alpha for gamma rates at sites</span><br><span class="line">       ncatG = 5    * No. categories in discrete gamma</span><br><span class="line"></span><br><span class="line">   cleandata = 0    * remove sites with ambiguity data (1:yes, 0:no)?</span><br><span class="line"></span><br><span class="line">     BDparas = 1 1 0.1  * birth, death, sampling</span><br><span class="line"> kappa_gamma = 6 2      * gamma prior for kappa</span><br><span class="line"> alpha_gamma = 1 1      * gamma prior for alpha</span><br><span class="line"></span><br><span class="line"> rgene_gamma = 2 20 1   * gammaDir prior for rate for genes</span><br><span class="line">sigma2_gamma = 1 10 1   * gammaDir prior for sigma^2     (for clock=2 or 3)</span><br><span class="line"></span><br><span class="line">    finetune = 1: .1 .1 .1 .1 .1 .1 * auto (0 or 1): times, musigma2, rates, mixing, paras, FossilErr</span><br><span class="line"></span><br><span class="line">       print = 1   * 0: no mcmc sample; 1: everything except branch rates 2: everything</span><br><span class="line">      burnin = 2000</span><br><span class="line">    sampfreq = 10</span><br><span class="line">     nsample = 20000</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">mcmctree mcmctree.ctl</span></span><br></pre></td></tr></tbody></table></figure>

<p>可以在相关文件中查看详细的序列和物种树校准信息，运行上述命令后，文件夹中会出现一个名为 <code>out.BV</code> 的文件，将其重命名为 <code>in.BV</code>。</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash"><span class="built_in">mv</span> out.BV in.BV</span></span><br></pre></td></tr></tbody></table></figure>

<p>随后修改控制文件，将 <code>usedata = 3</code> 换为 <code>usedata = 2</code> ，再次运行并得到结果。</p>
<p>比对一下结果，使用近似似然时（用时 10 s）：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">t_n8           0.1959 ( 0.1627,  0.2477) ( 0.1588,  0.2391)  0.0803  (Jnode 12)</span><br><span class="line">t_n9           0.1571 ( 0.1464,  0.1629) ( 0.1479,  0.1639)  0.0160  (Jnode 11)</span><br><span class="line">t_n10          0.0928 ( 0.0814,  0.1074) ( 0.0804,  0.1062)  0.0258  (Jnode 10)</span><br><span class="line">t_n11          0.0640 ( 0.0593,  0.0739) ( 0.0587,  0.0723)  0.0136  (Jnode  9)</span><br><span class="line">t_n12          0.0256 ( 0.0191,  0.0336) ( 0.0188,  0.0331)  0.0143  (Jnode  8)</span><br><span class="line">t_n13          0.0484 ( 0.0378,  0.0607) ( 0.0374,  0.0602)  0.0228  (Jnode  7)</span><br><span class="line">mu1            0.4790 ( 0.3951,  0.5705) ( 0.3933,  0.5675)  0.1742</span><br><span class="line">mu2            0.1643 ( 0.1270,  0.2092) ( 0.1262,  0.2078)  0.0816</span><br><span class="line">mu3            2.9379 ( 2.0577,  3.7933) ( 2.0485,  3.7801)  1.7316</span><br><span class="line">sigma2_1       0.0436 ( 0.0022,  0.1503) ( 0.0001,  0.1206)  0.1205</span><br><span class="line">sigma2_2       0.0941 ( 0.0103,  0.2748) ( 0.0001,  0.2292)  0.2291</span><br><span class="line">sigma2_3       0.2195 ( 0.0713,  0.5266) ( 0.0517,  0.4588)  0.4070</span><br><span class="line">lnL          -16.1493 (-24.6090, -9.3610) (-23.9400, -8.9710) 14.9690</span><br></pre></td></tr></tbody></table></figure>

<p>不使用近似似然时（<code>usedata = 1</code>，用时 24 min 38 s）：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">t_n8           0.1931 ( 0.1637,  0.2304) ( 0.1609,  0.2260)  0.0651  (Jnode 12)</span><br><span class="line">t_n9           0.1572 ( 0.1470,  0.1630) ( 0.1483,  0.1637)  0.0154  (Jnode 11)</span><br><span class="line">t_n10          0.0912 ( 0.0815,  0.1036) ( 0.0809,  0.1025)  0.0217  (Jnode 10)</span><br><span class="line">t_n11          0.0631 ( 0.0592,  0.0713) ( 0.0585,  0.0699)  0.0114  (Jnode  9)</span><br><span class="line">t_n12          0.0243 ( 0.0191,  0.0311) ( 0.0188,  0.0306)  0.0119  (Jnode  8)</span><br><span class="line">t_n13          0.0464 ( 0.0375,  0.0572) ( 0.0370,  0.0566)  0.0195  (Jnode  7)</span><br><span class="line">mu1            0.4824 ( 0.4064,  0.5676) ( 0.4033,  0.5633)  0.1600</span><br><span class="line">mu2            0.1677 ( 0.1323,  0.2090) ( 0.1300,  0.2061)  0.0761</span><br><span class="line">mu3            2.7935 ( 2.2004,  3.3122) ( 2.2439,  3.3458)  1.1019</span><br><span class="line">sigma2_1       0.0340 ( 0.0022,  0.1169) ( 0.0001,  0.0933)  0.0933</span><br><span class="line">sigma2_2       0.0712 ( 0.0049,  0.2187) ( 0.0001,  0.1797)  0.1796</span><br><span class="line">sigma2_3       0.0717 ( 0.0109,  0.2105) ( 0.0015,  0.1746)  0.1731</span><br><span class="line">kappa_1       12.3972 (10.4875, 14.6043) (10.3315, 14.4080)  4.0765</span><br><span class="line">kappa_2        9.3277 ( 7.2435, 11.8446) ( 7.1484, 11.6949)  4.5464</span><br><span class="line">kappa_3       33.8883 (30.1020, 38.0932) (30.0334, 37.9816)  7.9483</span><br><span class="line">alpha_1        0.2192 ( 0.1746,  0.2733) ( 0.1717,  0.2689)  0.0972</span><br><span class="line">alpha_2        0.0436 ( 0.0019,  0.1100) ( 0.0000,  0.0987)  0.0987</span><br><span class="line">alpha_3        3.9802 ( 2.7839,  5.7941) ( 2.6174,  5.5002)  2.8828</span><br><span class="line">lnL        -29997.5335 (-30011.0870, -29985.7710) (-30010.5710, -29985.3530) 25.2180</span><br></pre></td></tr></tbody></table></figure>

<p>对比不使用近似似然和使用近似似然时的结果可以发现，两次分析所得到的各节点分歧时间都是非常相近的，但是不使用近似似然分析所用的时间却花费了 140 倍以上。</p>
<p>此外，对于蛋白和密码子序列数据，若要使用近似似然方法节约时间可能需要进行额外的操作，详情可见 <em><strong>Tutorial 4: Approximate likelihood with protein data</strong></em>。</p>
<h3 id="要补充的一些话"><a href="#要补充的一些话" class="headerlink" title="要补充的一些话"></a>要补充的一些话</h3><p>本文中关于 <code>mcmctree</code> 的部分有很多相关的结果信息并没有详细说明，这主要是因为这一部分与之前讲的两个部分有个根本的不同 —— 它具有一个单独的 tutorial 并且 PAML DOC 中也有专门讲解，这些内容基本上涵盖了所有我未提及到的其他地方。</p>
<p>俗话说得好，好记性不如烂笔头。跟着看跟着学跟着做才是最好的学习方法，做出来了一遍还怕后面做不成吗？</p>
<h4 id="估计替换率的方法"><a href="#估计替换率的方法" class="headerlink" title="估计替换率的方法"></a>估计替换率的方法</h4><p>详尽版：<a target="_blank" rel="noopener" href="http://abacus.gene.ucl.ac.uk/software/MCMCtreeStepByStepManual.pdf">A Step-by-Step Tutorial: Divergence Time Estimation with Approximate Likelihood Calculation Using MCMCTREE in PAML</a></p>
<p>创建一个 <code>baseml.ctl</code> 文件，写入以下内容：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">seqfile = xxx</span><br><span class="line">treefile = xxx</span><br><span class="line">outfile = mlb</span><br><span class="line"></span><br><span class="line">noisy = 3</span><br><span class="line">verbose = 1</span><br><span class="line">runmode = 0</span><br><span class="line"></span><br><span class="line">model = 7</span><br><span class="line">Mgene = 0</span><br><span class="line"></span><br><span class="line">fix_kappa = 0</span><br><span class="line">kappa = 2</span><br><span class="line"></span><br><span class="line">fix_alpha = 0</span><br><span class="line">alpha = 0.5</span><br><span class="line">Malpha = 0</span><br><span class="line">ncatG = 5</span><br><span class="line"></span><br><span class="line">fix_rho = 1</span><br><span class="line">rho = 0.</span><br><span class="line">nparK = 0</span><br><span class="line"></span><br><span class="line">clock = 1</span><br><span class="line">nhomo = 1</span><br><span class="line">getSE = 1</span><br><span class="line">RateAncestor = 0</span><br><span class="line">cleandata = 0</span><br></pre></td></tr></tbody></table></figure>

<p>其中 <code>seq</code> 为序列文件（PAML 或 Phy 格式），<code>treefile</code> 为根处有校准的有根树文件，例子：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">7  1</span><br><span class="line">((((human, (chimpanzee, bonobo)), gorilla), (orangutan, sumatran)), gibbon)'@1.0';</span><br></pre></td></tr></tbody></table></figure>

<p>在准备好相关文件后，运行以下命令：</p>
<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta prompt_">$ </span><span class="language-bash">baseml baseml.ctl</span></span><br></pre></td></tr></tbody></table></figure>

<p>输出的 mlc 文件中存在以下内容：</p>
<figure class="highlight plaintext"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">Substitution rate is per time unit</span><br><span class="line"></span><br><span class="line">    0.090798 +- 0.006418</span><br></pre></td></tr></tbody></table></figure>

<p>根据该值，假设 <em>m</em> = <em>s</em> = 0.09，则：</p>
<ul>
<li><em>a</em> = (0.09/0.09)^2 = 1</li>
<li><em>b</em> = 0.09/0.09^2 = 11.1</li>
</ul>
<p>因此可在运行 mcmctree 时将 rgene_gamma 设置为：<code>rgene_gamma = 1 11.1</code></p>
<hr>
<p>除了已介绍的模型，还有一些模型例如 <code>Two-ratio model</code> 和 <code>Fixed ratio model</code> 等没有讲到，不过这些模型的运行只需改几个参数，包括它们之间的比较（比如说 <code>Two-ratio vs. Fixed</code>）同样是通过似然比检验，因此后面不会再特别补充，以后的变动可能主要是对现有的内容进行完善<del>（不过也说不定）</del>。</p>
<!-- hexo injector body_end start --> <script data-pjax="">if(document.getElementById('recent-posts') && (location.pathname ==='/'|| '/' ==='all')){
    var parent = document.getElementById('recent-posts');
    var child = '<div class="recent-post-item" style="width:100%;height: auto"><div id="catalog_magnet"><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/生信/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📚 Juseの生信分享 (27)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/绘图/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">👩‍💻 Juseの绘图经验 (4)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/杂项/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🤹 Juseの琐碎教程 (13)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/学习/文献阅读/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📕 Juseの文献笔记 (2)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/软件开发/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">🧰 Juseの软件开发 (8)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><div class="magnet_item"><a class="magnet_link" href="https://biojuse.com/categories/博客/"><div class="magnet_link_context" style=""><span style="font-weight:500;flex:1">📑 Juseの博客轻改 (1)</span><span style="padding:0px 4px;border-radius: 8px;"><i class="fas fa-arrow-circle-right"></i></span></div></a></div><a class="magnet_link_more"  target="_blank" rel="noopener" href="https://jusetiz.github.io/categories" style="flex:1;text-align: center;margin-bottom: 10px;">查看更多...</a></div></div>';
    console.log('已挂载magnet')
    parent.insertAdjacentHTML("afterbegin",child)}
     </script><style>#catalog_magnet{flex-wrap: wrap;display: flex;width:100%;justify-content:space-between;padding: 10px 10px 0 10px;align-content: flex-start;}.magnet_item{flex-basis: calc(33.333333333333336% - 5px);background: #f2f2f2;margin-bottom: 10px;border-radius: 8px;transition: all 0.2s ease-in-out;}.magnet_item:hover{background: #b30070}.magnet_link_more{color:#555}.magnet_link{color:black}.magnet_link:hover{color:white}@media screen and (max-width: 600px) {.magnet_item {flex-basis: 100%;}}.magnet_link_context{display:flex;padding: 10px;font-size:16px;transition: all 0.2s ease-in-out;}.magnet_link_context:hover{padding: 10px 20px;}</style>
    <style></style><script data-pjax="">
  function butterfly_swiper_injector_config(){
    var parent_div_git = document.getElementById('recent-posts');
    var item_html = '<div class="recent-post-item" style="height: auto;width: 100%"><div class="blog-slider swiper-container-fade swiper-container-horizontal" id="swiper_container"><div class="blog-slider__wrp swiper-wrapper" style="transition-duration: 0ms;"><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-09-01</span><a class="blog-slider__title" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt="">JuseKit（八） —— 计算转录组组装指标</a><div class="blog-slider__text">本次更新新增了计算转录组组装指标的功能，经实测比 Trinity 快且结果一致。</div><a class="blog-slider__button" href="2023/09/01/JuseKit（八） —— 计算转录组组装指标/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-08-04</span><a class="blog-slider__title" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt="">JuseKit（七） —— 绘制 GO 富集分析气泡图</a><div class="blog-slider__text">本次更新新增 GO 富集气泡图绘制方法，只需文件一键执行！</div><a class="blog-slider__button" href="2023/08/04/JuseKit（七） —— 绘制 GO 富集分析气泡图/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-07-12</span><a class="blog-slider__title" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt="">DiscoVista 可视化系统发育不一致</a><div class="blog-slider__text">使用合适的工具将系统发育的不一致更加清晰地呈现出来。</div><a class="blog-slider__button" href="2023/07/12/DiscoVista 可视化系统发育不一致/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/05/04/关于 PAML 的一二三事/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-05-04</span><a class="blog-slider__title" href="2023/05/04/关于 PAML 的一二三事/" alt="">关于 PAML 的一二三事</a><div class="blog-slider__text">PAML 进行正选择检测、选择压力分析和物种分歧时间估计的详细介绍，包括诸多需要注意的事项。</div><a class="blog-slider__button" href="2023/05/04/关于 PAML 的一二三事/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-07-04</span><a class="blog-slider__title" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt="">JuseKit（六） —— 绘制火山图</a><div class="blog-slider__text">此次更新新增了绘制火山图的功能，仅需输入对应格式的文件即可一键绘制火山图。</div><a class="blog-slider__button" href="2023/07/04/JuseKit（六） —— 绘制火山图/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic2/jusekit.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-05-14</span><a class="blog-slider__title" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt="">JuseKit（五） —— 用于系统发育分析的序列过滤</a><div class="blog-slider__text">本次更新带来序列过滤功能，可根据序列长度及物种序列数过滤比对。</div><a class="blog-slider__button" href="2023/05/14/JuseKit（五） —— 用于系统发育分析的序列过滤/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/wxyd1.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-04-06</span><a class="blog-slider__title" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt="">系统发育转录组学是否靠谱？</a><div class="blog-slider__text">向来只闻 Phylogenomics，那么 Phylotranscriptomics 又如何呢？</div><a class="blog-slider__button" href="2023/04/06/文献分享篇（一）—— 系统发育转录组学是否靠谱？/" alt="">详情   </a></div></div><div class="blog-slider__item swiper-slide" style="width: 750px; opacity: 1; transform: translate3d(0px, 0px, 0px); transition-duration: 0ms;"><a class="blog-slider__img" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt=""><img width="48" height="48" src="/pic/cover.gif" data-original="/pic/slwd.png" alt="" onerror="this.src=https://unpkg.zhimg.com/akilar-candyassets/image/loading.gif; this.onerror = null;"/></a><div class="blog-slider__content"><span class="blog-slider__code">2023-03-13</span><a class="blog-slider__title" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt="">单基因 dN、dS 和 ω 的滑动窗口绘制方法</a><div class="blog-slider__text">绘制可以观察基因内不同部位选择情况的滑动窗口图，基于 Kaks_calculator 计算。</div><a class="blog-slider__button" href="2023/03/13/单基因-dN、dS-和-ω-的滑动窗口绘制方法/" alt="">详情   </a></div></div></div><div class="blog-slider__pagination swiper-pagination-clickable swiper-pagination-bullets"></div></div></div>';
    console.log('已挂载butterfly_swiper')
    parent_div_git.insertAdjacentHTML("afterbegin",item_html)
    }
  var elist = 'undefined'.split(',');
  var cpage = location.pathname;
  var epage = 'all';
  var flag = 0;

  for (var i=0;i<elist.length;i++){
    if (cpage.includes(elist[i])){
      flag++;
    }
  }

  if ((epage ==='all')&&(flag == 0)){
    butterfly_swiper_injector_config();
  }
  else if (epage === cpage){
    butterfly_swiper_injector_config();
  }
  </script><script defer="" src="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper.min.js"></script><script defer="" data-pjax="" src="https://npm.elemecdn.com/hexo-butterfly-swiper/lib/swiper_init.js"></script><!-- hexo injector body_end end --></article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">文章作者: </span><span class="post-copyright-info"><a href="https://biojuse.com">Juse</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">文章链接: </span><span class="post-copyright-info"><a href="https://biojuse.com">https://biojuse.com</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">版权声明: </span><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://biojuse.com" target="_blank">Juse's Blog</a>！</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/">生物信息学</a><a class="post-meta__tags" href="/tags/%E7%94%9F%E4%BF%A1/">生信</a></div><div class="post_share"><div class="social-share" data-image="/pic/cover4.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer=""></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2023/05/14/JuseKit%EF%BC%88%E4%BA%94%EF%BC%89%20%E2%80%94%E2%80%94%20%E7%94%A8%E4%BA%8E%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E5%88%86%E6%9E%90%E7%9A%84%E5%BA%8F%E5%88%97%E8%BF%87%E6%BB%A4/"><img class="prev-cover" src="/pic/cover.gif" data-original="/pic2/jusekit.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">JuseKit（五） —— 用于系统发育分析的序列过滤</div></div></a></div><div class="next-post pull-right"><a href="/2023/04/26/JuseKit%EF%BC%88%E5%9B%9B%EF%BC%89%20%E2%80%94%E2%80%94%20%E5%BA%8F%E5%88%97%E6%A0%BC%E5%BC%8F%E8%BD%AC%E6%8D%A2%E4%BB%A5%E5%8F%8A%20Orthogroup%20%E7%9A%84%20cds%20%E6%8F%90%E5%8F%96/"><img class="next-cover" src="/pic/cover.gif" data-original="/pic2/jusekit.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">JuseKit（四） —— 序列格式转换以及 Orthogroup 的 cds 提取</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>相关推荐</span></div><div class="relatedPosts-list"><div><a href="/2023/07/12/DiscoVista%20%E5%8F%AF%E8%A7%86%E5%8C%96%E7%B3%BB%E7%BB%9F%E5%8F%91%E8%82%B2%E4%B8%8D%E4%B8%80%E8%87%B4/" title="DiscoVista 可视化系统发育不一致"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">DiscoVista 可视化系统发育不一致</div></div></a></div><div><a href="/2023/06/23/Astral%20%E5%BB%BA%E6%A0%91%E6%8C%87%E5%8D%97/" title="Astral 建树指南"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-12-13</div><div class="title">Astral 建树指南</div></div></a></div><div><a href="/2023/07/07/%E4%BD%BF%E7%94%A8%20IQTREE%20%E8%AE%A1%E7%AE%97%E4%B8%80%E8%87%B4%E6%80%A7%E5%9B%A0%E5%AD%90/" title="使用 IQTREE 计算一致性因子"><img class="cover" src="/pic/cover.gif" data-original="/pic/cover4.jpg" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">使用 IQTREE 计算一致性因子</div></div></a></div><div><a href="/2023/03/13/%E5%8D%95%E5%9F%BA%E5%9B%A0-dN%E3%80%81dS-%E5%92%8C-%CF%89-%E7%9A%84%E6%BB%91%E5%8A%A8%E7%AA%97%E5%8F%A3%E7%BB%98%E5%88%B6%E6%96%B9%E6%B3%95/" title="单基因 dN、dS 和 ω 的滑动窗口绘制方法"><img class="cover" src="/pic/cover.gif" data-original="/pic/slwd.png" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-10</div><div class="title">单基因 dN、dS 和 ω 的滑动窗口绘制方法</div></div></a></div><div><a href="/2023/03/02/%E5%9F%BA%E4%BA%8E%20Cytoscape%20%E7%9A%84%E5%AF%8C%E9%9B%86%E7%BD%91%E7%BB%9C%E5%9B%BE%E7%BB%98%E5%88%B6/" title="基于 Cytoscape 的富集网络图绘制"><img class="cover" src="/pic/cover.gif" data-original="/pic/cyto.png" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">基于 Cytoscape 的富集网络图绘制</div></div></a></div><div><a href="/2022/12/23/%E5%9F%BA%E4%BA%8E%20GraphPad%20Prism%20%E7%9A%84%E7%81%AB%E5%B1%B1%E5%9B%BE%E7%BB%98%E5%88%B6/" title="基于 GraphPad Prism 的火山图绘制"><img class="cover" src="/pic/cover.gif" data-original="/pic/prism17.png" alt="cover"><div class="content is-center"><div class="date"><i class="fas fa-history fa-fw"></i> 2023-10-09</div><div class="title">基于 GraphPad Prism 的火山图绘制</div></div></a></div></div></div><hr><div id="post-comment"><div class="comment-head"><div class="comment-headline"><i class="fas fa-comments fa-fw"></i><span> 评论</span></div></div><div class="comment-wrap"><div><div class="vcomment" id="vcomment"></div></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/pic/cover.gif" data-original="/pic/gitav.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"></div><div class="author-info__name">Juse</div><div class="author-info__description">偶尔徒步行进倒也不错</div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">文章</div><div class="length-num">57</div></a><a href="/tags/"><div class="headline">标签</div><div class="length-num">18</div></a><a href="/categories/"><div class="headline">分类</div><div class="length-num">10</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/JuseTiZ"><i class="fab fa-github"></i><span>Follow Juse</span></a><div class="card-info-social-icons is-center"><a class="social-icon" href="/qq.html" target="_blank" title=""><i class="fab fa-qq"></i></a><a class="social-icon" href="https://space.bilibili.com/31269335" target="_blank" title=""><i class="fab fa-bilibili"></i></a><a class="social-icon" href="https://github.com/JuseTiZ" target="_blank" title=""><i class="fab fa-github"></i></a><a class="social-icon" href="https://steamcommunity.com/id/291195857/" target="_blank" title=""><i class="fab fa-steam"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>公告</span></div><div class="announcement_content"><p align="center"><b>人生是旷野，不是轨道</b><br>⭐⭐🗞️ 新闻速递 📰⭐⭐<br>Juse 抽卡又保底了</p></div><div id="welcome-info"></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#PAML-%E7%AE%80%E4%BB%8B"><span class="toc-number">1.</span> <span class="toc-text">PAML 简介</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%9B%B4%E6%96%B0%E6%97%A5%E5%BF%97"><span class="toc-number">1.1.</span> <span class="toc-text">更新日志</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%AD%A3%E9%80%89%E6%8B%A9%E6%A3%80%E6%B5%8B%EF%BC%88%E6%94%AF%E4%BD%8D%E7%82%B9%E6%A8%A1%E5%9E%8B%EF%BC%89"><span class="toc-number">2.</span> <span class="toc-text">正选择检测（支位点模型）</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%8E%9F%E7%90%86%E5%8F%8A%E8%BF%90%E8%A1%8C"><span class="toc-number">2.1.</span> <span class="toc-text">原理及运行</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%B1%80%E9%99%90%E6%80%A7"><span class="toc-number">2.2.</span> <span class="toc-text">局限性</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#dN-dS-%E8%AE%A1%E7%AE%97%EF%BC%88%E8%87%AA%E7%94%B1%E6%AF%94%E6%A8%A1%E5%9E%8B%EF%BC%89"><span class="toc-number">3.</span> <span class="toc-text">dN / dS 计算（自由比模型）</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%8E%9F%E7%90%86%E5%8F%8A%E8%BF%90%E8%A1%8C-1"><span class="toc-number">3.1.</span> <span class="toc-text">原理及运行</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%89%B9%E9%87%8F%E6%8F%90%E5%8F%96%E7%BB%93%E6%9E%9C"><span class="toc-number">3.2.</span> <span class="toc-text">批量提取结果</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%BC%B0%E8%AE%A1%E7%89%A9%E7%A7%8D%E5%88%86%E6%AD%A7%E6%97%B6%E9%97%B4"><span class="toc-number">4.</span> <span class="toc-text">估计物种分歧时间</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%AE%9E%E6%88%98%E6%BC%94%E7%BB%83"><span class="toc-number">4.1.</span> <span class="toc-text">实战演练</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%A6%81%E8%A1%A5%E5%85%85%E7%9A%84%E4%B8%80%E4%BA%9B%E8%AF%9D"><span class="toc-number">4.2.</span> <span class="toc-text">要补充的一些话</span></a><ol class="toc-child"><li class="toc-item toc-level-4"><a class="toc-link" href="#%E4%BC%B0%E8%AE%A1%E6%9B%BF%E6%8D%A2%E7%8E%87%E7%9A%84%E6%96%B9%E6%B3%95"><span class="toc-number">4.2.1.</span> <span class="toc-text">估计替换率的方法</span></a></li></ol></li></ol></li></ol></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div id="ft"><div class="ft-item-1"><div class="t-top"><div class="t-t-l"><p class="ft-t t-l-t">想说的话😉</p><div class="bg-ad"><div>Genes are not destiny. While they play a large role in shaping how we act, how we feel, and how we interact with others, our environment and personal choices have just as much impact on our behavior.</div><div class="btn-xz-box"><a class="btn-xz" target="_blank" rel="noopener" href="https://habitica.com/static/home">游戏人生不游戏 🎯</a></div></div></div><div class="t-t-r"><p class="ft-t t-l-t">仙人指路😝</p><ul class="ft-links"><li><a target="_blank" rel="noopener" href="https://www.fomal.cc/">建站参考</a><a target="_blank" rel="noopener" href="https://emojixd.com/">表情大全</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/talk/">看看生活</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/messageboard/">我要留言</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/myself/">关于博主</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/archives/">文章归档</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/categories/">文章分类</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/tags/">文章标签</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/about/">有的没的</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/new/">更新日志</a></li><li><a target="_blank" rel="noopener" href="https://jusetiz.github.io/link/">友情链接</a><a target="_blank" rel="noopener" href="https://jusetiz.github.io/404.html">404页面</a></li></ul></div></div></div><div class="ft-item-2"><p class="ft-t">友人链接🔗</p><div class="ft-img-group"><div class="img-group-item"><a target="_blank" rel="noopener" href="https://linctanny.github.io/"><img src="/pic/cover.gif" data-original="https://linctanny.github.io/img/p7.jpeg" alt=""></a></div><div class="img-group-item"><a target="_blank" rel="noopener" href="https://robust-vase.github.io/"><img src="/pic/cover.gif" data-original="https://robust-vase.github.io/img/John_King.webp" alt=""></a></div><div class="img-group-item"><a target="_blank" rel="noopener" href="https://blog.stariverfeel.eu.org/"><img src="/pic/cover.gif" data-original="https://avatars.githubusercontent.com/u/26035995?v=4" alt=""></a></div></div></div></div><div class="copyright">©2022 - 2025  @ Juse</div><div class="framework-info"><span>框架 </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo </a><span class="footer-separator">|</span><span> 主题 </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly </a><span class="footer-separator">|</span><span> 部署 </span><a target="_blank" rel="noopener" href="https://github.com/">GitPage </a><span class="footer-separator">|</span><span> 加速 </span><a target="_blank" rel="noopener" href="https://dash.cloudflare.com/">Cloudflare </a><span class="footer-separator">|</span><span> </span><a href="https://icp.gov.moe/?keyword=20232400" target="_blank">萌ICP备20232400号</a></div><div id="workboard"></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="fas fa-book-open"></i></button><button id="translateLink" type="button" title="简繁转换">简</button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="设置"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="fas fa-list-ul"></i></button><a id="to_comment" href="#post-comment" title="直达评论"><i class="fas fa-comments"></i></a><button id="go-up" type="button" title="回到顶部"><i class="fas fa-arrow-up"></i></button></div></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="is-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"></div></div><hr><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div class="js-pjax" id="rightMenu"><div class="rightMenu-group rightMenu-small"><a class="rightMenu-item" href="javascript:window.history.back();"><i class="fa fa-arrow-left"></i></a><a class="rightMenu-item" href="javascript:window.history.forward();"><i class="fa fa-arrow-right"></i></a><a class="rightMenu-item" href="javascript:window.location.reload();"><i class="fa fa-refresh"></i></a><a class="rightMenu-item" href="/"><i class="fa fa-home"></i></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-text"><a class="rightMenu-item" href="javascript:rmf.copySelect();"><i class="fa fa-copy"></i><span>复制</span></a><a class="rightMenu-item" href="javascript:window.open(&quot;https://cn.bing.com/search?q=&quot;+window.getSelection().toString());window.location.reload();"><i class="fa fa-search"></i><span>必应搜索</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-too"><a class="rightMenu-item" href="javascript:window.open(window.getSelection().toString());window.location.reload();"><i class="fa fa-link"></i><span>转到链接</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-paste"><a class="rightMenu-item" href="javascript:rmf.paste()"><i class="fa fa-copy"></i><span>粘贴</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-post"><a class="rightMenu-item" href="#post-comment"><i class="fas fa-comment"></i><span>空降评论</span></a><a class="rightMenu-item" href="javascript:rmf.copyWordsLink()"><i class="fa fa-link"></i><span>复制本文地址</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-to"><a class="rightMenu-item" href="javascript:rmf.openWithNewTab()"><i class="fa fa-window-restore"></i><span>新窗口打开</span></a><a class="rightMenu-item" id="menu-too" href="javascript:rmf.open()"><i class="fa fa-link"></i><span>转到链接</span></a><a class="rightMenu-item" href="javascript:rmf.copyLink()"><i class="fa fa-copy"></i><span>复制链接</span></a></div><div class="rightMenu-group rightMenu-line hide" id="menu-img"><a class="rightMenu-item" href="javascript:rmf.saveAs()"><i class="fa fa-download"></i><span>保存图片</span></a><a class="rightMenu-item" href="javascript:rmf.openWithNewTab()"><i class="fa fa-window-restore"></i><span>在新窗口打开</span></a><a class="rightMenu-item" href="javascript:rmf.copyLink()"><i class="fa fa-copy"></i><span>复制图片链接</span></a></div><div class="rightMenu-group rightMenu-line"><a class="rightMenu-item" href="/about/"><i class="fa fa-sitemap"></i><span>有的没的</span></a><a class="rightMenu-item" href="/new/"><i class="fa fa-bell"></i><span>更新日志</span></a><a class="rightMenu-item" href="javascript:rmf.fullScreen();"><i class="fas fa-expand"></i><span>切换全屏</span></a><a class="rightMenu-item" href="javascript:window.print();"><i class="fa-solid fa-print"></i><span>打印页面</span></a></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="/js/tw_cn.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox.umd.min.js"></script><script src="/js/search/local-search.js"></script><script>var preloader = {
  endLoading: () => {
    document.body.style.overflow = 'auto';
    document.getElementById('loading-box').classList.add("loaded")
  },
  initLoading: () => {
    document.body.style.overflow = '';
    document.getElementById('loading-box').classList.remove("loaded")

  }
}
window.addEventListener('load',preloader.endLoading())</script><div class="js-pjax"><script>function loadValine () {
  function initValine () {
    const valine = new Valine(Object.assign({
      el: '#vcomment',
      appId: 'ORERCz6nkgV8OYWw8FovgBhQ-gzGzoHsz',
      appKey: 'BNxukaAtCZZ0K6TZldZpkY9e',
      avatar: 'mp',
      placeholder: '🐕温馨提示本站不具有邮件提示功能，Juse每隔一段时间会在后台检查评论并回复',
      serverURLs: '',
      emojiMaps: "",
      path: window.location.pathname,
      visitor: false
    }, null))
  }

  if (typeof Valine === 'function') initValine() 
  else getScript('https://cdn.jsdelivr.net/npm/valine/dist/Valine.min.js').then(initValine)
}

if ('Valine' === 'Valine' || !true) {
  if (true) btf.loadComment(document.getElementById('vcomment'),loadValine)
  else setTimeout(loadValine, 0)
} else {
  function loadOtherComment () {
    loadValine()
  }
}</script></div><script src="https://cdn.staticfile.org/jquery/3.6.3/jquery.min.js"></script><script type="text/javascript" src="/js/rightmenu.js"></script><script type="text/javascript" src="/js/txmap.js"></script><script type="text/javascript" src="/js/copy12.js"></script><script type="text/javascript" src="/js/ldex.js"></script><script type="text/javascript" src="/js/runtime.js"></script><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/dist/activate-power-mode.min.js"></script><script>POWERMODE.colorful = true;
POWERMODE.shake = false;
POWERMODE.mobile = true;
document.body.addEventListener('input', POWERMODE);
</script><script async="" data-pjax="" src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script></div>
        <style>
            [bg-lazy] {
                background-image: none !important;
                background-color: #eee !important;
            }
        </style>
        <script>
            window.imageLazyLoadSetting = {
                isSPA: false,
                preloadRatio: 1,
                processImages: null,
            };
        </script><script>window.addEventListener("load",function(){var t=/\.(gif|jpg|jpeg|tiff|png)$/i,r=/^data:image\/[a-z]+;base64,/;Array.prototype.slice.call(document.querySelectorAll("img[data-original]")).forEach(function(a){var e=a.parentNode;"A"===e.tagName&&(e.href.match(t)||e.href.match(r))&&(e.href=a.dataset.original)})});</script><script>!function(r){r.imageLazyLoadSetting.processImages=t;var e=r.imageLazyLoadSetting.isSPA,n=r.imageLazyLoadSetting.preloadRatio||1,c=a();function a(){var t=Array.prototype.slice.call(document.querySelectorAll("img[data-original]")),e=Array.prototype.slice.call(document.querySelectorAll("[bg-lazy]"));return t.concat(e)}function t(){e&&(c=a());for(var t,o=0;o<c.length;o++)0<=(t=(t=c[o]).getBoundingClientRect()).bottom&&0<=t.left&&t.top<=(r.innerHeight*n||document.documentElement.clientHeight*n)&&function(){var t,e,n,a,i=c[o];e=function(){c=c.filter(function(t){return i!==t}),r.imageLazyLoadSetting.onImageLoaded&&r.imageLazyLoadSetting.onImageLoaded(i)},(t=i).hasAttribute("bg-lazy")?(t.removeAttribute("bg-lazy"),e&&e()):(n=new Image,a=t.getAttribute("data-original"),n.onload=function(){t.src=a,t.removeAttribute("data-original"),e&&e()},t.src!==a&&(n.src=a))}()}function i(){clearTimeout(t.tId),t.tId=setTimeout(t,500)}t(),document.addEventListener("scroll",i),r.addEventListener("resize",i),r.addEventListener("orientationchange",i)}(this);</script><script type="text/javascript" charset="utf-8" src="/js/lazyload-plugin/lazyload.intersectionObserver.min.js"></script></body></html>